{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "6df9b4d2-a8df-49d4-ae19-bc4342520ac9",
   "metadata": {},
   "source": [
    "# Simulation experiment for detection thresholds for variant calling\n",
    "\n",
    "This notebook is to get detection thresholds for the limit of the variant calling based on: \n",
    "\n",
    "1. Read depth\n",
    "2. Error rate\n",
    "3. Number of mutations\n",
    "\n",
    "\n",
    "Also optionaly get the time it takes to complete experiments per sequence length.\n",
    "\n",
    "\n",
    "```\n",
    ">lcl|NC_000913.3_cds_NP_414902.1_362 [gene=tauD] [locus_tag=b0368] [db_xref=UniProtKB/Swiss-Prot:P37610] [protein=taurine dioxygenase] [protein_id=NP_414902.1] [location=387795..388646] [gbkey=CDS]\n",
    "```\n",
    "\n",
    "From: ecoli_GCF_000005845.2_ASM584v2_cds_from_genomic.fna\n",
    "and >NP_414902.1 taurine dioxygenase [Escherichia coli str. K-12 substr. MG1655]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "f774d659-f374-4432-8493-4cace2f22133",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "# Visualisation things to make the figures look nice\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from sciutil import SciUtil\n",
    "\n",
    "\n",
    "data_dir = 'ePCR_data/'\n",
    "fig_dir = 'ePCR_figures/'\n",
    "\n",
    "\n",
    "plt.rcParams['svg.fonttype'] = 'none'\n",
    "axis_line_width = 1.0\n",
    "axis_font_size = 12\n",
    "title_font_size = 12\n",
    "label_font_size = 10\n",
    "figsize = (5, 4) # Figure size\n",
    "font = 'Arial'\n",
    "style = 'ticks'\n",
    "font_family = 'sans-serif'\n",
    "\n",
    "# Create a figure with 2x2 subplots\n",
    "sns.set_style(\"whitegrid\")\n",
    "cmap = 'viridis'\n",
    "palette = sns.color_palette(\"viridis\", as_cmap=True)\n",
    "sns.set_palette(cmap)\n",
    "\n",
    "sns.set(rc={'figure.figsize': figsize, 'font.family': font_family,\n",
    "            'font.sans-serif': font, 'font.size': label_font_size}, style=style)\n",
    "\n",
    "def set_ax_params(ax):\n",
    "    ax.tick_params(direction='out', length=2, width=axis_line_width)\n",
    "    ax.spines['bottom'].set_linewidth(axis_line_width)\n",
    "    ax.spines['top'].set_linewidth(0)\n",
    "    ax.spines['left'].set_linewidth(axis_line_width)\n",
    "    ax.spines['right'].set_linewidth(0)\n",
    "    ax.tick_params(labelsize=axis_font_size)\n",
    "    ax.tick_params(axis='x', which='major', pad=2.0)\n",
    "    ax.tick_params(axis='y', which='major', pad=2.0)\n",
    "    ax.set_xticklabels(ax.get_xticklabels(), rotation=45, horizontalalignment='right')\n",
    "    ax.tick_params(labelsize=label_font_size)\n",
    "\n",
    "\n",
    "u = SciUtil()\n",
    "\n",
    "# Generate mutations on a \"real\" sequence\n",
    "# Decided to choose tauD from Ecoli K12\n",
    "parent_sequence_aa = 'MSERLSITPLGPYIGAQISGADLTRPLSDNQFEQLYHAVLRHQVVFLRDQAITPQQQRALAQRFGELHIHPVYPHAEGVD\\\n",
    "EIIVLDTHNDNPPDNDNWHTDVTFIETPPAGAILAAKELPSTGGDTLWTSGIAAYEALSVPFRQLLSGLRAEHDFRKSFP\\\n",
    "EYKYRKTEEEHQRWREAVAKNPPLLHPVVRTHPVSGKQALFVNEGFTTRIVDVSEKESEALLSFLFAHITKPEFQVRWRW\\\n",
    "QPNDIAIWDNRVTQHYANADYLPQRRIMHRATILGDKPFYRAG'\n",
    "\n",
    "parent_sequence = 'ATGAGTGAACGTCTGAGCATTACCCCGCTGGGGCCGTATATCGGCGCACAAATTTCGGGTGCCGACCTGACGCGCCCGTT\\\n",
    "AAGCGATAATCAGTTTGAACAGCTTTACCATGCGGTGCTGCGCCATCAGGTGGTGTTTCTACGCGATCAAGCTATTACGC\\\n",
    "CGCAGCAGCAACGCGCGCTGGCCCAGCGTTTTGGCGAATTGCATATTCACCCTGTTTACCCGCATGCCGAAGGGGTTGAC\\\n",
    "GAGATCATCGTGCTGGATACCCATAACGATAATCCGCCAGATAACGACAACTGGCATACCGATGTGACATTTATTGAAAC\\\n",
    "GCCACCCGCAGGGGCGATTCTGGCAGCTAAAGAGTTACCTTCGACCGGCGGTGATACGCTCTGGACCAGCGGTATTGCGG\\\n",
    "CCTATGAGGCGCTCTCTGTTCCCTTCCGCCAGCTGCTGAGTGGGCTGCGTGCGGAGCATGATTTCCGTAAATCGTTCCCG\\\n",
    "GAATACAAATACCGCAAAACCGAGGAGGAACATCAACGCTGGCGCGAGGCGGTCGCGAAAAACCCGCCGTTGCTACATCC\\\n",
    "GGTGGTGCGAACGCATCCGGTGAGCGGTAAACAGGCGCTGTTTGTGAATGAAGGCTTTACTACGCGAATTGTTGATGTGA\\\n",
    "GCGAGAAAGAGAGCGAAGCCTTGTTAAGTTTTTTGTTTGCCCATATCACCAAACCGGAGTTTCAGGTGCGCTGGCGCTGG\\\n",
    "CAACCAAATGATATTGCGATTTGGGATAACCGCGTGACCCAGCACTATGCCAATGCCGATTACCTGCCACAGCGACGGAT\\\n",
    "AATGCATCGGGCGACGATCCTTGGGGATAAACCGTTTTATCGGGCGGGGTAA'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eabfb04e-5fc0-4ef5-a62e-c37ec6835e92",
   "metadata": {},
   "source": [
    "# Introduce mutations at a given frequency and an error rate\n",
    "\n",
    "Test setup:\n",
    "\n",
    "1. For number of mutations from 1 to the sequence length test mutating each one and correlate this to the p value\n",
    "2. For sequencing error rates from 0 to 100% make sequences ranging with this and see what the results are for the p value and the error\n",
    "3. For different sequence lengths also check how sequence length corresponds to the pvalue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "8928921b-8b9b-4a86-9f7b-9874071e7a15",
   "metadata": {},
   "outputs": [],
   "source": [
    "from minION import *\n",
    "from tqdm import tqdm\n",
    "\n",
    "def make_experiment(run_label, read_depth, sequencing_error_rate, parent_sequence, positions, library_number, \n",
    "                    number_of_wells, epcr_mutation_rate, frequency_cutoff=0.5):\n",
    "    # Make a full experiment setup\n",
    "    mutated_sequence = make_epcr_de_experiment(read_depth, sequencing_error_rate, parent_sequence, library_number,\n",
    "                                                   epcr_mutation_rate)\n",
    "    \n",
    "    variant_df = get_dummy_plate_df(run_label, 'Well', number_of_wells)  # i.e. 20 - 1 since we don't have the other variant    \n",
    "    mutant_to_well_df = {}\n",
    "    current_well = 0\n",
    "    for mutant in tqdm(mutated_sequence):\n",
    "        parent_name = 'Parent'\n",
    "        reads = []\n",
    "        read_ids = []\n",
    "        quals = []\n",
    "        for i, seq in enumerate(mutated_sequence[mutant]):\n",
    "            read_ids.append(f'read_{i}')\n",
    "            reads.append(seq)\n",
    "            quals.append(100)  # Dummy don't need\n",
    "    \n",
    "        well_df = make_well_df_from_reads(reads, read_ids, quals)\n",
    "        rows_all = make_row_from_read_pileup_across_well(well_df, parent_sequence, parent_name)\n",
    "        well_df = pd.DataFrame(rows_all)\n",
    "        well_df.columns = ['gene_name', 'position', 'ref', 'most_frequent', 'freq_non_ref', 'total_other',\n",
    "                           'total_reads', 'p_value', 'percent_most_freq_mutation', 'A', 'p(a)', 'T', 'p(t)', 'G',\n",
    "                           'p(g)',\n",
    "                           'C', 'p(c)', 'N', 'p(n)']\n",
    "        well_df = calculate_mutation_significance_across_well(well_df)\n",
    "        label, frequency, combined_p_value, mixed_well = get_variant_label_for_well(well_df, frequency_cutoff)\n",
    "        # This should be mutated at 100% - the rate of our sequencing errror\n",
    "        if verbose:\n",
    "            u.dp([\"label\", label, f\"frequency\", frequency, f\"combined_p_value\", combined_p_value, \"mixed_well\", mixed_well])\n",
    "        mutant_to_well_df[f'{mutant}_{current_well}'] = well_df\n",
    "        variant_df.at[current_well, \"Mixed Well\"] = mixed_well\n",
    "        variant_df.at[current_well, \"Variant\"] = label\n",
    "        variant_df.at[current_well, \"frequency\"] = frequency\n",
    "        variant_df.at[current_well, \"P value\"] = combined_p_value\n",
    "        variant_df.at[current_well, \"Well\"] = f'Well {current_well}'\n",
    "        variant_df.at[current_well, \"Alignment_count\"] = read_depth\n",
    "        current_well += 1\n",
    "\n",
    "    # Before returning adjust the pvalues\n",
    "    variant_df['P adj.'] = multipletests(list(variant_df[\"P value\"].values), alpha=0.05, method='fdr_bh')[1]\n",
    "    return variant_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bcb70656-53c2-4d2c-9c1c-094a33b5cac0",
   "metadata": {},
   "source": [
    "## Experiment 1: Varying the sequencing error rate for a single mutation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "88a93209-a3f8-4121-b3d5-7182dba919e3",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|█| 96/96 [00:40<00:00,  2.38i\n",
      "  0%|      | 0/96 [00:00<?, ?it/s]/var/folders/gq/6ljhmvm1713fykdjqbl188pm0000gn/T/ipykernel_28643/2604269417.py:38: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '0.954' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  variant_df.at[current_well, \"frequency\"] = frequency\n",
      "100%|█| 96/96 [00:53<00:00,  1.80i\n",
      "  0%|      | 0/96 [00:00<?, ?it/s]/var/folders/gq/6ljhmvm1713fykdjqbl188pm0000gn/T/ipykernel_28643/2604269417.py:38: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '0.9079999999999999' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  variant_df.at[current_well, \"frequency\"] = frequency\n",
      "100%|█| 96/96 [00:58<00:00,  1.65i\n",
      "  0%|      | 0/96 [00:00<?, ?it/s]/var/folders/gq/6ljhmvm1713fykdjqbl188pm0000gn/T/ipykernel_28643/2604269417.py:38: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '0.845625' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  variant_df.at[current_well, \"frequency\"] = frequency\n",
      " 41%|▍| 39/96 [00:23<00:36,  1.58i"
     ]
    }
   ],
   "source": [
    "# We're going to make an experiment of 10 plates with different sequencing error rates\n",
    "read_depth = 100\n",
    "number_of_wells = 19\n",
    "epcr_mutation_rate = 0.02\n",
    "frequency_cutoff=0.5\n",
    "library_number = 96 # Usually do a 96 well plate\n",
    "verbose = False\n",
    "for position in [1, 20, 100]:\n",
    "    experiment_df = pd.DataFrame()\n",
    "    positions = [position]\n",
    "    for sequencing_error in range(0, 100, 5):\n",
    "        sequencing_error_rate = sequencing_error/100.0\n",
    "        run_df = make_experiment(f'SeqError_{sequencing_error}', read_depth, sequencing_error_rate, parent_sequence, positions, \n",
    "                                 library_number, number_of_wells, epcr_mutation_rate, frequency_cutoff)\n",
    "        run_df.reset_index(inplace=True)\n",
    "        experiment_df = pd.concat([experiment_df, run_df])\n",
    "\n",
    "    # Also plot each one\n",
    "    experiment_df.to_csv(f'{data_dir}Experiment1_position:{position}.csv', index=False)\n",
    "    experiment_df['-log10(P.adj)'] = -1*np.log10(experiment_df['P adj.'].values)\n",
    "    experiment_df['Error rate'] = [int(c.split('_')[-1])/100.0 for c in experiment_df['Plate']]\n",
    "    experiment_df['# Mutations'] = [len(v.split('_')) for v in experiment_df['Variant'].values]\n",
    "    experiment_df['# Mutations'] = [f'{v}' if v < 5 else f'>=5' for v in experiment_df['# Mutations']]\n",
    "    experiment_df = experiment_df.reset_index()\n",
    "    ax = sns.scatterplot(experiment_df, x='frequency', y='-log10(P.adj)', hue='Error rate', palette='viridis', \n",
    "                         style=\"# Mutations\", style_order=['1', '2', '3', '4', '>=5'], s=80)\n",
    "    set_ax_params(ax)\n",
    "    plt.xlabel('Mutation frequency')\n",
    "    plt.ylabel('-log10(padj)')\n",
    "    plt.title('Effect of sequencing error on significance', fontsize=title_font_size, fontweight=\"bold\")\n",
    "    plt.legend(bbox_to_anchor=(1.05, 1), loc=2, borderaxespad=0.)\n",
    "    plt.savefig(f'{fig_dir}Experiment1_position:{position}.svg')\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "72beba50-d357-4b88-90d5-ff9897c2f9a5",
   "metadata": {},
   "source": [
    "## Experiment 2: Varying the sequencing error rate for a multiple mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6aa7af5f-ba65-4d6c-a55f-0b1517049818",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "verbose = False\n",
    "for positions in [[50, 14, 90]]:\n",
    "    experiment_df = pd.DataFrame()\n",
    "    positions = positions\n",
    "    number_of_wells = (19*3)  # 19 aas * 3 positions\n",
    "    for sequencing_error in range(0, 100, 10):\n",
    "        sequencing_error_rate = sequencing_error/100.0\n",
    "        run_df = make_experiment(f'SeqError_{sequencing_error}', read_depth, sequencing_error_rate, parent_sequence, positions, \n",
    "                                 library_number, number_of_wells, epcr_mutation_rate, frequency_cutoff)\n",
    "        run_df.reset_index(inplace=True)\n",
    "        experiment_df = pd.concat([experiment_df, run_df])\n",
    "\n",
    "    # Also plot each one\n",
    "    positions = [str(s) for s in positions]\n",
    "    experiment_df.to_csv(f'{data_dir}Experiment2_position:{\"-\".join(positions)}.csv', index=False)\n",
    "    # experiment_df['-log10(P.adj)'] = -1*np.log10(experiment_df['P adj.'].values)\n",
    "    # experiment_df['Error rate'] = [int(c.split('_')[-1])/100.0 for c in experiment_df['Plate']]\n",
    "    # experiment_df['# Mutations'] = [len(v.split('_')) for v in experiment_df['Variant'].values]\n",
    "    # experiment_df['# Mutations'] = [f'{v}' if v < 5 else f'>=5' for v in experiment_df['# Mutations']]\n",
    "    # experiment_df = experiment_df.reset_index()\n",
    "    # ax = sns.scatterplot(experiment_df, x='frequency', y='-log10(P.adj)', hue='Error rate', palette='viridis', \n",
    "    #                      style=\"# Mutations\", style_order=['1', '2', '3', '4', '>=5'], s=80)\n",
    "    # set_ax_params(ax)\n",
    "    # plt.xlabel('Mutation frequency')\n",
    "    # plt.ylabel('-log10(padj)')\n",
    "    # plt.title('Effect of sequencing error on significance', fontsize=title_font_size, fontweight=\"bold\")\n",
    "    # plt.legend(bbox_to_anchor=(1.05, 1), loc=2, borderaxespad=0.)\n",
    "    # plt.savefig(f'{fig_dir}Experiment2_positions:{\"-\".join(positions)}.svg')\n",
    "    # plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ebf4be57-5c55-47d9-959f-a2f44e50426a",
   "metadata": {},
   "source": [
    "## Experiment 3: varying read depth and it's effect on significance (1 position)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "60ea640d-45e3-4524-b519-e74d1f6f2592",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# We're going to make an experiment of 10 plates with different sequencing error rates\n",
    "sequencing_error_rate = 0.1\n",
    "number_of_wells = 19\n",
    "\n",
    "for position in [1, 20, 100]:\n",
    "    positions = [position]\n",
    "    experiment_df = pd.DataFrame()\n",
    "    for read_depth in range(5, 100, 5):\n",
    "        print(sequencing_error)\n",
    "        run_df = make_experiment(f'ReadDepth_{read_depth}', read_depth, sequencing_error_rate, parent_sequence, positions, \n",
    "                         library_number, number_of_wells, epcr_mutation_rate, frequency_cutoff)\n",
    "        run_df.reset_index(inplace=True)\n",
    "        experiment_df = pd.concat([experiment_df, run_df])\n",
    "    \n",
    "     # Also plot each one\n",
    "    experiment_df.to_csv(f'{data_dir}Experiment3_position:{position}.csv', index=False)\n",
    "    experiment_df['-log10(P.adj)'] = -1*np.log10(experiment_df['P adj.'].values)\n",
    "    experiment_df['Read depth'] = [int(c.split('_')[-1]) for c in experiment_df['Plate']]\n",
    "    experiment_df['# Mutations'] = [len(v.split('_')) for v in experiment_df['Variant'].values]\n",
    "    experiment_df['# Mutations'] = [f'{v}' if v < 5 else f'>=5' for v in experiment_df['# Mutations']]\n",
    "    experiment_df = experiment_df.reset_index()\n",
    "    ax = sns.scatterplot(experiment_df, x='frequency', y='-log10(P.adj)', hue='Read depth', palette='viridis', \n",
    "                         style=\"# Mutations\", style_order=['1', '2', '3', '4', '>=5'], s=80)\n",
    "    set_ax_params(ax)\n",
    "    plt.xlabel('Mutation frequency')\n",
    "    plt.ylabel('-log10(padj)')\n",
    "    plt.title('Effect of read depth on significance', fontsize=title_font_size, fontweight=\"bold\")\n",
    "    plt.legend(bbox_to_anchor=(1.05, 1), loc=2, borderaxespad=0.)\n",
    "    plt.savefig(f'{fig_dir}Experiment3_position:{position}.svg')\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f62040aa-85bd-4bf6-a4cc-d36f8b90a3e4",
   "metadata": {},
   "source": [
    "## Experiment 4: effect of sequence length on significance\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "af0a7dd6-a1ad-4fba-ad8d-fd5ca61b4098",
   "metadata": {},
   "outputs": [],
   "source": [
    "# We're going to make an experiment of 10 plates with different sequencing error rates\n",
    "sequencing_error_rate = 0.1\n",
    "number_of_wells = 19\n",
    "read_depth = 25\n",
    "\n",
    "for position in [1, 2, 3]:\n",
    "    positions = [position]\n",
    "    experiment_df = pd.DataFrame()\n",
    "    for seq_len in range(5, 200, 20):\n",
    "        print(sequencing_error)\n",
    "        run_df = make_experiment(f'SeqLen_{seq_len}', read_depth, sequencing_error_rate, parent_sequence[:seq_len*3], positions, \n",
    "                 library_number, number_of_wells, epcr_mutation_rate, frequency_cutoff)\n",
    "        run_df.reset_index(inplace=True)\n",
    "        experiment_df = pd.concat([experiment_df, run_df])\n",
    "    \n",
    "     # Also plot each one\n",
    "    experiment_df.to_csv(f'{data_dir}Experiment4_position:{position}.csv', index=False)\n",
    "    experiment_df['-log10(P.adj)'] = -1*np.log10(experiment_df['P adj.'].values)\n",
    "    experiment_df['Seq length'] = [int(c.split('_')[-1]) for c in experiment_df['Plate']]\n",
    "    experiment_df['# Mutations'] = [len(v.split('_')) for v in experiment_df['Variant'].values]\n",
    "    experiment_df['# Mutations'] = [f'{v}' if v < 5 else f'>=5' for v in experiment_df['# Mutations']]\n",
    "    experiment_df = experiment_df.reset_index()\n",
    "    ax = sns.scatterplot(experiment_df, x='frequency', y='-log10(P.adj)', hue='Seq length', palette='viridis', \n",
    "                         style=\"# Mutations\", style_order=['1', '2', '3', '4', '>=5'], s=80)\n",
    "    set_ax_params(ax)\n",
    "    plt.xlabel('Mutation frequency')\n",
    "    plt.ylabel('-log10(padj)')\n",
    "    plt.title('Effect of sequence length on significance', fontsize=title_font_size, fontweight=\"bold\")\n",
    "    plt.legend(bbox_to_anchor=(1.05, 1), loc=2, borderaxespad=0.)\n",
    "    plt.savefig(f'{fig_dir}Experiment4_position:{position}.svg')\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f98f649b-3a5a-4cbf-8f72-241c5a73553d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# We're going to make an experiment of 10 plates with different sequencing error rates\n",
    "sequencing_error_rate = 0.1\n",
    "number_of_wells = 19\n",
    "read_depth = 25\n",
    "\n",
    "for position in [1, 20, 100]:\n",
    "    positions = [position]\n",
    "    experiment_df = pd.DataFrame()\n",
    "    for frequency_cutoff in range(10, 90, 10):\n",
    "        run_df = make_experiment(f'FreqCutoff_{frequency_cutoff}', read_depth, sequencing_error_rate, parent_sequence, positions, library_number, number_of_wells, epcr_mutation_rate, frequency_cutoff/100.0)\n",
    "        run_df.reset_index(inplace=True)\n",
    "        experiment_df = pd.concat([experiment_df, run_df])\n",
    "    \n",
    "     # Also plot each one\n",
    "    experiment_df.to_csv(f'{data_dir}Experiment5_position:{position}.csv', index=False)\n",
    "    # experiment_df['-log10(P.adj)'] = -1*np.log10(experiment_df['P adj.'].values)\n",
    "    # experiment_df['Seq length'] = [int(c.split('_')[-1]) for c in experiment_df['Plate']]\n",
    "    # experiment_df['# Mutations'] = [len(v.split('_')) for v in experiment_df['Variant'].values]\n",
    "    # experiment_df['# Mutations'] = [f'{v}' if v < 5 else f'>=5' for v in experiment_df['# Mutations']]\n",
    "    # experiment_df = experiment_df.reset_index()\n",
    "    # ax = sns.scatterplot(experiment_df, x='frequency', y='-log10(P.adj)', hue='Seq length', palette='viridis', \n",
    "    #                      style=\"# Mutations\", style_order=['1', '2', '3', '4', '>=5'], s=80)\n",
    "    # set_ax_params(ax)\n",
    "    # plt.xlabel('Mutation frequency')\n",
    "    # plt.ylabel('-log10(padj)')\n",
    "    # plt.title('Effect of frequency cutoff on significance', fontsize=title_font_size, fontweight=\"bold\")\n",
    "    # plt.legend(bbox_to_anchor=(1.05, 1), loc=2, borderaxespad=0.)\n",
    "    # plt.savefig(f'{fig_dir}Experiment5_position:{position}.svg')\n",
    "    # plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "42728d5e-515c-4fbc-9092-469544b08a4f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# We're going to make an experiment of 10 plates with different sequencing error rates\n",
    "sequencing_error_rate = 0.1\n",
    "number_of_wells = 19\n",
    "read_depth = 25\n",
    "frequency_cutoff = 0.5 \n",
    "\n",
    "for position in [1, 20, 100]:\n",
    "    positions = [position]\n",
    "    experiment_df = pd.DataFrame()\n",
    "    for epcr_mutation_rate in range(1, 10, 1):\n",
    "        run_df = make_experiment(f'ePCR_{epcr_mutation_rate}', read_depth, sequencing_error_rate, parent_sequence, positions, library_number, number_of_wells, epcr_mutation_rate/100.0, frequency_cutoff)\n",
    "        run_df.reset_index(inplace=True)\n",
    "        experiment_df = pd.concat([experiment_df, run_df])\n",
    "    \n",
    "     # Also plot each one\n",
    "    experiment_df.to_csv(f'{data_dir}Experiment6_position:{position}.csv', index=False)\n",
    "    # experiment_df['-log10(P.adj)'] = -1*np.log10(experiment_df['P adj.'].values)\n",
    "    # experiment_df['Seq length'] = [int(c.split('_')[-1]) for c in experiment_df['Plate']]\n",
    "    # experiment_df['# Mutations'] = [len(v.split('_')) for v in experiment_df['Variant'].values]\n",
    "    # experiment_df['# Mutations'] = [f'{v}' if v < 5 else f'>=5' for v in experiment_df['# Mutations']]\n",
    "    # experiment_df = experiment_df.reset_index()\n",
    "    # ax = sns.scatterplot(experiment_df, x='frequency', y='-log10(P.adj)', hue='Seq length', palette='viridis', \n",
    "    #                      style=\"# Mutations\", style_order=['1', '2', '3', '4', '>=5'], s=80)\n",
    "    # set_ax_params(ax)\n",
    "    # plt.xlabel('Mutation frequency')\n",
    "    # plt.ylabel('-log10(padj)')\n",
    "    # plt.title('Effect of frequency cutoff on significance', fontsize=title_font_size, fontweight=\"bold\")\n",
    "    # plt.legend(bbox_to_anchor=(1.05, 1), loc=2, borderaxespad=0.)\n",
    "    # plt.savefig(f'{fig_dir}Experiment5_position:{position}.svg')\n",
    "    # plt.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
