{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Try to implement bayes based variant calling\n",
    "\n",
    "- For Parent first\n",
    "- Check for more mutatios later"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import\n",
    "import sys\n",
    "sys.path.append(\"/home/emre/github_repo/MinION\")\n",
    "from minION.util import IO_processor\n",
    "from minION import analyser\n",
    "from minION import consensus\n",
    "\n",
    "import importlib\n",
    "importlib.reload(analyser)\n",
    "importlib.reload(consensus)\n",
    "importlib.reload(IO_processor)\n",
    "from pathlib import Path\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import os\n",
    "import glob\n",
    "import numpy as np\n",
    "from Bio import SeqIO\n",
    "import matplotlib.pyplot as plt\n",
    "import gzip\n",
    "import math\n",
    "import re\n",
    "import pickle\n",
    "import itertools\n",
    "import pysam\n",
    "import subprocess"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_bases_from_pileup_simulation(bam_file, chrom, positions):\n",
    "    bases_dict = {position: {} for position in positions}\n",
    "    \n",
    "    with pysam.AlignmentFile(bam_file, 'rb') as bam:\n",
    "        for pileup_column in bam.pileup(chrom, min(positions) - 1, max(positions) + 1,\n",
    "                                        min_base_quality=0, \n",
    "                                        min_mapping_quality=0, \n",
    "                                        truncate=True):\n",
    "            pos = pileup_column.pos + 1\n",
    "            if pos in positions:\n",
    "                for pileup_read in pileup_column.pileups:\n",
    "                    read_name = pileup_read.alignment.query_name\n",
    "\n",
    "                    # Handle deletions\n",
    "                    if pileup_read.is_del:\n",
    "                        base = '-'  # Symbol to represent a deletion\n",
    "                    elif not pileup_read.is_refskip:\n",
    "                        base = pileup_read.alignment.query_sequence[pileup_read.query_position]\n",
    "                    else:\n",
    "                        continue\n",
    "\n",
    "                    # Add base to the dictionary\n",
    "                    if read_name not in bases_dict[pos]:\n",
    "                        bases_dict[pos][read_name] = base\n",
    "\n",
    "    # Get unique read names and sort them\n",
    "    read_names = sorted(set().union(*[bases_dict[pos].keys() for pos in bases_dict]))\n",
    "\n",
    "    # Create DataFrame for bases\n",
    "    df_bases = pd.DataFrame(index=read_names, columns=positions)\n",
    "    \n",
    "    # Populate DataFrame\n",
    "    for pos in positions:\n",
    "        for read_name in bases_dict[pos]:\n",
    "            df_bases.at[read_name, pos] = bases_dict[pos][read_name]\n",
    "    \n",
    "    # Fill NaN with \"-\"\n",
    "    df_bases = df_bases.fillna(\"-\")\n",
    "\n",
    "    return df_bases\n",
    "\n",
    "\n",
    "\n",
    "def get_bases_from_pileup(bam_file, chrom, positions):\n",
    "    bases_dict = {position: {} for position in positions}\n",
    "    qualities_dict = {position: {} for position in positions}\n",
    "    \n",
    "    with pysam.AlignmentFile(bam_file, 'rb') as bam:\n",
    "        for pileup_column in bam.pileup(chrom, min(positions) - 1, max(positions) + 1,\n",
    "                                        min_base_quality=0, \n",
    "                                        min_mapping_quality=0, \n",
    "                                        truncate=True):\n",
    "            pos = pileup_column.pos + 1\n",
    "            if pos in positions:\n",
    "                for pileup_read in pileup_column.pileups:\n",
    "                    read_name = pileup_read.alignment.query_name\n",
    "\n",
    "                    # Handle deletions\n",
    "                    if pileup_read.is_del:\n",
    "                        base = '-'  # or any symbol you prefer to represent a deletion\n",
    "                        quality = 0  # Assign a default quality for deletions\n",
    "                    elif not pileup_read.is_refskip:\n",
    "                        base = pileup_read.alignment.query_sequence[pileup_read.query_position]\n",
    "                        quality = pileup_read.alignment.query_qualities[pileup_read.query_position]\n",
    "                    else:\n",
    "                        continue\n",
    "\n",
    "                    # Add base and quality to the dictionaries\n",
    "                    if read_name not in bases_dict[pos]:\n",
    "                        bases_dict[pos][read_name] = base\n",
    "                        qualities_dict[pos][read_name] = quality\n",
    "\n",
    "    # Get unique read names and sort them\n",
    "    read_names = sorted(set().union(*[bases_dict[pos].keys() for pos in bases_dict]))\n",
    "\n",
    "    # Create DataFrames\n",
    "    df_bases = pd.DataFrame(index=read_names, columns=positions)\n",
    "    df_qualities = pd.DataFrame(index=read_names, columns=positions)\n",
    "    \n",
    "    # Populate DataFrames\n",
    "    for pos in positions:\n",
    "        for read_name in bases_dict[pos]:\n",
    "            df_bases.at[read_name, pos] = bases_dict[pos][read_name]\n",
    "            df_qualities.at[read_name, pos] = qualities_dict[pos][read_name]\n",
    "    \n",
    "    # Fill NaN with \"-\" for bases and 0 for qualities\n",
    "    df_bases = df_bases.fillna(\"-\")\n",
    "    df_qualities = df_qualities.fillna(10) # 10 is the lowest quality filter we used for filtering\n",
    "\n",
    "    return df_bases, df_qualities\n",
    "\n",
    "\n",
    "def get_soft_pop_frequency(bam_file, template, reference, nb_positions, min_depth = 5):\n",
    "\n",
    "    # Min depth based on the alphabet size\n",
    "    \n",
    "    # Check also for variant by sampling random positions\n",
    "\n",
    "    bases_df = get_bases_from_pileup(bam_file, reference, nb_positions)\n",
    "    \n",
    "    frequency_df = bases_df.apply(get_variant_name, axis=1, args=(template, nb_positions)).value_counts().reset_index()\n",
    "\n",
    "    frequency_df.columns = ['Population', 'N_reads']\n",
    "    \n",
    "    frequency_df[\"Frequency\"] = frequency_df[\"N_reads\"] / frequency_df[\"N_reads\"].sum()\n",
    "\n",
    "\n",
    "    # Filter for frequency > 0.4 and depth > 15\n",
    "    frequency_df = frequency_df[(frequency_df[\"Frequency\"] > min_freq) & (frequency_df[\"N_reads\"] > min_depth)]\n",
    "\n",
    "    return frequency_df\n",
    "\n",
    "\n",
    "def add_neighbouring_positions(positions, nb_neighbours, max_index):\n",
    "    if positions is np.nan:\n",
    "        return np.nan\n",
    "    \n",
    "    elif isinstance(positions, int):\n",
    "        positions = [positions]\n",
    "\n",
    "    new_positions = []\n",
    "    for position in positions:\n",
    "        for new_pos in range(position - nb_neighbours, position + nb_neighbours + 1):\n",
    "            if 1 <= new_pos <= max_index:  # Check if the new position is within valid index range\n",
    "                new_positions.append(new_pos)\n",
    "    return sorted(set(new_positions))\n",
    "\n",
    "def calculate_mean_quality_for_reads(bases_df, qual_df, nb_positions, nb_neighbours):\n",
    "    if isinstance(nb_positions, int):\n",
    "        nb_positions = [nb_positions]  # Convert single integer to a list\n",
    "\n",
    "    read_mean_qualities = {}\n",
    "\n",
    "    for nb_position in nb_positions:\n",
    "        if nb_position not in bases_df.columns or nb_position not in qual_df.columns:\n",
    "            continue  # Skip positions that are not present in either DataFrame\n",
    "\n",
    "        neighbor_positions = range(nb_position - nb_neighbours, nb_position + nb_neighbours + 1)\n",
    "\n",
    "        for read_name in qual_df.index:\n",
    "            total_qual = 0\n",
    "            valid_count = 0\n",
    "\n",
    "            for position in neighbor_positions:\n",
    "                if position not in bases_df.columns:\n",
    "                    continue  # Skip positions that are outside the DataFrame's columns\n",
    "\n",
    "                base = bases_df.at[read_name, position]\n",
    "                quality = qual_df.at[read_name, position]\n",
    "\n",
    "                if base != \"-\" and not pd.isna(quality):\n",
    "                    total_qual += quality\n",
    "                    valid_count += 1\n",
    "\n",
    "            if valid_count == 0:\n",
    "                continue  # Skip if no valid qualities were found\n",
    "            else:\n",
    "                if read_name not in read_mean_qualities:\n",
    "                    read_mean_qualities[read_name] = {}\n",
    "                read_mean_qualities[read_name][nb_position] = total_qual / valid_count\n",
    "\n",
    "    # Convert the dictionary into a DataFrame\n",
    "    mean_quality_df = pd.DataFrame.from_dict(read_mean_qualities, orient='index')\n",
    "    updated_base_df = bases_df[nb_positions] \n",
    "\n",
    "    return updated_base_df, mean_quality_df\n",
    "\n",
    "def add_neighbouring_positions(positions, nb_neighbours, max_index):\n",
    "    if positions is None:\n",
    "        return None\n",
    "    \n",
    "    elif isinstance(positions, int):\n",
    "        positions = [positions]\n",
    "\n",
    "    new_positions = []\n",
    "    for position in positions:\n",
    "        for new_pos in range(position - nb_neighbours, position + nb_neighbours + 1):\n",
    "            if 1 <= new_pos <= max_index:  # Check if the new position is within valid index range\n",
    "                new_positions.append(new_pos)\n",
    "    return sorted(set(new_positions))\n",
    "\n",
    "def calculate_mean_quality_for_reads(bases_df, qual_df, nb_positions, nb_neighbours):\n",
    "    if isinstance(nb_positions, int):\n",
    "        nb_positions = [nb_positions]  # Convert single integer to a list\n",
    "\n",
    "    read_mean_qualities = {}\n",
    "\n",
    "    for nb_position in nb_positions:\n",
    "        if nb_position not in bases_df.columns or nb_position not in qual_df.columns:\n",
    "            continue  # Skip positions that are not present in either DataFrame\n",
    "\n",
    "        neighbor_positions = range(nb_position - nb_neighbours, nb_position + nb_neighbours + 1)\n",
    "\n",
    "        for read_name in qual_df.index:\n",
    "            total_qual = 0\n",
    "            valid_count = 0\n",
    "\n",
    "            for position in neighbor_positions:\n",
    "                if position not in bases_df.columns:\n",
    "                    continue  # Skip positions that are outside the DataFrame's columns\n",
    "\n",
    "                base = bases_df.at[read_name, position]\n",
    "                quality = qual_df.at[read_name, position]\n",
    "\n",
    "                if base != \"-\" and not pd.isna(quality):\n",
    "                    total_qual += quality\n",
    "                    valid_count += 1\n",
    "\n",
    "            if valid_count == 0:\n",
    "                continue  # Skip if no valid qualities were found\n",
    "            else:\n",
    "                if read_name not in read_mean_qualities:\n",
    "                    read_mean_qualities[read_name] = {}\n",
    "                read_mean_qualities[read_name][nb_position] = total_qual / valid_count\n",
    "\n",
    "    # Convert the dictionary into a DataFrame\n",
    "    mean_quality_df = pd.DataFrame.from_dict(read_mean_qualities, orient='index')\n",
    "    updated_base_df = bases_df[nb_positions] \n",
    "\n",
    "    return updated_base_df, mean_quality_df\n",
    "\n",
    "def get_non_error_prop(quality_score):\n",
    "    \"\"\"Convert quality score to non-error probability.\"\"\"\n",
    "    return 1 - 10 ** (-quality_score / 10)\n",
    "\n",
    "def get_softmax_count_df(bases_df, qual_df, nb_positions):\n",
    "\n",
    "    alphabet = \"ACTG-\"\n",
    "    softmax_counts = {position: [] for position in nb_positions}\n",
    "    \n",
    "    for position in nb_positions:\n",
    "        for base in alphabet:\n",
    "            base_mask = bases_df[position] == base\n",
    "            base_counts = base_mask.sum()\n",
    "            # Calculate the non-error probability for each base and sum them up\n",
    "            soft_count = sum(base_mask * qual_df[position].apply(get_non_error_prop))\n",
    "            softmax_counts[position].append(soft_count)\n",
    "\n",
    "    softmax_count_df = pd.DataFrame(softmax_counts, columns=nb_positions, index=list(alphabet))\n",
    "\n",
    "    # Apply softmax to each column (position)\n",
    "    softmax_count_df = softmax_count_df.apply(lambda x: x / x.sum(), axis=0)\n",
    "\n",
    "    return softmax_count_df\n",
    "\n",
    "def get_softmax_count_df_Simulation(bases_df, qual_df, nb_positions):\n",
    "    \n",
    "    alphabet = \"ACTG-\"\n",
    "    softmax_counts = {position: [] for position in nb_positions}\n",
    "\n",
    "    for position in nb_positions:\n",
    "        for base in alphabet:\n",
    "            base_mask = bases_df[position] == base\n",
    "            base_counts = base_mask.sum()\n",
    "            soft_count = sum(base_mask * 0.99)\n",
    "            softmax_counts[position].append(soft_count)\n",
    "    \n",
    "    softmax_count_df = pd.DataFrame(softmax_counts, columns=nb_positions, index=list(alphabet))\n",
    "    softmax_count_df = softmax_count_df.apply(lambda x: x / x.sum(), axis=0)\n",
    "    return softmax_count_df\n",
    "    \n",
    "def get_softmax(soft_count):\n",
    "    \"\"\"Calculate the softmax of a dictionary of soft counts.\"\"\"\n",
    "    # Calculate the sum of the non-error probabilities\n",
    "    total = sum(soft_count.values())\n",
    "    # Calculate the softmax for each base\n",
    "    return {base: count / total for base, count in soft_count.items()}\n",
    "\n",
    "def call_potential_populations(softmax_df, ref_seq):\n",
    "    positions = softmax_df.columns\n",
    "    top_combinations = []\n",
    "    \n",
    "    # Get the top 2 variants for each position\n",
    "    for position in positions:\n",
    "        top_variants = softmax_df[position].nlargest(2)\n",
    "\n",
    "        if top_variants.iloc[1] < 0.1:\n",
    "            top_combinations.append([top_variants.index[0]])\n",
    "        \n",
    "        else:\n",
    "            top_combinations.append(top_variants.index.tolist())\n",
    "\n",
    "        potential_combinations = list(itertools.product(*top_combinations))\n",
    "\n",
    "    \n",
    "    variants = {\"Variant\" : [], \"Probability\" : []}\n",
    "    \n",
    "    for combination in potential_combinations:\n",
    "        final_variant = []\n",
    "        for i, pos in enumerate(positions):\n",
    "\n",
    "            if combination[i] == ref_seq[pos - 1]:\n",
    "                continue\n",
    "\n",
    "            elif combination[i] == \"-\":\n",
    "                var = f\"{ref_seq[pos - 1]}{pos}DEL\"\n",
    "                final_variant.append(var)\n",
    "            else:\n",
    "                var = f\"{ref_seq[pos - 1]}{pos}{combination[i]}\"\n",
    "                final_variant.append(var)\n",
    "\n",
    "        final_variant = '_'.join(final_variant)\n",
    "        if final_variant == \"\":\n",
    "            final_variant = \"#PARENT#\"\n",
    "\n",
    "        joint_prob = np.prod([softmax_df.at[combination[i], positions[i]] for i in range(len(positions))])\n",
    "    \n",
    "        variants[\"Variant\"].append(final_variant)\n",
    "        variants[\"Probability\"].append(joint_prob)\n",
    "\n",
    "    return variants\n",
    "\n",
    "def get_variant_soft(bam_file, template_seq, ref_name, padding = 50):\n",
    "\n",
    "    variants = {\"Variant\" : [], \"Position\" : [], \"Alignment Probability\" : [], \"Alignment Count\" : []}\n",
    "\n",
    "    alignment_count = int(subprocess.run(f\"samtools view -c {bam_file}\", shell=True, capture_output=True).stdout.decode(\"utf-8\").strip())\n",
    "\n",
    "    # if alignment_count < 5:\n",
    "    #     print(\"Not enough alignments\")\n",
    "    #     return None\n",
    "\n",
    "\n",
    "    template = analyser.get_template_sequence(template_seq)\n",
    "\n",
    "    padding_start, padding_end = padding, padding\n",
    "    range_positions = range(padding_start + 1, len(template) - padding_end + 1) \n",
    "\n",
    "    freq_dist = pd.DataFrame(analyser.get_highest_non_ref_base_freq_2(bam_file, ref_name, range_positions, template, qualities=False)[0]).T.rename(columns={0:\"Base\", 1:\"Frequency\"})\n",
    "\n",
    "    nb_positions = analyser.get_nb_positions(freq_dist, 0.3)\n",
    "\n",
    "    available_positions = [pos for pos in range_positions if pos not in nb_positions]\n",
    "\n",
    "\n",
    "    if len(nb_positions) == 0:\n",
    "        # Select random 3 positions\n",
    "        nb_positions = np.random.choice(available_positions, 3, replace=False)\n",
    "\n",
    "    elif len(nb_positions) == 1:\n",
    "        add_pos  = np.random.choice(available_positions, 2, replace=False)\n",
    "        nb_positions = np.append(nb_positions, add_pos)\n",
    "\n",
    "    elif len(nb_positions) > 15:\n",
    "        print(\"Too many positions, either contaminated or sequencing error\")\n",
    "        nb_positions = np.random.choice(range_positions, 3, replace=False)\n",
    "\n",
    "    #bases_df, qual_df = get_bases_from_pileup(bam_file, ref_name, add_neighbouring_positions(nb_positions, 2, len(template)))\n",
    "    bases_df = get_bases_from_pileup_simulation(bam_file, ref_name, add_neighbouring_positions(nb_positions, 2, len(template)))\n",
    "    #bases_df, qual_df = calculate_mean_quality_for_reads(bases_df, qual_df, nb_positions, 2)\n",
    "    #qual_df = qual_df.fillna(10)\n",
    "    #softmax_df = get_softmax_count_df(bases_df, qual_df, nb_positions)\n",
    "    qual_df = pd.DataFrame() # Filler\n",
    "    softmax_df = get_softmax_count_df_Simulation(bases_df, qual_df, nb_positions)\n",
    "    variant_df = pd.DataFrame(call_potential_populations(softmax_df, template)).sort_values(by=\"Probability\", ascending=False)\n",
    "\n",
    "    # Take top variant\n",
    "    variants[\"Variant\"] = variant_df[\"Variant\"].iloc[0]\n",
    "    variants[\"Position\"] = nb_positions\n",
    "    variants[\"Alignment Probability\"] = variant_df[\"Probability\"].iloc[0]\n",
    "    variants[\"Alignment Count\"] = alignment_count\n",
    "\n",
    "    \n",
    "    # print(\"Error in getting variant\")\n",
    "    # variants[\"Variant\"] = np.nan\n",
    "    # variants[\"Position\"] = None\n",
    "    # variants[\"Alignment Probability\"] = None\n",
    "    # variants[\"Alignment Count\"] = alignment_count\n",
    "\n",
    "    return variants\n",
    "\n",
    "\n",
    "def get_variant_df_soft(demultiplex_folder: Path, ref_seq : Path, ref_name : str, barcode_dicts : dict = None, merge = True, min_depth= 5, padding=50, rowwise = False, alignment_name = \"alignment_minimap.bam\"):\n",
    "\n",
    "\n",
    "    if barcode_dicts is None:\n",
    "        barcode_dicts = get_barcode_dict(demultiplex_folder)\n",
    "    \n",
    "    variant_template_df = analyser.template_df(barcode_dicts, rowwise=False)\n",
    "\n",
    "    variants = {\"RBC\": [], \"FBC\": [], \"Position\": [], \"Variant\": [], \"Alignment Probability\": [], \"Alignment Count\": []}\n",
    "\n",
    "    template = analyser.get_template_sequence(ref_seq) # Reference sequence\n",
    "\n",
    "    summary = analyser.read_summary_file(demultiplex_folder)\n",
    "    n_counts = summary.groupby([\"RBC\",\"FBC\"])[\"FBC\"].value_counts().reset_index() \n",
    "\n",
    "\n",
    "\n",
    "    for barcode_id, barcode_dict in barcode_dicts.items():\n",
    "\n",
    "        rbc = os.path.basename(barcode_id)\n",
    "\n",
    "        for front_barcode in barcode_dict:\n",
    "\n",
    "            fbc = os.path.basename(front_barcode)\n",
    "            print(\"Processing\", rbc, fbc)\n",
    "            count = n_counts[(n_counts[\"RBC\"] == rbc) & (n_counts[\"FBC\"] == fbc)][\"count\"].values[0]\n",
    "\n",
    "            # # If alignment file exist continue\n",
    "            if not os.path.exists(os.path.join(front_barcode, \"alignment_minimap.bam\")):\n",
    "                print(f\"Alignment file in {front_barcode} does not exist, running alignment and indexing\")\n",
    "                analyser.run_alignment_and_indexing(ref_seq, front_barcode, site_saturation=True)\n",
    "            \n",
    "            else: \n",
    "                print(\"Alignment file already exists, skipping alignment and indexing\")\n",
    "\n",
    "\n",
    "            bam_file = front_barcode / alignment_name\n",
    "\n",
    "\n",
    "            if not bam_file.exists() or count < min_depth:\n",
    "                print(f\"{bam_file} does not exist.\")\n",
    "                variants[\"RBC\"].append(rbc)\n",
    "                variants[\"FBC\"].append(fbc)\n",
    "                variants[\"Position\"].append(np.nan)\n",
    "                variants[\"Variant\"].append(np.nan)\n",
    "                variants[\"Alignment Count\"].append(np.nan)\n",
    "                variants[\"Alignment Probability\"].append(np.nan)\n",
    "                print(f\"Skipping Variant: {fbc}/{rbc}\")\n",
    "                continue\n",
    "\n",
    "            # try:\n",
    "            if padding == 0:\n",
    "                print(\"Padding is 0. Implementing soft alignment\")\n",
    "                nn_variants = get_variant_soft(bam_file, ref_seq, ref_name, padding = padding)\n",
    "            \n",
    "            else: \n",
    "                nn_variants = get_variant_soft(bam_file, ref_seq, ref_name, padding = padding)\n",
    "                print(nn_variants)\n",
    "\n",
    "            if nn_variants is None:\n",
    "                print(\"Empty variant list\")\n",
    "                variants[\"RBC\"].append(rbc)\n",
    "                variants[\"FBC\"].append(fbc)\n",
    "                variants[\"Position\"].append(np.nan)\n",
    "                variants[\"Variant\"].append(np.nan)\n",
    "                variants[\"Alignment Count\"].append(np.nan)\n",
    "                variants[\"Alignment Probability\"].append(np.nan)\n",
    "                print(f\"Skipping Variant: {fbc}/{rbc}\")\n",
    "                continue\n",
    "\n",
    "            \n",
    "            variants[\"RBC\"].append(rbc)\n",
    "            variants[\"FBC\"].append(fbc)\n",
    "            variants[\"Position\"].append(nn_variants[\"Position\"])\n",
    "            variants[\"Variant\"].append(nn_variants[\"Variant\"])\n",
    "            #variants[\"Reads\"].append(count)\n",
    "            # Check if Alignment count is a number\n",
    "            if isinstance(nn_variants[\"Alignment Count\"], int) & (isinstance(nn_variants[\"Alignment Probability\"], float) or nn_variants[\"Alignment Probability\"] == \"-\"):\n",
    "                variants[\"Alignment Count\"].append(nn_variants[\"Alignment Count\"])\n",
    "                variants[\"Alignment Probability\"].append(nn_variants[\"Alignment Probability\"])\n",
    "\n",
    "\n",
    "            else:\n",
    "                print(f\"Skipping {rbc}/{fbc} due incomplete data\")\n",
    "                variants[\"Alignment Count\"].append(np.nan)\n",
    "                variants[\"Alignment Probability\"].append(np.nan)\n",
    "\n",
    "            print(f\"Variant: {fbc}/{rbc} {nn_variants['Alignment Count']} {nn_variants['Alignment Probability']}\")\n",
    "        \n",
    "        # except Exception as e:\n",
    "        #     # Append 'NA' in case of an exception\n",
    "        #     print(f\"Error processing {rbc}/{fbc}: {e}\")\n",
    "        #     variants[\"RBC\"].append(rbc)\n",
    "        #     variants[\"FBC\"].append(fbc)\n",
    "        #     variants[\"Position\"].append(None)\n",
    "        #     variants[\"Variant\"].append(\"NA\")\n",
    "        #     variants[\"Alignment Count\"].append(\"NA\")\n",
    "        #     variants[\"Alignment Frequency\"].append(\"NA\")\n",
    "\n",
    "\n",
    "\n",
    "    if merge:\n",
    "        variant_template_df = analyser.template_df(barcode_dicts, rowwise=rowwise)\n",
    "        variant_df = analyser.rename_barcode(pd.DataFrame(variants).merge(n_counts, on=[\"RBC\",\"FBC\"] , how=\"left\"))\n",
    "        variant_df[\"Variant\"] = variant_df[\"Variant\"].apply(analyser.format_variant_list)\n",
    "        variant_df[\"Variant\"] = variant_df[\"Variant\"].apply(lambda x: analyser.adjust_variant(x, padding))\n",
    "\n",
    "        return variant_df.merge(variant_template_df, on=[\"Plate\", \"Well\"], how=\"right\")\n",
    "    else:\n",
    "        return variants\n",
    "\n",
    "def call_variant_BF(bam_file, chrom, positions, reference_sequence, qualities = False):\n",
    "    \"\"\" Calls variants from a BAM file.\n",
    "    Args:\n",
    "        - bam_file (str): Path to the BAM file.\n",
    "        - chrom (str): Chromosome or contig name.\n",
    "        - positions (list): List of positions to call variants.\n",
    "        - reference_sequence (str): Reference sequence of the chromosome or contig.\n",
    "    Returns:\n",
    "        - variants (list): List of variants.\n",
    "    \"\"\"\n",
    "\n",
    "    variants = {\"Variant\" : [], \"Position\" : [], \"Frequency\" : []}\n",
    "\n",
    "    if qualities:\n",
    "        bases, qualities = get_highest_non_ref_base_freq(bam_file, chrom, positions, reference_sequence)\n",
    "    \n",
    "    bases = analyser.get_highest_non_ref_base_freq_2(bam_file, chrom, positions, reference_sequence, qualities=False)\n",
    "\n",
    "    for position in positions:\n",
    "        ref_base = reference_sequence[position - 1].upper()\n",
    "        non_ref_base, freq = bases[0][position]\n",
    "        if non_ref_base and freq >= 0.35:\n",
    "\n",
    "            if non_ref_base == \"-\":\n",
    "                non_ref_base = \"DEL\"\n",
    "\n",
    "            variant = f\"{ref_base}{position}{non_ref_base}\"\n",
    "\n",
    "            variants[\"Variant\"].append(variant)\n",
    "            variants[\"Position\"].append(int(position))\n",
    "            variants[\"Frequency\"].append(freq)\n",
    "            #variants[\"Quality-Score\"].append(qualities[position])\n",
    "    \n",
    "    if variants[\"Variant\"] == []:\n",
    "        variants[\"Variant\"].append(\"#PARENT#\")\n",
    "        variants[\"Position\"].append(np.nan)\n",
    "        variants[\"Frequency\"].append(np.nan)\n",
    "        #variants[\"Quality-Score\"].append(\"-\")\n",
    "            \n",
    "\n",
    "    return variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get alignment which were difficult to align \n",
    "variant_df = pd.read_pickle('/home/emre/github_repo/MinION/results/2_hetcpiii_minion_errorprone/local/variants_SW_1Mio.pkl')\n",
    "variant_df_guppy = pd.read_pickle('/home/emre/github_repo/MinION/results/2_hetcpiii_minion_errorprone/local/variants_SW_BF_40k.pkl')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "variant_df_guppy.head(29)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bam_file = Path(\"/home/emre/minION_results/MinION_RBC_0902723_sup/Demultiplex_cpp_70_1Mio_reads/RB01/NB38/alignment_minimap.bam\")\n",
    "template_seq = Path(\"/home/emre/github_repo/MinION/minION/refseq/hetcpiii_padded.fasta\")\n",
    "ref_name = \"HetCPIII\"\n",
    "\n",
    "\n",
    "\n",
    "get_variant_soft(bam_file, template_seq, ref_name, padding = 50)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "nb_positions = [166,104,339]\n",
    "bases_df, qual_df = get_bases_from_pileup(bam_file, ref_name, add_neighbouring_positions(nb_positions, 2, len(template)))\n",
    "bases_df, qual_df = calculate_mean_quality_for_reads(bases_df, qual_df, nb_positions, 2)\n",
    "# Replace NaN with 10\n",
    "qual_df = qual_df.fillna(10)\n",
    "get_softmax_count_df(bases_df, qual_df, nb_positions)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "qual_df\n",
    "variant_df = pd.DataFrame(call_potential_populations(softmax_df, template)).sort_values(by=\"Probability\", ascending=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "demultiplex_folder = Path(\"/home/emre/minION_results/MinION_RBC_0902723_sup/Demultiplex_cpp_70_40k_reads\")\n",
    "template_seq = Path(\"/home/emre/github_repo/MinION/minION/refseq/hetcpiii_padded.fasta\")\n",
    "ref_name = \"HetCPIII\"\n",
    "barcode_dicts = analyser.get_barcode_dict(demultiplex_folder, \"NB\", \"RB\")\n",
    "\n",
    "\n",
    "variant_df_soft = get_variant_df_soft(demultiplex_folder, template_seq, ref_name, barcode_dicts, merge = True, min_depth= 5, padding=50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "variant_df_soft.to_pickle(\"/home/emre/github_repo/MinION/results/2_hetcpiii_minion_errorprone/local/variants_SW_soft_40k.pkl\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "variant_df_soft.head(29)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run for barcode simulater\n",
    "demultiplex_folder = Path(\"/home/emre/minION_results/TamLQV96_sup/Demultiplex_cpp_70\")\n",
    "template_seq = \"/home/emre/tam-lqv.fasta\"\n",
    "ref_name = \"Tam-LQV\"\n",
    "barcode_dicts = analyser.get_barcode_dict(demultiplex_folder, \"NB\", \"RB\")\n",
    "\n",
    "\n",
    "variants = get_variant_df_soft(demultiplex_folder, template_seq, ref_name, barcode_dicts, merge = False, min_depth= 5, padding=0, rowwise=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from tqdm import tqdm\n",
    "for var_path in tqdm(folders):\n",
    "    var_name = os.path.basename(var_path)\n",
    "    depths = glob.glob(f\"{var_path}/depth*\")\n",
    "    for depth in depths:\n",
    "        \n",
    "        bam_file = os.path.join(depth, \"alignment_minimap.bam\")\n",
    "\n",
    "        if not os.path.exists(os.path.join(depth, \"alignment_minimap.bam\")):\n",
    "            print(f\"Alignment file in {depth} does not exist, running alignment and indexing\")\n",
    "            analyser.run_alignment_and_indexing(ref_seq, depth)   \n",
    "                 \n",
    "        else:\n",
    "            print(\"Variant exists\", depth)\n",
    "            nn_variants = get_variant_soft(bam_file, template_seq, ref_name, padding = 50)\n",
    "            print(nn_variants)\n",
    "            break\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "padding_start = 50\n",
    "padding_end = 50\n",
    "template = analyser.get_template_sequence(template_seq)\n",
    "range_positions = range(padding_start + 1, len(template) - padding_end + 1)\n",
    "nb_positions = np.random.choice(range_positions, 3, replace=False)\n",
    "bases_df, qual_df = get_bases_from_pileup(bam_file, ref_name, add_neighbouring_positions(nb_positions, 2, len(template)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def single_plate_annotation(entry):\n",
    "    row = [\"A\", \"B\", \"C\", \"D\", \"E\", \"F\", \"G\", \"H\"]\n",
    "    new_well_name = row[int(entry[\"Plate\"]) - 1] + entry[\"Well\"][1:]\n",
    "    entry[\"Well\"] = new_well_name\n",
    "    return entry"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "variant_df = analyser.rename_barcode(pd.DataFrame(variants), rowwise=True)\n",
    "variant_df = variant_df.apply(single_plate_annotation, axis=1)\n",
    "#variant_df = variant_df.merge(variant_template_df, on=[\"Well\"], how=\"left\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "variant_df = analyser.rename_barcode(pd.DataFrame(variants), rowwise=True)\n",
    "variant_df = variant_df.apply(single_plate_annotation, axis=1)\n",
    "#Drop Plate column\n",
    "variant_df = variant_df.drop(columns=[\"Plate\"])\n",
    "variant_df = variant_df.merge(variant_template_df, on=[\"Well\"], how=\"right\")\n",
    "variant_df.to_pickle(\"../results/4_Tam-LQV/Tam-LQF.pkl\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "demultiplexer_path = Path(\"/home/emre/minION_results/20231130_RL-5sites-8plates_flongle_sup/Demultiplex_cpp_70\")\n",
    "#demultiplexer_path = Path(\"/home/emre/minION_results/20231130_RL-5sites-8plates_flongle_sup/test_alignment\")\n",
    "ref_name = \"ParPgb\"\n",
    "template_seq = \"/home/emre/minION_results/ParPgb.fasta\"\n",
    "barcode_dicts = analyser.get_barcode_dict(demultiplexer_path, \"NB\", \"RB\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing RB12 NB87\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB87/RB12 174 nan\n",
      "Processing RB12 NB03\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB03/RB12 130 nan\n",
      "Processing RB12 NB20\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB20/RB12 69 0.9708044982698963\n",
      "Processing RB12 NB48\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB48/RB12 107 1.0\n",
      "Processing RB12 NB70\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB70/RB12 152 nan\n",
      "Processing RB12 NB04\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Variant: NB04/RB12 13 nan\n",
      "Processing RB12 NB37\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB37/RB12 165 0.9640390683696467\n",
      "Processing RB12 NB88\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB88/RB12 89 0.9121323065269535\n",
      "Processing RB12 NB14\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB14/RB12 163 0.9037865753875717\n",
      "Processing RB12 NB82\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB82/RB12 124 0.9679370447450572\n",
      "Processing RB12 NB44\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB44/RB12 275 0.9496708039068368\n",
      "Processing RB12 NB28\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB28/RB12 103 0.9326043924969366\n",
      "Processing RB12 NB40\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB40/RB12 111 0.8607584940674786\n",
      "Processing RB12 NB81\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB81/RB12 173 0.9321946332717682\n",
      "Processing RB12 NB13\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB13/RB12 110 0.8857580766341098\n",
      "Processing RB12 NB80\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB80/RB12 360 0.8976959019204391\n",
      "Processing RB12 NB47\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB47/RB12 118 0.9495116345877621\n",
      "Processing RB12 NB84\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB84/RB12 191 0.9585797440853885\n",
      "Processing RB12 NB72\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB72/RB12 85 0.9193079584775087\n",
      "Processing RB12 NB10\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB10/RB12 107 0.8217993079584773\n",
      "Processing RB12 NB43\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB43/RB12 239 0.9501409289053063\n",
      "Processing RB12 NB79\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB79/RB12 116 0.9156510107015456\n",
      "Processing RB12 NB36\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB36/RB12 125 0.93708288\n",
      "Processing RB12 NB30\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB30/RB12 89 0.9887640449438203\n",
      "Processing RB12 NB45\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB45/RB12 135 0.6726210435401104\n",
      "Processing RB12 NB93\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB93/RB12 114 0.9212160375306992\n",
      "Processing RB12 NB31\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB31/RB12 125 0.960509952\n",
      "Processing RB12 NB75\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB75/RB12 146 0.9727434790767499\n",
      "Processing RB12 NB21\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB21/RB12 145 0.9524812005412278\n",
      "Processing RB12 NB86\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB86/RB12 199 0.9799752531501729\n",
      "Processing RB12 NB07\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB07/RB12 144 0.774556213017751\n",
      "Processing RB12 NB34\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB34/RB12 162 0.9814814814814815\n",
      "Processing RB12 NB62\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB62/RB12 103 0.9331699500424169\n",
      "Processing RB12 NB23\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB23/RB12 176 0.9886686466942147\n",
      "Processing RB12 NB92\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB92/RB12 40 0.9493754109138725\n",
      "Processing RB12 NB73\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB73/RB12 134 0.9412411101099533\n",
      "Processing RB12 NB67\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB67/RB12 143 0.958286468775979\n",
      "Processing RB12 NB29\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB29/RB12 134 0.9777785997612735\n",
      "Processing RB12 NB33\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB33/RB12 126 0.9761904761904762\n",
      "Processing RB12 NB06\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB06/RB12 170 0.9364657032363117\n",
      "Processing RB12 NB78\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB78/RB12 104 0.9619989616522531\n",
      "Processing RB12 NB95\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB95/RB12 148 0.7782362347738531\n",
      "Processing RB12 NB26\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB26/RB12 133 0.8752731509365871\n",
      "Processing RB12 NB76\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB76/RB12 129 0.9689922480620154\n",
      "Processing RB12 NB68\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB68/RB12 148 0.9338384449094821\n",
      "Processing RB12 NB74\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB74/RB12 140 0.9786734693877549\n",
      "Processing RB12 NB89\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB89/RB12 45 0.934320987654321\n",
      "Processing RB12 NB83\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB83/RB12 358 0.972231370716024\n",
      "Processing RB12 NB22\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB22/RB12 103 0.942592248567117\n",
      "Processing RB12 NB96\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB96/RB12 166 0.9002565640701868\n",
      "Processing RB12 NB32\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB32/RB12 143 0.9584807094709776\n",
      "Processing RB12 NB41\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB41/RB12 253 0.9116761433684261\n",
      "Processing RB12 NB17\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB17/RB12 123 0.7576733115391482\n",
      "Processing RB12 NB66\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB66/RB12 116 0.797907099512075\n",
      "Processing RB12 NB85\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB85/RB12 32 0.9375\n",
      "Processing RB12 NB61\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB61/RB12 257 0.9806505522695007\n",
      "Processing RB12 NB46\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB46/RB12 229 0.8011919605482509\n",
      "Processing RB12 NB77\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB77/RB12 271 0.9490987845704494\n",
      "Processing RB12 NB12\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB12/RB12 75 0.8715875555555559\n",
      "Processing RB12 NB38\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB38/RB12 221 0.9057741247548028\n",
      "Processing RB12 NB64\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB64/RB12 144 0.9520613640260627\n",
      "Processing RB12 NB71\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB71/RB12 34 1.0\n",
      "Processing RB12 NB24\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB24/RB12 84 0.9301202623906707\n",
      "Processing RB12 NB94\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB94/RB12 144 0.9623507805325985\n",
      "Processing RB12 NB65\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB65/RB12 117 0.899543148063858\n",
      "Processing RB12 NB15\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB15/RB12 111 0.9732157285115649\n",
      "Processing RB12 NB02\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB02/RB12 116 0.9403983353151009\n",
      "Processing RB12 NB01\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB01/RB12 29 0.6122448979591837\n",
      "Processing RB12 NB27\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB27/RB12 194 0.9289734658525014\n",
      "Processing RB12 NB08\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB08/RB12 90 0.9554349198333545\n",
      "Processing RB12 NB91\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB91/RB12 99 0.9699010305070912\n",
      "Processing RB12 NB19\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB19/RB12 5 0.7999999999999999\n",
      "Processing RB12 NB35\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB35/RB12 212 0.9295567817728728\n",
      "Processing RB12 NB09\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB09/RB12 111 0.9642074506939371\n",
      "Processing RB12 NB11\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB11/RB12 164 0.8692139369713149\n",
      "Processing RB12 NB63\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB63/RB12 146 0.9132865658827246\n",
      "Processing RB12 NB42\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB42/RB12 135 0.9560493827160493\n",
      "Processing RB12 NB18\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB18/RB12 250 0.9097920000000003\n",
      "Processing RB12 NB90\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB90/RB12 72 0.9722222222222223\n",
      "Processing RB12 NB05\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB05/RB12 180 0.8496468820435762\n",
      "Processing RB12 NB39\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "/home/emre/minION_results/20231130_RL-5sites-8plates_flongle_sup/Demultiplex_cpp_70/RB12/NB39/alignment_minimap.bam does not exist.\n",
      "Skipping Variant: NB39/RB12\n",
      "Processing RB12 NB69\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB69/RB12 255 0.9574142072053736\n",
      "Processing RB12 NB25\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB25/RB12 154 0.9674911358879063\n",
      "Processing RB09 NB87\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB87/RB09 103 0.44752018454440606\n",
      "Processing RB09 NB03\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB03/RB09 8 0.6562500000000001\n",
      "Processing RB09 NB20\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "/home/emre/minION_results/20231130_RL-5sites-8plates_flongle_sup/Demultiplex_cpp_70/RB09/NB20/alignment_minimap.bam does not exist.\n",
      "Skipping Variant: NB20/RB09\n",
      "Processing RB09 NB70\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB70/RB09 125 0.9218457599999997\n",
      "Processing RB09 NB04\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB04/RB09 68 0.7822104620394874\n",
      "Processing RB09 NB37\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB37/RB09 24 0.9583333333333334\n",
      "Processing RB09 NB88\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB88/RB09 52 0.9619082840236686\n",
      "Processing RB09 NB14\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB14/RB09 55 0.9113869271224644\n",
      "Processing RB09 NB82\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB82/RB09 117 0.9495895700147289\n",
      "Processing RB09 NB54\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB54/RB09 120 0.8720833333333332\n",
      "Processing RB09 NB44\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB44/RB09 121 0.8387847222222219\n",
      "Processing RB09 NB57\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB57/RB09 117 0.9168475780353931\n",
      "Processing RB09 NB28\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB28/RB09 108 0.8918895747599449\n",
      "Processing RB09 NB40\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB40/RB09 157 0.9069615103053236\n",
      "Processing RB09 NB81\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB81/RB09 52 1.0\n",
      "Processing RB09 NB13\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB13/RB09 67 0.9701492537313432\n",
      "Processing RB09 NB53\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB53/RB09 102 0.8770335315979524\n",
      "Processing RB09 NB80\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB80/RB09 167 0.6332464691614293\n",
      "Processing RB09 NB47\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB47/RB09 94 0.9071689317395956\n",
      "Processing RB09 NB84\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB84/RB09 154 0.9178416456203604\n",
      "Processing RB09 NB72\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB72/RB09 9 0.7500000000000001\n",
      "Processing RB09 NB10\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB10/RB09 10 1.0\n",
      "Processing RB09 NB43\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB43/RB09 146 0.7867553405223662\n",
      "Processing RB09 NB58\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "/home/emre/minION_results/20231130_RL-5sites-8plates_flongle_sup/Demultiplex_cpp_70/RB09/NB58/alignment_minimap.bam does not exist.\n",
      "Skipping Variant: NB58/RB09\n",
      "Processing RB09 NB79\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB79/RB09 64 0.968994140625\n",
      "Processing RB09 NB36\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB36/RB09 19 0.8502697186178745\n",
      "Processing RB09 NB30\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB30/RB09 109 0.8933359793487252\n",
      "Processing RB09 NB60\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB60/RB09 101 0.9414763258504067\n",
      "Processing RB09 NB45\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB45/RB09 117 0.9165603675794339\n",
      "Processing RB09 NB93\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB93/RB09 165 0.9174455296769346\n",
      "Processing RB09 NB31\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB31/RB09 100 0.9032639999999998\n",
      "Processing RB09 NB75\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB75/RB09 129 0.8732145177992712\n",
      "Processing RB09 NB21\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB21/RB09 95 0.9686426592797784\n",
      "Processing RB09 NB86\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB86/RB09 123 0.9283414666389377\n",
      "Processing RB09 NB07\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB07/RB09 5 0.7999999999999999\n",
      "Processing RB09 NB34\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB34/RB09 74 0.882541507906738\n",
      "Processing RB09 NB62\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB62/RB09 135 0.985185185185185\n",
      "Processing RB09 NB23\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB23/RB09 58 0.8837231948829389\n",
      "Processing RB09 NB92\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB92/RB09 51 0.9046294411651626\n",
      "Processing RB09 NB73\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "/home/emre/minION_results/20231130_RL-5sites-8plates_flongle_sup/Demultiplex_cpp_70/RB09/NB73/alignment_minimap.bam does not exist.\n",
      "Skipping Variant: NB73/RB09\n",
      "Processing RB09 NB67\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB67/RB09 113 0.9472942282089435\n",
      "Processing RB09 NB29\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB29/RB09 110 0.8947483095416979\n",
      "Processing RB09 NB33\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB33/RB09 11 0.9090909090909091\n",
      "Processing RB09 NB06\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB06/RB09 19 0.9473684210526316\n",
      "Processing RB09 NB78\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB78/RB09 99 0.8827382075009867\n",
      "Processing RB09 NB95\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB95/RB09 111 0.9732157285115649\n",
      "Processing RB09 NB26\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB26/RB09 38 0.8738883219128153\n",
      "Processing RB09 NB76\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB76/RB09 90 0.9453045267489714\n",
      "Processing RB09 NB59\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB59/RB09 97 0.90016906383776\n",
      "Processing RB09 NB68\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB68/RB09 135 0.9852400548696842\n",
      "Processing RB09 NB55\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB55/RB09 113 0.9823008849557522\n",
      "Processing RB09 NB74\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB74/RB09 88 0.74511328125\n",
      "Processing RB09 NB89\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB89/RB09 73 1.0\n",
      "Processing RB09 NB83\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB83/RB09 152 0.9224376731301936\n",
      "Processing RB09 NB22\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB22/RB09 6 0.8333333333333333\n",
      "Processing RB09 NB96\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB96/RB09 31 0.935483870967742\n",
      "Processing RB09 NB32\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "/home/emre/minION_results/20231130_RL-5sites-8plates_flongle_sup/Demultiplex_cpp_70/RB09/NB32/alignment_minimap.bam does not exist.\n",
      "Skipping Variant: NB32/RB09\n",
      "Processing RB09 NB41\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB41/RB09 174 0.9714625445897739\n",
      "Processing RB09 NB17\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB17/RB09 106 0.8526837590762841\n",
      "Processing RB09 NB66\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB66/RB09 108 0.954301459381192\n",
      "Processing RB09 NB61\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB61/RB09 146 0.9593263276412085\n",
      "Processing RB09 NB56\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB56/RB09 135 0.9276146928821822\n",
      "Processing RB09 NB46\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB46/RB09 131 0.8465422440639633\n",
      "Processing RB09 NB49\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB49/RB09 38 0.7084487534626036\n",
      "Processing RB09 NB77\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB77/RB09 182 0.9566450938265937\n",
      "Processing RB09 NB12\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB12/RB09 65 0.9098224852071007\n",
      "Processing RB09 NB38\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB38/RB09 74 0.8208793161313251\n",
      "Processing RB09 NB64\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB64/RB09 105 0.785373501781665\n",
      "Processing RB09 NB71\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB71/RB09 25 0.9216000000000002\n",
      "Processing RB09 NB24\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB24/RB09 77 0.923920067114535\n",
      "Processing RB09 NB94\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB94/RB09 39 0.9493754109138725\n",
      "Processing RB09 NB50\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB50/RB09 42 0.845754446923183\n",
      "Processing RB09 NB65\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB65/RB09 83 0.9522427057628103\n",
      "Processing RB09 NB15\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB15/RB09 12 0.9166666666666666\n",
      "Processing RB09 NB02\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB02/RB09 10 0.7200000000000001\n",
      "Processing RB09 NB27\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB27/RB09 157 0.9498123074720668\n",
      "Processing RB09 NB08\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "/home/emre/minION_results/20231130_RL-5sites-8plates_flongle_sup/Demultiplex_cpp_70/RB09/NB08/alignment_minimap.bam does not exist.\n",
      "Skipping Variant: NB08/RB09\n",
      "Processing RB09 NB91\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB91/RB09 100 0.902286\n",
      "Processing RB09 NB19\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB19/RB09 51 0.9419454056132256\n",
      "Processing RB09 NB51\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "/home/emre/minION_results/20231130_RL-5sites-8plates_flongle_sup/Demultiplex_cpp_70/RB09/NB51/alignment_minimap.bam does not exist.\n",
      "Skipping Variant: NB51/RB09\n",
      "Processing RB09 NB52\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB52/RB09 124 0.9129728777147459\n",
      "Processing RB09 NB35\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB35/RB09 198 0.8771832187810151\n",
      "Processing RB09 NB09\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB09/RB09 89 0.9443252114631991\n",
      "Processing RB09 NB63\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB63/RB09 70 0.8714285714285714\n",
      "Processing RB09 NB42\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB42/RB09 105 0.9809523809523809\n",
      "Processing RB09 NB18\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB18/RB09 16 0.9375\n",
      "Processing RB09 NB90\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB90/RB09 8 1.0\n",
      "Processing RB09 NB05\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB05/RB09 106 0.9262948608582922\n",
      "Processing RB09 NB39\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB39/RB09 108 0.9452128740537521\n",
      "Processing RB09 NB69\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB69/RB09 165 0.831741325096697\n",
      "Processing RB11 NB87\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB87/RB11 80 0.9629667968750001\n",
      "Processing RB11 NB03\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB03/RB11 107 0.9448811184586988\n",
      "Processing RB11 NB20\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB20/RB11 77 0.8864638481774595\n",
      "Processing RB11 NB48\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB48/RB11 94 0.9893617021276596\n",
      "Processing RB11 NB70\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB70/RB11 161 0.9320628062188957\n",
      "Processing RB11 NB04\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB04/RB11 93 0.9055382125101169\n",
      "Processing RB11 NB37\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB37/RB11 129 0.9311306854416265\n",
      "Processing RB11 NB88\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB88/RB11 71 0.9307677048204721\n",
      "Processing RB11 NB14\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB14/RB11 155 0.9366028666375751\n",
      "Processing RB11 NB82\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB82/RB11 132 0.9625786097893535\n",
      "Processing RB11 NB54\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB54/RB11 123 0.8700759652282871\n",
      "Processing RB11 NB44\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB44/RB11 142 0.8401533481787368\n",
      "Processing RB11 NB57\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB57/RB11 239 0.9708163372489977\n",
      "Processing RB11 NB28\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB28/RB11 88 0.9010494928625096\n",
      "Processing RB11 NB40\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB40/RB11 146 0.8129085116067362\n",
      "Processing RB11 NB81\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB81/RB11 198 0.9650530403514793\n",
      "Processing RB11 NB13\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB13/RB11 139 0.8328881400509571\n",
      "Processing RB11 NB53\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB53/RB11 194 0.9343105361942339\n",
      "Processing RB11 NB80\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB80/RB11 197 0.94993193922651\n",
      "Processing RB11 NB47\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB47/RB11 114 0.8877272629095052\n",
      "Processing RB11 NB84\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB84/RB11 153 0.9486044834440248\n",
      "Processing RB11 NB72\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB72/RB11 80 0.8340820312500001\n",
      "Processing RB11 NB10\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB10/RB11 110 0.9198347107438015\n",
      "Processing RB11 NB43\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB43/RB11 164 0.975609756097561\n",
      "Processing RB11 NB58\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB58/RB11 17 0.9411764705882353\n",
      "Processing RB11 NB79\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB79/RB11 115 0.9482041587901701\n",
      "Processing RB11 NB36\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB36/RB11 101 0.7968461869694842\n",
      "Processing RB11 NB30\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB30/RB11 54 0.9451303155006858\n",
      "Processing RB11 NB60\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB60/RB11 134 0.948318110937848\n",
      "Processing RB11 NB45\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB45/RB11 135 0.9418316313570084\n",
      "Processing RB11 NB93\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB93/RB11 172 0.9540610417950621\n",
      "Processing RB11 NB31\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB31/RB11 137 0.949752912182808\n",
      "Processing RB11 NB75\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB75/RB11 127 0.9307341449328566\n",
      "Processing RB11 NB21\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB21/RB11 120 0.9670127314814816\n",
      "Processing RB11 NB86\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB86/RB11 145 0.8458265611546187\n",
      "Processing RB11 NB07\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB07/RB11 95 0.9373961218836565\n",
      "Processing RB11 NB34\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB34/RB11 120 0.9267916666666666\n",
      "Processing RB11 NB62\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB62/RB11 119 0.9420238683708777\n",
      "Processing RB11 NB23\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB23/RB11 136 0.7644824311490975\n",
      "Processing RB11 NB92\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB92/RB11 78 0.9871794871794872\n",
      "Processing RB11 NB73\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB73/RB11 108 0.9633058984910836\n",
      "Processing RB11 NB67\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB67/RB11 99 0.9796959600166597\n",
      "Processing RB11 NB29\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB29/RB11 114 0.9311259064867463\n",
      "Processing RB11 NB33\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB33/RB11 109 0.9727287960346835\n",
      "Processing RB11 NB06\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB06/RB11 153 0.948060941828255\n",
      "Processing RB11 NB78\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB78/RB11 101 0.9320577190549169\n",
      "Processing RB11 NB95\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB95/RB11 41 0.975609756097561\n",
      "Processing RB11 NB26\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB26/RB11 69 0.9712245326612057\n",
      "Processing RB11 NB16\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB16/RB11 18 0.9444444444444445\n",
      "Processing RB11 NB76\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB76/RB11 99 0.98989898989899\n",
      "Processing RB11 NB59\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB59/RB11 242 0.8552910399359664\n",
      "Processing RB11 NB68\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB68/RB11 152 0.9610007153010643\n",
      "Processing RB11 NB55\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB55/RB11 215 0.9361949010778925\n",
      "Processing RB11 NB74\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB74/RB11 108 0.9723936899862827\n",
      "Processing RB11 NB89\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB89/RB11 58 0.9322235434007136\n",
      "Processing RB11 NB83\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB83/RB11 162 0.8982370573591428\n",
      "Processing RB11 NB22\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB22/RB11 122 0.9432948132222522\n",
      "Processing RB11 NB96\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB96/RB11 100 0.9107999999999999\n",
      "Processing RB11 NB32\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB32/RB11 101 0.9508871679247132\n",
      "Processing RB11 NB41\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB41/RB11 157 0.9556574303217168\n",
      "Processing RB11 NB17\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB17/RB11 128 0.8151359558105465\n",
      "Processing RB11 NB66\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB66/RB11 131 0.9277665311193447\n",
      "Processing RB11 NB85\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB85/RB11 22 0.594838897798513\n",
      "Processing RB11 NB61\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB61/RB11 206 0.947492832152953\n",
      "Processing RB11 NB56\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB56/RB11 326 0.884796087141549\n",
      "Processing RB11 NB46\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB46/RB11 163 0.8811412904055581\n",
      "Processing RB11 NB49\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB49/RB11 316 0.949997997115847\n",
      "Processing RB11 NB77\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB77/RB11 148 0.9729729729729729\n",
      "Processing RB11 NB12\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB12/RB11 68 0.9708044982698963\n",
      "Processing RB11 NB38\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB38/RB11 134 0.962909333927378\n",
      "Processing RB11 NB64\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB64/RB11 97 0.9896907216494846\n",
      "Processing RB11 NB71\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB71/RB11 135 0.9778875171467764\n",
      "Processing RB11 NB24\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB24/RB11 119 0.9915966386554622\n",
      "Processing RB11 NB94\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB94/RB11 92 0.9465922166515987\n",
      "Processing RB11 NB50\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB50/RB11 220 0.8902900075131481\n",
      "Processing RB11 NB65\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB65/RB11 89 0.8917555240017647\n",
      "Processing RB11 NB15\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB15/RB11 88 0.5361774882963568\n",
      "Processing RB11 NB02\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB02/RB11 99 0.9308141098774708\n",
      "Processing RB11 NB01\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB01/RB11 89 0.9554349198333545\n",
      "Processing RB11 NB27\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB27/RB11 159 0.9624619279300659\n",
      "Processing RB11 NB08\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB08/RB11 92 0.9569943289224953\n",
      "Processing RB11 NB91\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB91/RB11 100 0.9313919999999999\n",
      "Processing RB11 NB19\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB19/RB11 99 0.9303132333435363\n",
      "Processing RB11 NB51\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB51/RB11 220 0.924370773854245\n",
      "Processing RB11 NB52\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB52/RB11 163 0.849333333333333\n",
      "Processing RB11 NB35\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB35/RB11 173 0.9713647926296476\n",
      "Processing RB11 NB09\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB09/RB11 130 0.9923076923076923\n",
      "Processing RB11 NB11\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB11/RB11 146 0.9528015485184449\n",
      "Processing RB11 NB63\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB63/RB11 87 0.9106640364584522\n",
      "Processing RB11 NB42\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB42/RB11 98 0.9498083281625854\n",
      "Processing RB11 NB18\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB18/RB11 161 0.9272646629833934\n",
      "Processing RB11 NB90\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB90/RB11 98 0.9693877551020408\n",
      "Processing RB11 NB05\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB05/RB11 163 0.7522379001668943\n",
      "Processing RB11 NB39\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB39/RB11 120 0.9749311489301603\n",
      "Processing RB11 NB69\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB69/RB11 179 0.8967675668343232\n",
      "Processing RB11 NB25\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB25/RB11 155 0.7735780838607408\n",
      "Processing RB10 NB87\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB87/RB10 91 0.9346697258785172\n",
      "Processing RB10 NB03\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB03/RB10 141 0.937321745235213\n",
      "Processing RB10 NB20\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB20/RB10 104 0.9615420869073429\n",
      "Processing RB10 NB48\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB48/RB10 83 0.8600667731165627\n",
      "Processing RB10 NB70\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB70/RB10 122 0.9354340231120667\n",
      "Processing RB10 NB04\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB04/RB10 5 1.0\n",
      "Processing RB10 NB37\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB37/RB10 137 0.9567881189397174\n",
      "Processing RB10 NB88\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB88/RB10 97 0.8708157247995723\n",
      "Processing RB10 NB14\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB14/RB10 161 0.9751552795031055\n",
      "Processing RB10 NB82\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB82/RB10 86 0.9541612688190978\n",
      "Processing RB10 NB54\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB54/RB10 115 0.9482041587901702\n",
      "Processing RB10 NB44\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB44/RB10 150 0.7880106666666662\n",
      "Processing RB10 NB57\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB57/RB10 102 0.9327106467346645\n",
      "Processing RB10 NB28\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB28/RB10 96 0.8596718437264617\n",
      "Processing RB10 NB40\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB40/RB10 172 0.9711662573735644\n",
      "Processing RB10 NB81\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB81/RB10 110 0.9640480841472576\n",
      "Processing RB10 NB13\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB13/RB10 124 0.9679370447450572\n",
      "Processing RB10 NB53\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB53/RB10 141 0.964939974408011\n",
      "Processing RB10 NB80\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB80/RB10 176 0.9007710514885422\n",
      "Processing RB10 NB47\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB47/RB10 102 0.8866273152859759\n",
      "Processing RB10 NB84\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB84/RB10 117 0.9745050770691797\n",
      "Processing RB10 NB72\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB72/RB10 60 0.9183333333333336\n",
      "Processing RB10 NB10\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB10/RB10 102 0.885123368840039\n",
      "Processing RB10 NB43\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB43/RB10 184 0.8992202268431004\n",
      "Processing RB10 NB58\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "/home/emre/minION_results/20231130_RL-5sites-8plates_flongle_sup/Demultiplex_cpp_70/RB10/NB58/alignment_minimap.bam does not exist.\n",
      "Skipping Variant: NB58/RB10\n",
      "Processing RB10 NB79\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB79/RB10 89 0.8910235769232406\n",
      "Processing RB10 NB36\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB36/RB10 131 0.9847910960899712\n",
      "Processing RB10 NB30\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB30/RB10 144 0.9654706790123455\n",
      "Processing RB10 NB60\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB60/RB10 111 0.9023618212807402\n",
      "Processing RB10 NB45\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB45/RB10 122 0.8882692824509539\n",
      "Processing RB10 NB93\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB93/RB10 101 0.9128400341259496\n",
      "Processing RB10 NB31\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB31/RB10 125 0.9292799999999999\n",
      "Processing RB10 NB75\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB75/RB10 127 0.9231672006651098\n",
      "Processing RB10 NB21\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB21/RB10 112 0.9645248724489796\n",
      "Processing RB10 NB86\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB86/RB10 207 0.9804878048780488\n",
      "Processing RB10 NB07\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB07/RB10 117 0.9409014537219665\n",
      "Processing RB10 NB34\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB34/RB10 141 0.8190917968749998\n",
      "Processing RB10 NB62\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB62/RB10 149 0.9666231250844555\n",
      "Processing RB10 NB23\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB23/RB10 145 0.952061364026063\n",
      "Processing RB10 NB92\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB92/RB10 64 0.9538536071777344\n",
      "Processing RB10 NB73\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB73/RB10 92 0.8848730171776116\n",
      "Processing RB10 NB67\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB67/RB10 92 0.9238014732520228\n",
      "Processing RB10 NB29\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB29/RB10 132 0.9772727272727271\n",
      "Processing RB10 NB33\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB33/RB10 116 0.9408365451638033\n",
      "Processing RB10 NB06\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB06/RB10 135 0.9348038408779149\n",
      "Processing RB10 NB78\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB78/RB10 106 0.6808472766108935\n",
      "Processing RB10 NB95\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB95/RB10 108 0.9453827541533303\n",
      "Processing RB10 NB26\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB26/RB10 155 0.9806451612903225\n",
      "Processing RB10 NB16\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "/home/emre/minION_results/20231130_RL-5sites-8plates_flongle_sup/Demultiplex_cpp_70/RB10/NB16/alignment_minimap.bam does not exist.\n",
      "Skipping Variant: NB16/RB10\n",
      "Processing RB10 NB76\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB76/RB10 85 0.9417301038062282\n",
      "Processing RB10 NB59\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB59/RB10 102 0.8856944161747744\n",
      "Processing RB10 NB68\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB68/RB10 149 0.8767638195147504\n",
      "Processing RB10 NB55\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB55/RB10 107 0.9628788540483885\n",
      "Processing RB10 NB74\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB74/RB10 118 0.8613538872036575\n",
      "Processing RB10 NB89\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB89/RB10 101 0.9320577190549171\n",
      "Processing RB10 NB83\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB83/RB10 145 0.9658026159334124\n",
      "Processing RB10 NB22\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB22/RB10 150 0.967021037037037\n",
      "Processing RB10 NB96\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB96/RB10 110 0.8695749258506468\n",
      "Processing RB10 NB32\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB32/RB10 121 0.9752066115702479\n",
      "Processing RB10 NB41\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB41/RB10 223 0.9910514991252589\n",
      "Processing RB10 NB17\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB17/RB10 147 0.9569418189455996\n",
      "Processing RB10 NB66\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB66/RB10 145 0.9794054696789534\n",
      "Processing RB10 NB85\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "/home/emre/minION_results/20231130_RL-5sites-8plates_flongle_sup/Demultiplex_cpp_70/RB10/NB85/alignment_minimap.bam does not exist.\n",
      "Skipping Variant: NB85/RB10\n",
      "Processing RB10 NB61\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB61/RB10 194 0.7541806802852321\n",
      "Processing RB10 NB56\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB56/RB10 185 0.9520500661362605\n",
      "Processing RB10 NB46\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB46/RB10 174 0.9372818174496597\n",
      "Processing RB10 NB49\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB49/RB10 195 0.9393849166130697\n",
      "Processing RB10 NB77\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB77/RB10 154 0.9485984583808836\n",
      "Processing RB10 NB12\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB12/RB10 77 0.9741946365322991\n",
      "Processing RB10 NB38\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB38/RB10 198 0.9453580803443062\n",
      "Processing RB10 NB64\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB64/RB10 108 0.935185185185185\n",
      "Processing RB10 NB71\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB71/RB10 106 0.9080482545994344\n",
      "Processing RB10 NB24\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB24/RB10 108 0.9542181069958847\n",
      "Processing RB10 NB94\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB94/RB10 129 0.9463951229078826\n",
      "Processing RB10 NB50\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB50/RB10 119 0.9915966386554622\n",
      "Processing RB10 NB65\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB65/RB10 68 0.8035060044779161\n",
      "Processing RB10 NB15\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB15/RB10 135 0.9706439059086518\n",
      "Processing RB10 NB02\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB02/RB10 107 0.9813084112149534\n",
      "Processing RB10 NB01\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB01/RB10 45 0.8104687500000001\n",
      "Processing RB10 NB27\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB27/RB10 118 0.390545280612245\n",
      "Processing RB10 NB08\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB08/RB10 66 0.9403122130394859\n",
      "Processing RB10 NB91\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB91/RB10 119 0.9422303770742109\n",
      "Processing RB10 NB19\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB19/RB10 74 0.9466764061358655\n",
      "Processing RB10 NB51\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB51/RB10 124 0.9523891443724615\n",
      "Processing RB10 NB52\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB52/RB10 120 0.9752077546296297\n",
      "Processing RB10 NB35\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB35/RB10 157 0.9007132755350082\n",
      "Processing RB10 NB09\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB09/RB10 118 0.9580580293019247\n",
      "Processing RB10 NB11\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB11/RB10 139 0.9085451063609543\n",
      "Processing RB10 NB63\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB63/RB10 86 0.931710415435119\n",
      "Processing RB10 NB42\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB42/RB10 134 0.9628565732835107\n",
      "Processing RB10 NB18\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB18/RB10 124 0.8983837400557214\n",
      "Processing RB10 NB90\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB90/RB10 102 0.9047293273326249\n",
      "Processing RB10 NB05\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB05/RB10 137 0.9563689446366781\n",
      "Processing RB10 NB39\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB39/RB10 148 0.9203798392987581\n",
      "Processing RB10 NB69\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB69/RB10 171 0.9258757280442765\n",
      "Processing RB10 NB25\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB25/RB10 143 0.9047731573038432\n",
      "Processing RB06 NB87\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB87/RB06 118 0.8164455459979587\n",
      "Processing RB06 NB03\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB03/RB06 105 0.9248979591836735\n",
      "Processing RB06 NB20\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB20/RB06 48 0.9791666666666666\n",
      "Processing RB06 NB70\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB70/RB06 101 0.9900990099009901\n",
      "Processing RB06 NB04\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB04/RB06 51 0.8477508650519032\n",
      "Processing RB06 NB37\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB37/RB06 80 0.950625\n",
      "Processing RB06 NB88\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB88/RB06 46 1.0\n",
      "Processing RB06 NB14\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB14/RB06 112 0.9646828876639942\n",
      "Processing RB06 NB82\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB82/RB06 70 0.9575510204081633\n",
      "Processing RB06 NB54\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB54/RB06 88 1.0\n",
      "Processing RB06 NB44\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB44/RB06 211 0.9763931627771164\n",
      "Processing RB06 NB57\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB57/RB06 63 0.9375157470395568\n",
      "Processing RB06 NB28\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB28/RB06 61 0.8720773985487772\n",
      "Processing RB06 NB40\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB40/RB06 136 0.9346317327752899\n",
      "Processing RB06 NB81\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB81/RB06 70 0.9857142857142858\n",
      "Processing RB06 NB13\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB13/RB06 73 0.9461386006267081\n",
      "Processing RB06 NB53\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB53/RB06 91 0.9565269894940225\n",
      "Processing RB06 NB80\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB80/RB06 176 0.9169503662659654\n",
      "Processing RB06 NB47\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB47/RB06 66 0.8718888888888888\n",
      "Processing RB06 NB43\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB43/RB06 194 0.9392279600689403\n",
      "Processing RB06 NB58\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB58/RB06 23 0.7594312484589463\n",
      "Processing RB06 NB79\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB79/RB06 58 0.9310344827586208\n",
      "Processing RB06 NB36\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB36/RB06 79 0.9254927095016825\n",
      "Processing RB06 NB30\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB30/RB06 117 0.9321352911096501\n",
      "Processing RB06 NB60\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB60/RB06 84 0.9175170068027212\n",
      "Processing RB06 NB45\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB45/RB06 109 0.899924248800606\n",
      "Processing RB06 NB93\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB93/RB06 160 0.8471669921874998\n",
      "Processing RB06 NB31\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB31/RB06 73 0.9459560893225746\n",
      "Processing RB06 NB75\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB75/RB06 75 0.9472\n",
      "Processing RB06 NB21\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB21/RB06 56 0.9646045918367346\n",
      "Processing RB06 NB86\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB86/RB06 130 0.9846153846153847\n",
      "Processing RB06 NB07\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB07/RB06 76 0.9608725761772855\n",
      "Processing RB06 NB34\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB34/RB06 67 0.5145912229895299\n",
      "Processing RB06 NB62\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB62/RB06 82 0.9167017309673395\n",
      "Processing RB06 NB23\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB23/RB06 118 0.8498178976428943\n",
      "Processing RB06 NB92\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB92/RB06 82 0.975609756097561\n",
      "Processing RB06 NB73\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB73/RB06 71 0.9164848244395953\n",
      "Processing RB06 NB67\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB67/RB06 77 0.9613762860516109\n",
      "Processing RB06 NB29\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB29/RB06 60 0.9505555555555556\n",
      "Processing RB06 NB33\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB33/RB06 64 0.968994140625\n",
      "Processing RB06 NB06\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB06/RB06 81 0.9513755995491505\n",
      "Processing RB06 NB78\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB78/RB06 58 0.856271909769852\n",
      "Processing RB06 NB95\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB95/RB06 79 0.8499408284023671\n",
      "Processing RB06 NB26\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB26/RB06 71 0.9718309859154929\n",
      "Processing RB06 NB16\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB16/RB06 20 1.0\n",
      "Processing RB06 NB76\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB76/RB06 61 0.9836065573770492\n",
      "Processing RB06 NB59\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB59/RB06 62 0.9838709677419355\n",
      "Processing RB06 NB68\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB68/RB06 103 0.9616345162149373\n",
      "Processing RB06 NB55\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB55/RB06 77 0.9870129870129871\n",
      "Processing RB06 NB74\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB74/RB06 64 0.8496093750000001\n",
      "Processing RB06 NB89\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB89/RB06 83 0.9760487734068807\n",
      "Processing RB06 NB83\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB83/RB06 146 0.9197181614171102\n",
      "Processing RB06 NB22\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB22/RB06 61 0.9196452566514378\n",
      "Processing RB06 NB96\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB96/RB06 124 0.8675858480749217\n",
      "Processing RB06 NB32\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB32/RB06 70 0.5578775510204083\n",
      "Processing RB06 NB41\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB41/RB06 167 0.9124212044196592\n",
      "Processing RB06 NB17\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB17/RB06 70 0.9714285714285715\n",
      "Processing RB06 NB66\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB66/RB06 61 0.9196452566514378\n",
      "Processing RB06 NB85\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "/home/emre/minION_results/20231130_RL-5sites-8plates_flongle_sup/Demultiplex_cpp_70/RB06/NB85/alignment_minimap.bam does not exist.\n",
      "Skipping Variant: NB85/RB06\n",
      "Processing RB06 NB61\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB61/RB06 100 0.9030699999999999\n",
      "Processing RB06 NB56\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB56/RB06 195 0.9540565417488495\n",
      "Processing RB06 NB46\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB46/RB06 127 0.9303680024682884\n",
      "Processing RB06 NB49\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB49/RB06 99 0.9406100593734508\n",
      "Processing RB06 NB77\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB77/RB06 162 0.6489777777777775\n",
      "Processing RB06 NB12\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB12/RB06 54 0.9814814814814814\n",
      "Processing RB06 NB38\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB38/RB06 117 0.9163561984074804\n",
      "Processing RB06 NB64\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB64/RB06 60 0.9430473372781064\n",
      "Processing RB06 NB71\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB71/RB06 85 0.9764705882352942\n",
      "Processing RB06 NB24\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB24/RB06 42 0.8832199546485261\n",
      "Processing RB06 NB94\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB94/RB06 80 0.9504687500000001\n",
      "Processing RB06 NB50\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB50/RB06 86 0.8560252556378684\n",
      "Processing RB06 NB65\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB65/RB06 54 0.9451303155006858\n",
      "Processing RB06 NB15\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB15/RB06 93 0.9679731760897212\n",
      "Processing RB06 NB02\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB02/RB06 62 1.0\n",
      "Processing RB06 NB27\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB27/RB06 142 0.9172671418313492\n",
      "Processing RB06 NB91\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB91/RB06 63 0.9221466364323508\n",
      "Processing RB06 NB19\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB19/RB06 39 0.9243918474687706\n",
      "Processing RB06 NB51\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB51/RB06 69 0.9710144927536232\n",
      "Processing RB06 NB52\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB52/RB06 71 0.9442570918468558\n",
      "Processing RB06 NB35\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB35/RB06 129 0.8476280704721125\n",
      "Processing RB06 NB09\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB09/RB06 75 0.8587946666666668\n",
      "Processing RB06 NB11\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB11/RB06 132 0.9404791718841301\n",
      "Processing RB06 NB63\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB63/RB06 51 0.9419454056132256\n",
      "Processing RB06 NB42\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB42/RB06 74 0.9464937910883857\n",
      "Processing RB06 NB18\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB18/RB06 158 0.9197102054807025\n",
      "Processing RB06 NB90\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB90/RB06 67 0.9850746268656717\n",
      "Processing RB06 NB05\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB05/RB06 137 0.9497529121828078\n",
      "Processing RB06 NB39\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB39/RB06 73 1.0\n",
      "Processing RB06 NB69\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB69/RB06 131 0.9620651477186642\n",
      "Processing RB06 NB25\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB25/RB06 87 0.9885057471264368\n",
      "Processing RB08 NB87\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB87/RB08 95 0.9556845762012233\n",
      "Processing RB08 NB03\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB03/RB08 36 1.0\n",
      "Processing RB08 NB20\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB20/RB08 40 1.0\n",
      "Processing RB08 NB48\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB48/RB08 81 0.9382716049382718\n",
      "Processing RB08 NB70\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB70/RB08 135 0.9276687496824667\n",
      "Processing RB08 NB04\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB04/RB08 65 0.9098224852071007\n",
      "Processing RB08 NB37\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB37/RB08 111 0.9289786499428574\n",
      "Processing RB08 NB88\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB88/RB08 62 0.9838709677419355\n",
      "Processing RB08 NB14\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB14/RB08 110 0.9728925619834712\n",
      "Processing RB08 NB82\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB82/RB08 87 0.8976086669309022\n",
      "Processing RB08 NB54\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB54/RB08 109 0.9726454002188369\n",
      "Processing RB08 NB44\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB44/RB08 81 0.9144947416552356\n",
      "Processing RB08 NB57\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB57/RB08 76 1.0\n",
      "Processing RB08 NB28\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB28/RB08 107 0.9176249323493135\n",
      "Processing RB08 NB40\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB40/RB08 128 0.9313225746154785\n",
      "Processing RB08 NB81\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB81/RB08 95 0.9480332409972299\n",
      "Processing RB08 NB13\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB13/RB08 97 0.9087044319268786\n",
      "Processing RB08 NB53\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB53/RB08 106 0.9534506673294062\n",
      "Processing RB08 NB80\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB80/RB08 150 0.928256\n",
      "Processing RB08 NB47\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB47/RB08 88 0.9528061224489797\n",
      "Processing RB08 NB84\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB84/RB08 96 0.9895833333333334\n",
      "Processing RB08 NB72\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB72/RB08 64 0.8800582885742188\n",
      "Processing RB08 NB10\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB10/RB08 80 0.9751562500000001\n",
      "Processing RB08 NB43\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB43/RB08 121 0.9753432142613209\n",
      "Processing RB08 NB58\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB58/RB08 38 0.9736842105263157\n",
      "Processing RB08 NB79\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB79/RB08 54 0.9259259259259259\n",
      "Processing RB08 NB36\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB36/RB08 98 0.9387755102040816\n",
      "Processing RB08 NB30\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB30/RB08 117 0.949589570014729\n",
      "Processing RB08 NB60\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB60/RB08 72 0.9724151234567903\n",
      "Processing RB08 NB45\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB45/RB08 120 0.9026921296296295\n",
      "Processing RB08 NB93\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB93/RB08 168 0.9471371882086166\n",
      "Processing RB08 NB31\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB31/RB08 112 0.6715561224489797\n",
      "Processing RB08 NB75\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB75/RB08 73 0.7807833590819941\n",
      "Processing RB08 NB21\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB21/RB08 49 0.9795918367346939\n",
      "Processing RB08 NB86\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB86/RB08 159 0.8487006051975791\n",
      "Processing RB08 NB07\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB07/RB08 64 1.0\n",
      "Processing RB08 NB34\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB34/RB08 98 0.9497052673630885\n",
      "Processing RB08 NB62\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB62/RB08 118 0.9416785552563797\n",
      "Processing RB08 NB23\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB23/RB08 116 0.9404694483168642\n",
      "Processing RB08 NB92\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB92/RB08 52 0.9245562130177513\n",
      "Processing RB08 NB73\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB73/RB08 76 0.9349030470914129\n",
      "Processing RB08 NB67\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB67/RB08 115 0.9654442344045369\n",
      "Processing RB08 NB29\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB29/RB08 109 0.9282765096380083\n",
      "Processing RB08 NB33\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB33/RB08 116 0.948870392390012\n",
      "Processing RB08 NB06\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB06/RB08 91 0.9230769230769231\n",
      "Processing RB08 NB78\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB78/RB08 27 1.0\n",
      "Processing RB08 NB95\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB95/RB08 78 0.9492110453648915\n",
      "Processing RB08 NB26\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB26/RB08 103 0.9518333490432652\n",
      "Processing RB08 NB16\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB16/RB08 41 0.975609756097561\n",
      "Processing RB08 NB76\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB76/RB08 84 0.9880952380952381\n",
      "Processing RB08 NB59\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB59/RB08 58 0.7848568790397047\n",
      "Processing RB08 NB68\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB68/RB08 63 0.8928721400376731\n",
      "Processing RB08 NB55\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB55/RB08 87 0.9885057471264368\n",
      "Processing RB08 NB74\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB74/RB08 94 0.6277311626907085\n",
      "Processing RB08 NB89\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB89/RB08 64 0.95361328125\n",
      "Processing RB08 NB83\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB83/RB08 113 0.9823791996240896\n",
      "Processing RB08 NB22\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB22/RB08 84 0.964709197170932\n",
      "Processing RB08 NB96\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB96/RB08 40 0.975\n",
      "Processing RB08 NB32\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB32/RB08 112 0.9734527320973033\n",
      "Processing RB08 NB41\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB41/RB08 112 0.973373724489796\n",
      "Processing RB08 NB17\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB17/RB08 52 0.4789497510185604\n",
      "Processing RB08 NB66\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB66/RB08 93 0.9679731760897214\n",
      "Processing RB08 NB61\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB61/RB08 137 0.9927007299270072\n",
      "Processing RB08 NB56\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB56/RB08 119 0.9096815196666902\n",
      "Processing RB08 NB46\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB46/RB08 136 0.9563148788927333\n",
      "Processing RB08 NB49\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB49/RB08 115 0.9655941481055316\n",
      "Processing RB08 NB77\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB77/RB08 93 0.9364088333911434\n",
      "Processing RB08 NB12\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB12/RB08 32 1.0\n",
      "Processing RB08 NB38\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB38/RB08 131 0.9251209136996678\n",
      "Processing RB08 NB64\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB64/RB08 104 0.9713387573964497\n",
      "Processing RB08 NB71\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB71/RB08 97 0.9794877245190775\n",
      "Processing RB08 NB24\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB24/RB08 100 0.9506000000000001\n",
      "Processing RB08 NB94\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB94/RB08 111 0.9642074506939371\n",
      "Processing RB08 NB50\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB50/RB08 129 0.9540273416410108\n",
      "Processing RB08 NB65\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB65/RB08 65 0.9846153846153847\n",
      "Processing RB08 NB15\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB15/RB08 97 0.9002731536925055\n",
      "Processing RB08 NB02\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB02/RB08 64 0.9375\n",
      "Processing RB08 NB27\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB27/RB08 163 0.9756106740938685\n",
      "Processing RB08 NB08\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB08/RB08 58 0.93251055803846\n",
      "Processing RB08 NB91\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB91/RB08 75 0.9086103703703705\n",
      "Processing RB08 NB19\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB19/RB08 77 0.9357395850902346\n",
      "Processing RB08 NB51\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB51/RB08 91 0.9673939151055442\n",
      "Processing RB08 NB52\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB52/RB08 113 0.9823008849557522\n",
      "Processing RB08 NB35\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB35/RB08 132 0.8614032890892394\n",
      "Processing RB08 NB09\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB09/RB08 88 0.8453964359504131\n",
      "Processing RB08 NB11\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB11/RB08 59 0.9322033898305085\n",
      "Processing RB08 NB63\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB63/RB08 71 0.8219146728515625\n",
      "Processing RB08 NB42\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB42/RB08 110 0.9639669421487606\n",
      "Processing RB08 NB18\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB18/RB08 115 0.991304347826087\n",
      "Processing RB08 NB90\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB90/RB08 91 0.9350320009660668\n",
      "Processing RB08 NB05\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB05/RB08 86 0.9653866955110872\n",
      "Processing RB08 NB39\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB39/RB08 117 0.9007892280140223\n",
      "Processing RB08 NB69\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB69/RB08 122 0.9111076477766861\n",
      "Processing RB08 NB25\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB25/RB08 104 0.9615384615384616\n",
      "Processing RB05 NB87\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB87/RB05 181 0.9130673666859985\n",
      "Processing RB05 NB03\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB03/RB05 166 0.9642532971193817\n",
      "Processing RB05 NB20\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB20/RB05 90 0.9888888888888889\n",
      "Processing RB05 NB48\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB48/RB05 124 0.9521331945889697\n",
      "Processing RB05 NB70\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB70/RB05 146 0.7552263374485594\n",
      "Processing RB05 NB04\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB04/RB05 136 0.9415549307958476\n",
      "Processing RB05 NB37\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB37/RB05 145 0.9524812005412278\n",
      "Processing RB05 NB88\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB88/RB05 84 0.929705215419501\n",
      "Processing RB05 NB14\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB14/RB05 180 0.8656440329218105\n",
      "Processing RB05 NB82\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB82/RB05 114 0.8802708525700215\n",
      "Processing RB05 NB54\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB54/RB05 119 0.9831932773109243\n",
      "Processing RB05 NB44\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB44/RB05 179 0.9833026434880309\n",
      "Processing RB05 NB57\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB57/RB05 104 0.9341609296768322\n",
      "Processing RB05 NB28\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB28/RB05 135 0.9705898491083675\n",
      "Processing RB05 NB40\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB40/RB05 280 0.9537244897959183\n",
      "Processing RB05 NB81\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB81/RB05 92 0.863268728938933\n",
      "Processing RB05 NB13\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB13/RB05 136 0.9708044982698963\n",
      "Processing RB05 NB53\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB53/RB05 69 0.8066406250000001\n",
      "Processing RB05 NB80\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB80/RB05 202 0.9704195667091462\n",
      "Processing RB05 NB47\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB47/RB05 131 0.9772736068068415\n",
      "Processing RB05 NB84\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB84/RB05 262 0.9585666238599772\n",
      "Processing RB05 NB72\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB72/RB05 100 0.8831999999999999\n",
      "Processing RB05 NB10\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB10/RB05 110 0.8521412471825693\n",
      "Processing RB05 NB43\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB43/RB05 266 0.8902537875583868\n",
      "Processing RB05 NB58\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB58/RB05 9 1.0\n",
      "Processing RB05 NB79\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB79/RB05 89 0.9775280898876404\n",
      "Processing RB05 NB36\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB36/RB05 129 1.0\n",
      "Processing RB05 NB30\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB30/RB05 202 0.9512511392301863\n",
      "Processing RB05 NB60\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB60/RB05 137 0.8864360513706205\n",
      "Processing RB05 NB45\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB45/RB05 145 0.7648113898509052\n",
      "Processing RB05 NB93\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB93/RB05 182 0.7773381982089624\n",
      "Processing RB05 NB31\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB31/RB05 125 0.9359999999999998\n",
      "Processing RB05 NB75\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB75/RB05 120 0.7839669421487602\n",
      "Processing RB05 NB21\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB21/RB05 111 0.9381477898643712\n",
      "Processing RB05 NB86\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB86/RB05 211 0.8635336109458658\n",
      "Processing RB05 NB07\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB07/RB05 133 0.9553394765108255\n",
      "Processing RB05 NB34\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB34/RB05 139 0.8949623904209791\n",
      "Processing RB05 NB62\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB62/RB05 130 0.9770414201183433\n",
      "Processing RB05 NB23\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB23/RB05 217 0.909617957484763\n",
      "Processing RB05 NB92\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB92/RB05 90 1.0\n",
      "Processing RB05 NB73\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB73/RB05 90 0.956170096021948\n",
      "Processing RB05 NB67\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB67/RB05 163 0.9575802973482404\n",
      "Processing RB05 NB29\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB29/RB05 126 0.9841899722852104\n",
      "Processing RB05 NB33\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB33/RB05 114 0.8654322787578363\n",
      "Processing RB05 NB06\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB06/RB05 209 0.9575303946215543\n",
      "Processing RB05 NB78\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB78/RB05 106 0.9625311498754006\n",
      "Processing RB05 NB95\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB95/RB05 115 0.9401837757869649\n",
      "Processing RB05 NB26\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB26/RB05 139 0.9504132939184593\n",
      "Processing RB05 NB16\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB16/RB05 9 0.888888888888889\n",
      "Processing RB05 NB76\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB76/RB05 104 0.9618158284023669\n",
      "Processing RB05 NB59\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB59/RB05 131 0.9544898315948953\n",
      "Processing RB05 NB68\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB68/RB05 157 0.9560202310503157\n",
      "Processing RB05 NB55\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB55/RB05 110 0.8412802404207363\n",
      "Processing RB05 NB74\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB74/RB05 124 0.9679370447450572\n",
      "Processing RB05 NB89\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB89/RB05 106 0.9255962976148093\n",
      "Processing RB05 NB83\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB83/RB05 245 0.9716284881299458\n",
      "Processing RB05 NB22\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB22/RB05 116 0.9742865636147444\n",
      "Processing RB05 NB96\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB96/RB05 179 0.9449736101039538\n",
      "Processing RB05 NB32\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB32/RB05 118 0.6059955010005892\n",
      "Processing RB05 NB41\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB41/RB05 231 0.9357395850902346\n",
      "Processing RB05 NB17\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB17/RB05 136 0.9491221472369221\n",
      "Processing RB05 NB66\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB66/RB05 124 0.9838709677419355\n",
      "Processing RB05 NB85\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB85/RB05 15 0.7865270823850704\n",
      "Processing RB05 NB61\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB61/RB05 207 0.8550236313742524\n",
      "Processing RB05 NB56\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB56/RB05 265 0.9440917862480231\n",
      "Processing RB05 NB46\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB46/RB05 218 0.9282765096380082\n",
      "Processing RB05 NB49\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB49/RB05 194 0.853703900520778\n",
      "Processing RB05 NB77\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB77/RB05 233 0.9871613033947946\n",
      "Processing RB05 NB12\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB12/RB05 69 0.9287964713295527\n",
      "Processing RB05 NB38\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB38/RB05 247 0.9170658633162441\n",
      "Processing RB05 NB64\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB64/RB05 129 0.94609698936362\n",
      "Processing RB05 NB71\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB71/RB05 134 0.9555023390510133\n",
      "Processing RB05 NB24\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB24/RB05 80 0.9625\n",
      "Processing RB05 NB94\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB94/RB05 135 0.9852400548696842\n",
      "Processing RB05 NB50\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB50/RB05 160 0.95046875\n",
      "Processing RB05 NB65\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB65/RB05 93 0.8828766331367788\n",
      "Processing RB05 NB15\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB15/RB05 131 0.9621816910436454\n",
      "Processing RB05 NB02\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB02/RB05 126 0.8996988566149562\n",
      "Processing RB05 NB01\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB01/RB05 95 0.9687522962530983\n",
      "Processing RB05 NB27\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB27/RB05 220 0.9117308414725773\n",
      "Processing RB05 NB08\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB08/RB05 118 0.9745762711864406\n",
      "Processing RB05 NB91\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB91/RB05 124 0.9082343862094027\n",
      "Processing RB05 NB19\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB19/RB05 72 0.9724151234567903\n",
      "Processing RB05 NB51\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB51/RB05 111 0.8229362099037076\n",
      "Processing RB05 NB52\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB52/RB05 138 0.7517666346316917\n",
      "Processing RB05 NB35\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB35/RB05 241 0.9346555007847462\n",
      "Processing RB05 NB09\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB09/RB05 114 0.9146458019471578\n",
      "Processing RB05 NB11\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB11/RB05 207 0.9524109097636765\n",
      "Processing RB05 NB63\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB63/RB05 127 0.968751937503875\n",
      "Processing RB05 NB42\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB42/RB05 121 0.9589508913325593\n",
      "Processing RB05 NB18\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB18/RB05 239 0.9668935187793014\n",
      "Processing RB05 NB90\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB90/RB05 67 0.9409668077522835\n",
      "Processing RB05 NB05\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB05/RB05 175 0.9549038833819241\n",
      "Processing RB05 NB39\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB39/RB05 126 0.968442932728647\n",
      "Processing RB05 NB69\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB69/RB05 199 0.9261826417002055\n",
      "Processing RB05 NB25\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB25/RB05 161 0.9752710157787122\n",
      "Processing RB07 NB87\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB87/RB07 62 0.9362643080124871\n",
      "Processing RB07 NB03\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB03/RB07 85 0.9309244860573989\n",
      "Processing RB07 NB20\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB20/RB07 57 0.9307479224376731\n",
      "Processing RB07 NB48\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB48/RB07 111 0.6984126984126984\n",
      "Processing RB07 NB70\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB70/RB07 114 0.9482107855048519\n",
      "Processing RB07 NB04\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB04/RB07 51 0.9803921568627451\n",
      "Processing RB07 NB37\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB37/RB07 98 0.8097487016464225\n",
      "Processing RB07 NB88\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB88/RB07 46 0.9782608695652174\n",
      "Processing RB07 NB14\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB14/RB07 115 0.9484310018903591\n",
      "Processing RB07 NB82\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB82/RB07 90 1.0\n",
      "Processing RB07 NB54\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB54/RB07 97 0.9693899129260973\n",
      "Processing RB07 NB44\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB44/RB07 123 0.9437504131138872\n",
      "Processing RB07 NB57\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB57/RB07 82 0.9634146341463414\n",
      "Processing RB07 NB28\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB28/RB07 76 0.8470850342615545\n",
      "Processing RB07 NB40\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB40/RB07 120 0.8522222222222221\n",
      "Processing RB07 NB81\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB81/RB07 103 0.923461212178339\n",
      "Processing RB07 NB13\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB13/RB07 92 0.9782608695652173\n",
      "Processing RB07 NB53\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB53/RB07 93 0.989247311827957\n",
      "Processing RB07 NB80\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB80/RB07 105 0.9346861030126337\n",
      "Processing RB07 NB47\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB47/RB07 82 0.9637120761451519\n",
      "Processing RB07 NB84\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB84/RB07 81 0.6378632812500001\n",
      "Processing RB07 NB72\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB72/RB07 64 0.9687500000000001\n",
      "Processing RB07 NB10\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB10/RB07 80 0.9750000000000001\n",
      "Processing RB07 NB43\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB43/RB07 80 0.9504687500000001\n",
      "Processing RB07 NB58\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB58/RB07 24 0.9130434782608695\n",
      "Processing RB07 NB79\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB79/RB07 85 0.9082190107877064\n",
      "Processing RB07 NB36\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB36/RB07 104 0.9525675992831134\n",
      "Processing RB07 NB30\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB30/RB07 84 0.9245562130177516\n",
      "Processing RB07 NB60\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB60/RB07 101 0.9312812469365748\n",
      "Processing RB07 NB45\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB45/RB07 110 0.9372689706987227\n",
      "Processing RB07 NB31\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB31/RB07 117 0.8233499603212511\n",
      "Processing RB07 NB75\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB75/RB07 89 0.9776543365736651\n",
      "Processing RB07 NB21\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB21/RB07 86 0.9767441860465116\n",
      "Processing RB07 NB86\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB86/RB07 112 0.9302327236698251\n",
      "Processing RB07 NB07\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB07/RB07 90 0.9561700960219479\n",
      "Processing RB07 NB34\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB34/RB07 99 0.9600040812162026\n",
      "Processing RB07 NB62\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB62/RB07 71 0.9720293592541163\n",
      "Processing RB07 NB23\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB23/RB07 119 0.9011968603556103\n",
      "Processing RB07 NB92\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB92/RB07 30 0.9344444444444444\n",
      "Processing RB07 NB73\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB73/RB07 95 0.9789473684210526\n",
      "Processing RB07 NB67\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB67/RB07 91 0.8288854003139717\n",
      "Processing RB07 NB29\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB29/RB07 88 0.966167355371901\n",
      "Processing RB07 NB33\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB33/RB07 91 0.8503511945125278\n",
      "Processing RB07 NB06\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB06/RB07 145 0.9590487514863257\n",
      "Processing RB07 NB78\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB78/RB07 82 0.9284543172617927\n",
      "Processing RB07 NB95\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB95/RB07 37 0.9189189189189189\n",
      "Processing RB07 NB26\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB26/RB07 89 0.9662921348314607\n",
      "Processing RB07 NB16\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB16/RB07 27 1.0\n",
      "Processing RB07 NB76\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB76/RB07 104 0.89860771601658\n",
      "Processing RB07 NB59\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB59/RB07 100 0.8244999999999999\n",
      "Processing RB07 NB68\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB68/RB07 111 0.9731352974596218\n",
      "Processing RB07 NB55\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB55/RB07 64 1.0\n",
      "Processing RB07 NB74\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB74/RB07 76 0.9736842105263159\n",
      "Processing RB07 NB89\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB89/RB07 77 0.9870129870129871\n",
      "Processing RB07 NB83\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB83/RB07 106 0.9812210751156996\n",
      "Processing RB07 NB22\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB22/RB07 123 0.8977138075961365\n",
      "Processing RB07 NB96\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB96/RB07 45 0.9555555555555555\n",
      "Processing RB07 NB32\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB32/RB07 89 0.8327562326869806\n",
      "Processing RB07 NB41\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB41/RB07 107 0.953966513828494\n",
      "Processing RB07 NB17\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB17/RB07 81 0.9393328704409333\n",
      "Processing RB07 NB66\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB66/RB07 79 0.9006569460022433\n",
      "Processing RB07 NB85\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB85/RB07 43 0.9767441860465116\n",
      "Processing RB07 NB61\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB61/RB07 136 0.8723229505902707\n",
      "Processing RB07 NB56\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB56/RB07 82 0.9637120761451518\n",
      "Processing RB07 NB46\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB46/RB07 93 0.8616024973985432\n",
      "Processing RB07 NB49\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB49/RB07 129 0.9692326182320775\n",
      "Processing RB07 NB77\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB77/RB07 105 0.8546809199870423\n",
      "Processing RB07 NB12\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB12/RB07 59 0.9663889686871587\n",
      "Processing RB07 NB38\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB38/RB07 122 0.9355628885237091\n",
      "Processing RB07 NB64\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB64/RB07 80 0.9751562500000001\n",
      "Processing RB07 NB71\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB71/RB07 88 0.9011376381990067\n",
      "Processing RB07 NB24\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB24/RB07 78 0.9745233399079554\n",
      "Processing RB07 NB94\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB94/RB07 42 1.0\n",
      "Processing RB07 NB50\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB50/RB07 118 0.9335606366765833\n",
      "Processing RB07 NB65\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB65/RB07 51 0.8858131487889274\n",
      "Processing RB07 NB15\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB15/RB07 74 0.7610192837465567\n",
      "Processing RB07 NB02\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB02/RB07 81 0.9632677945435149\n",
      "Processing RB07 NB01\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB01/RB07 72 0.905456961591221\n",
      "Processing RB07 NB27\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB27/RB07 118 0.8623082204120186\n",
      "Processing RB07 NB08\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB08/RB07 102 0.9422281023135899\n",
      "Processing RB07 NB19\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB19/RB07 72 0.9587191358024691\n",
      "Processing RB07 NB51\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB51/RB07 117 0.949589570014729\n",
      "Processing RB07 NB35\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB35/RB07 81 0.9754610577655847\n",
      "Processing RB07 NB09\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB09/RB07 104 0.9245562130177515\n",
      "Processing RB07 NB11\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB11/RB07 109 0.9637220479232511\n",
      "Processing RB07 NB63\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB63/RB07 81 0.9032046831162821\n",
      "Processing RB07 NB42\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB42/RB07 149 0.9600918877528037\n",
      "Processing RB07 NB18\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB18/RB07 126 0.8826465624462599\n",
      "Processing RB07 NB05\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB05/RB07 117 0.9575571626853677\n",
      "Processing RB07 NB39\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB39/RB07 81 0.9042527768839816\n",
      "Processing RB07 NB69\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB69/RB07 91 0.989010989010989\n",
      "Processing RB07 NB25\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "Padding is 0. Implementing soft alignment\n",
      "Too many positions, either contaminated or sequencing error\n",
      "Variant: NB25/RB07 97 0.9587628865979381\n"
     ]
    }
   ],
   "source": [
    "variant_df = get_variant_df_soft(demultiplexer_path, template_seq, ref_name, barcode_dicts, merge = False, min_depth= 5, padding=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.DataFrame(variant_df).to_pickle(\"/home/emre/github_repo/MinION/results/6_5site_ParPgb/ParPgb_variant_df_5site_whole_Sequence_adj_param.pkl\")\n",
    "pd.DataFrame(variant_df).to_csv(\"/home/emre/github_repo/MinION/results/6_5site_ParPgb/ParPgb_variant_df_5site_whole_Sequence_adj_param.pkl\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "variant_df = get_variant_df_soft(demultiplexer_path, template_seq, ref_name, barcode_dicts, merge = False, min_depth= 5, padding=0, alignment_name=\"alignment_minimap_site_saturation.bam\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.DataFrame(variant_df).to_pickle(\"/home/emre/github_repo/MinION/results/6_5site_ParPgb/ParPgb_variant_df_5site_whole_Sequence_adj_param_site_saturation.pkl\")\n",
    "pd.DataFrame(variant_df).to_csv(\"/home/emre/github_repo/MinION/results/6_5site_ParPgb/ParPgb_variant_df_5site_whole_Sequence_adj_param_site_saturation.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>RBC</th>\n",
       "      <th>FBC</th>\n",
       "      <th>Position</th>\n",
       "      <th>Variant</th>\n",
       "      <th>Alignment Probability</th>\n",
       "      <th>Alignment Count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RB12</td>\n",
       "      <td>NB87</td>\n",
       "      <td>[354, 432, 374]</td>\n",
       "      <td>T354A_C432A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>174.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>RB12</td>\n",
       "      <td>NB03</td>\n",
       "      <td>[523, 599, 412]</td>\n",
       "      <td>G523A_G599A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>130.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>RB12</td>\n",
       "      <td>NB20</td>\n",
       "      <td>[367, 132, 278]</td>\n",
       "      <td>A367DEL_C132T_T278G</td>\n",
       "      <td>0.970804</td>\n",
       "      <td>69.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>RB12</td>\n",
       "      <td>NB48</td>\n",
       "      <td>[376, 499, 262]</td>\n",
       "      <td>G376DEL_T499DEL_C262A</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>107.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>RB12</td>\n",
       "      <td>NB70</td>\n",
       "      <td>[342, 376, 370]</td>\n",
       "      <td>T342A_G376A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>152.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>732</th>\n",
       "      <td>RB07</td>\n",
       "      <td>NB18</td>\n",
       "      <td>[135, 217, 171]</td>\n",
       "      <td>A217DEL_C171G</td>\n",
       "      <td>0.882647</td>\n",
       "      <td>126.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>733</th>\n",
       "      <td>RB07</td>\n",
       "      <td>NB05</td>\n",
       "      <td>[486, 479, 329]</td>\n",
       "      <td>C486G_C479T_A329C</td>\n",
       "      <td>0.957557</td>\n",
       "      <td>117.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>734</th>\n",
       "      <td>RB07</td>\n",
       "      <td>NB39</td>\n",
       "      <td>[219, 274, 501]</td>\n",
       "      <td>G219T_T274G_G501C</td>\n",
       "      <td>0.904253</td>\n",
       "      <td>81.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>735</th>\n",
       "      <td>RB07</td>\n",
       "      <td>NB69</td>\n",
       "      <td>[262, 200, 328]</td>\n",
       "      <td>C262G_T200A</td>\n",
       "      <td>0.989011</td>\n",
       "      <td>91.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>736</th>\n",
       "      <td>RB07</td>\n",
       "      <td>NB25</td>\n",
       "      <td>[157, 251, 445]</td>\n",
       "      <td>T157A</td>\n",
       "      <td>0.958763</td>\n",
       "      <td>97.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>737 rows  6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      RBC   FBC         Position                Variant  \\\n",
       "0    RB12  NB87  [354, 432, 374]            T354A_C432A   \n",
       "1    RB12  NB03  [523, 599, 412]            G523A_G599A   \n",
       "2    RB12  NB20  [367, 132, 278]    A367DEL_C132T_T278G   \n",
       "3    RB12  NB48  [376, 499, 262]  G376DEL_T499DEL_C262A   \n",
       "4    RB12  NB70  [342, 376, 370]            T342A_G376A   \n",
       "..    ...   ...              ...                    ...   \n",
       "732  RB07  NB18  [135, 217, 171]          A217DEL_C171G   \n",
       "733  RB07  NB05  [486, 479, 329]      C486G_C479T_A329C   \n",
       "734  RB07  NB39  [219, 274, 501]      G219T_T274G_G501C   \n",
       "735  RB07  NB69  [262, 200, 328]            C262G_T200A   \n",
       "736  RB07  NB25  [157, 251, 445]                  T157A   \n",
       "\n",
       "     Alignment Probability  Alignment Count  \n",
       "0                      NaN            174.0  \n",
       "1                      NaN            130.0  \n",
       "2                 0.970804             69.0  \n",
       "3                 1.000000            107.0  \n",
       "4                      NaN            152.0  \n",
       "..                     ...              ...  \n",
       "732               0.882647            126.0  \n",
       "733               0.957557            117.0  \n",
       "734               0.904253             81.0  \n",
       "735               0.989011             91.0  \n",
       "736               0.958763             97.0  \n",
       "\n",
       "[737 rows x 6 columns]"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.DataFrame(variant_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Simulator - Call Variant of Alignments Minimap.bam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import glob\n",
    "template_seq = Path(\"/home/emre/github_repo/MinION/minION/refseq/hetcpiii_padded.fasta\")\n",
    "ref_name = \"HetCPIII\"\n",
    "Variant_dict = {\"Original Variant\" : [], \"Predicted Variant\": [] ,\"Depth\" : [], \"Alignment Probability\" : [], \"Alignment Count\" : []}\n",
    "\n",
    "sequence_folder = Path(\"/home/emre/github_repo/MinION/examples/data/min_read_depth/seq\")\n",
    "variant_folder = glob.glob(f\"{sequence_folder}/*\")\n",
    "\n",
    "for var in variant_folder:  \n",
    "    variant_name = os.path.basename(var)\n",
    "    \n",
    "    depth_folders = glob.glob(f\"{var}/depth*\")\n",
    "\n",
    "    for depth in depth_folders:\n",
    "\n",
    "        depth_name = os.path.basename(depth)\n",
    "        \n",
    "        bam_file = os.path.join(depth, \"alignment_minimap_Q10.bam\")\n",
    "\n",
    "        if not os.path.exists(os.path.join(depth, \"alignment_minimap_Q10.bam\")):\n",
    "            print(f\"Alignment file in {depth} does not exist, skipping\")\n",
    "            continue\n",
    "        else:\n",
    "\n",
    "            nn_variants = get_variant_soft(bam_file, template_seq, ref_name, padding = 50)\n",
    "            print(nn_variants)\n",
    "\n",
    "            if \"wt\" in variant_name:\n",
    "                variant_name = \"#PARENT#\"\n",
    "                Variant_dict[\"Original Variant\"].append(variant_name)\n",
    "            else:\n",
    "                Variant_dict[\"Original Variant\"].append(variant_name)\n",
    "            Variant_dict[\"Depth\"].append(depth_name)\n",
    "            Variant_dict[\"Predicted Variant\"].append(nn_variants[\"Variant\"])\n",
    "            Variant_dict[\"Alignment Count\"].append(nn_variants[\"Alignment Count\"])\n",
    "            Variant_dict[\"Alignment Probability\"].append(nn_variants[\"Alignment Probability\"])\n",
    "\n",
    "\n",
    "        \n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Call Variant with BF only"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import glob\n",
    "template_seq = Path(\"/home/emre/github_repo/MinION/minION/refseq/hetcpiii_padded.fasta\")\n",
    "ref_name = \"HetCPIII\"\n",
    "Variant_dict = {\"Original Variant\" : [], \"Predicted Variant\": [] ,\"Depth\" : [], \"Frequency\" : []}\n",
    "template = analyser.get_template_sequence(template_seq)\n",
    "sequence_folder = Path(\"/home/emre/github_repo/MinION/examples/data/min_read_depth/seq\")\n",
    "variant_folder = glob.glob(f\"{sequence_folder}/*\")\n",
    "\n",
    "for var in variant_folder:  \n",
    "    variant_name = os.path.basename(var)\n",
    "    \n",
    "    depth_folders = glob.glob(f\"{var}/depth*\")\n",
    "\n",
    "    for depth in depth_folders:\n",
    "\n",
    "        depth_name = os.path.basename(depth)\n",
    "        \n",
    "        bam_file = os.path.join(depth, \"alignment_minimap_Q10.bam\")\n",
    "\n",
    "        if not os.path.exists(os.path.join(depth, \"alignment_minimap_Q10.bam\")):\n",
    "            print(f\"Alignment file in {depth} does not exist, skipping\")\n",
    "            continue\n",
    "        else:\n",
    "            padding = 50\n",
    "            nn_variants = call_variant_BF(bam_file, \"HetCPIII\", range(padding, len(template) - padding + 1), template, qualities=False)\n",
    "            nn_variants[\"Variant\"] = \"_\".join(nn_variants[\"Variant\"])\n",
    "            if \"wt\" in variant_name:\n",
    "                variant_name = \"#PARENT#\"\n",
    "                Variant_dict[\"Original Variant\"].append(variant_name)\n",
    "            else:\n",
    "                Variant_dict[\"Original Variant\"].append(variant_name)\n",
    "            Variant_dict[\"Depth\"].append(depth_name)\n",
    "            Variant_dict[\"Predicted Variant\"].append(nn_variants[\"Variant\"])\n",
    "            Variant_dict[\"Frequency\"].append(nn_variants[\"Frequency\"])\n",
    "  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.DataFrame(Variant_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.DataFrame(Variant_dict).to_pickle(\"/home/emre/github_repo/MinION/results/2_hetcpiii_minion_errorprone/Simulation_Q10_BF_results.pkl\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "variant_df = pd.read_pickle(\"/home/emre/github_repo/MinION/results/2_hetcpiii_minion_errorprone/Simulation_Q20_results.pkl\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "variant_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "medaka",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
