{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Try to implement bayes based variant calling\n",
    "\n",
    "- For Parent first\n",
    "- Check for more mutatios later"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import\n",
    "import sys\n",
    "sys.path.append(\"/home/emre/github_repo/MinION\")\n",
    "from minION.util import IO_processor\n",
    "from minION import analyser\n",
    "from minION import consensus\n",
    "\n",
    "import importlib\n",
    "importlib.reload(analyser)\n",
    "importlib.reload(consensus)\n",
    "importlib.reload(IO_processor)\n",
    "from pathlib import Path\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import os\n",
    "import glob\n",
    "import numpy as np\n",
    "from Bio import SeqIO\n",
    "import matplotlib.pyplot as plt\n",
    "import gzip\n",
    "import math\n",
    "import re\n",
    "import pickle\n",
    "import itertools\n",
    "import pysam\n",
    "import subprocess"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_bases_from_pileup(bam_file, chrom, positions):\n",
    "    bases_dict = {position: {} for position in positions}\n",
    "    qualities_dict = {position: {} for position in positions}\n",
    "    \n",
    "    with pysam.AlignmentFile(bam_file, 'rb') as bam:\n",
    "        for pileup_column in bam.pileup(chrom, min(positions) - 1, max(positions) + 1,\n",
    "                                        min_base_quality=0, \n",
    "                                        min_mapping_quality=0, \n",
    "                                        truncate=True):\n",
    "            pos = pileup_column.pos + 1\n",
    "            if pos in positions:\n",
    "                for pileup_read in pileup_column.pileups:\n",
    "                    read_name = pileup_read.alignment.query_name\n",
    "\n",
    "                    # Handle deletions\n",
    "                    if pileup_read.is_del:\n",
    "                        base = '-'  # or any symbol you prefer to represent a deletion\n",
    "                        quality = 0  # Assign a default quality for deletions\n",
    "                    elif not pileup_read.is_refskip:\n",
    "                        base = pileup_read.alignment.query_sequence[pileup_read.query_position]\n",
    "                        quality = pileup_read.alignment.query_qualities[pileup_read.query_position]\n",
    "                    else:\n",
    "                        continue\n",
    "\n",
    "                    # Add base and quality to the dictionaries\n",
    "                    if read_name not in bases_dict[pos]:\n",
    "                        bases_dict[pos][read_name] = base\n",
    "                        qualities_dict[pos][read_name] = quality\n",
    "\n",
    "    # Get unique read names and sort them\n",
    "    read_names = sorted(set().union(*[bases_dict[pos].keys() for pos in bases_dict]))\n",
    "\n",
    "    # Create DataFrames\n",
    "    df_bases = pd.DataFrame(index=read_names, columns=positions)\n",
    "    df_qualities = pd.DataFrame(index=read_names, columns=positions)\n",
    "    \n",
    "    # Populate DataFrames\n",
    "    for pos in positions:\n",
    "        for read_name in bases_dict[pos]:\n",
    "            df_bases.at[read_name, pos] = bases_dict[pos][read_name]\n",
    "            df_qualities.at[read_name, pos] = qualities_dict[pos][read_name]\n",
    "    \n",
    "    # Fill NaN with \"-\" for bases and 0 for qualities\n",
    "    df_bases = df_bases.fillna(\"-\")\n",
    "    df_qualities = df_qualities.fillna(10) # 10 is the lowest quality filter we used for filtering\n",
    "\n",
    "    return df_bases, df_qualities\n",
    "\n",
    "\n",
    "def get_soft_pop_frequency(bam_file, template, reference, nb_positions, min_depth = 5):\n",
    "\n",
    "    # Min depth based on the alphabet size\n",
    "    \n",
    "    # Check also for variant by sampling random positions\n",
    "\n",
    "    bases_df = get_bases_from_pileup(bam_file, reference, nb_positions)\n",
    "    \n",
    "    frequency_df = bases_df.apply(get_variant_name, axis=1, args=(template, nb_positions)).value_counts().reset_index()\n",
    "\n",
    "    frequency_df.columns = ['Population', 'N_reads']\n",
    "    \n",
    "    frequency_df[\"Frequency\"] = frequency_df[\"N_reads\"] / frequency_df[\"N_reads\"].sum()\n",
    "\n",
    "\n",
    "    # Filter for frequency > 0.4 and depth > 15\n",
    "    frequency_df = frequency_df[(frequency_df[\"Frequency\"] > min_freq) & (frequency_df[\"N_reads\"] > min_depth)]\n",
    "\n",
    "    return frequency_df\n",
    "\n",
    "\n",
    "def add_neighbouring_positions(positions, nb_neighbours, max_index):\n",
    "    if positions is None:\n",
    "        return None\n",
    "    \n",
    "    elif isinstance(positions, int):\n",
    "        positions = [positions]\n",
    "\n",
    "    new_positions = []\n",
    "    for position in positions:\n",
    "        for new_pos in range(position - nb_neighbours, position + nb_neighbours + 1):\n",
    "            if 1 <= new_pos <= max_index:  # Check if the new position is within valid index range\n",
    "                new_positions.append(new_pos)\n",
    "    return sorted(set(new_positions))\n",
    "\n",
    "def calculate_mean_quality_for_reads(bases_df, qual_df, nb_positions, nb_neighbours):\n",
    "    if isinstance(nb_positions, int):\n",
    "        nb_positions = [nb_positions]  # Convert single integer to a list\n",
    "\n",
    "    read_mean_qualities = {}\n",
    "\n",
    "    for nb_position in nb_positions:\n",
    "        if nb_position not in bases_df.columns or nb_position not in qual_df.columns:\n",
    "            continue  # Skip positions that are not present in either DataFrame\n",
    "\n",
    "        neighbor_positions = range(nb_position - nb_neighbours, nb_position + nb_neighbours + 1)\n",
    "\n",
    "        for read_name in qual_df.index:\n",
    "            total_qual = 0\n",
    "            valid_count = 0\n",
    "\n",
    "            for position in neighbor_positions:\n",
    "                if position not in bases_df.columns:\n",
    "                    continue  # Skip positions that are outside the DataFrame's columns\n",
    "\n",
    "                base = bases_df.at[read_name, position]\n",
    "                quality = qual_df.at[read_name, position]\n",
    "\n",
    "                if base != \"-\" and not pd.isna(quality):\n",
    "                    total_qual += quality\n",
    "                    valid_count += 1\n",
    "\n",
    "            if valid_count == 0:\n",
    "                continue  # Skip if no valid qualities were found\n",
    "            else:\n",
    "                if read_name not in read_mean_qualities:\n",
    "                    read_mean_qualities[read_name] = {}\n",
    "                read_mean_qualities[read_name][nb_position] = total_qual / valid_count\n",
    "\n",
    "    # Convert the dictionary into a DataFrame\n",
    "    mean_quality_df = pd.DataFrame.from_dict(read_mean_qualities, orient='index')\n",
    "    updated_base_df = bases_df[nb_positions] \n",
    "\n",
    "    return updated_base_df, mean_quality_df\n",
    "\n",
    "def add_neighbouring_positions(positions, nb_neighbours, max_index):\n",
    "    if positions is None:\n",
    "        return None\n",
    "    \n",
    "    elif isinstance(positions, int):\n",
    "        positions = [positions]\n",
    "\n",
    "    new_positions = []\n",
    "    for position in positions:\n",
    "        for new_pos in range(position - nb_neighbours, position + nb_neighbours + 1):\n",
    "            if 1 <= new_pos <= max_index:  # Check if the new position is within valid index range\n",
    "                new_positions.append(new_pos)\n",
    "    return sorted(set(new_positions))\n",
    "\n",
    "def calculate_mean_quality_for_reads(bases_df, qual_df, nb_positions, nb_neighbours):\n",
    "    if isinstance(nb_positions, int):\n",
    "        nb_positions = [nb_positions]  # Convert single integer to a list\n",
    "\n",
    "    read_mean_qualities = {}\n",
    "\n",
    "    for nb_position in nb_positions:\n",
    "        if nb_position not in bases_df.columns or nb_position not in qual_df.columns:\n",
    "            continue  # Skip positions that are not present in either DataFrame\n",
    "\n",
    "        neighbor_positions = range(nb_position - nb_neighbours, nb_position + nb_neighbours + 1)\n",
    "\n",
    "        for read_name in qual_df.index:\n",
    "            total_qual = 0\n",
    "            valid_count = 0\n",
    "\n",
    "            for position in neighbor_positions:\n",
    "                if position not in bases_df.columns:\n",
    "                    continue  # Skip positions that are outside the DataFrame's columns\n",
    "\n",
    "                base = bases_df.at[read_name, position]\n",
    "                quality = qual_df.at[read_name, position]\n",
    "\n",
    "                if base != \"-\" and not pd.isna(quality):\n",
    "                    total_qual += quality\n",
    "                    valid_count += 1\n",
    "\n",
    "            if valid_count == 0:\n",
    "                continue  # Skip if no valid qualities were found\n",
    "            else:\n",
    "                if read_name not in read_mean_qualities:\n",
    "                    read_mean_qualities[read_name] = {}\n",
    "                read_mean_qualities[read_name][nb_position] = total_qual / valid_count\n",
    "\n",
    "    # Convert the dictionary into a DataFrame\n",
    "    mean_quality_df = pd.DataFrame.from_dict(read_mean_qualities, orient='index')\n",
    "    updated_base_df = bases_df[nb_positions] \n",
    "\n",
    "    return updated_base_df, mean_quality_df\n",
    "\n",
    "def get_non_error_prop(quality_score):\n",
    "    \"\"\"Convert quality score to non-error probability.\"\"\"\n",
    "    return 1 - 10 ** (-quality_score / 10)\n",
    "\n",
    "def get_softmax_count_df(bases_df, qual_df, nb_positions):\n",
    "\n",
    "    alphabet = \"ACTG-\"\n",
    "    softmax_counts = {position: [] for position in nb_positions}\n",
    "    \n",
    "    for position in nb_positions:\n",
    "        for base in alphabet:\n",
    "            base_mask = bases_df[position] == base\n",
    "            base_counts = base_mask.sum()\n",
    "            # Calculate the non-error probability for each base and sum them up\n",
    "            soft_count = sum(base_mask * qual_df[position].apply(get_non_error_prop))\n",
    "            softmax_counts[position].append(soft_count)\n",
    "\n",
    "    softmax_count_df = pd.DataFrame(softmax_counts, columns=nb_positions, index=list(alphabet))\n",
    "\n",
    "    # Apply softmax to each column (position)\n",
    "    softmax_count_df = softmax_count_df.apply(lambda x: x / x.sum(), axis=0)\n",
    "\n",
    "    return softmax_count_df\n",
    "\n",
    "def get_softmax(soft_count):\n",
    "    \"\"\"Calculate the softmax of a dictionary of soft counts.\"\"\"\n",
    "    # Calculate the sum of the non-error probabilities\n",
    "    total = sum(soft_count.values())\n",
    "    # Calculate the softmax for each base\n",
    "    return {base: count / total for base, count in soft_count.items()}\n",
    "\n",
    "def call_potential_populations(softmax_df, ref_seq):\n",
    "    positions = softmax_df.columns\n",
    "    top_combinations = []\n",
    "    \n",
    "    # Get the top 2 variants for each position\n",
    "    for position in positions:\n",
    "        top_variants = softmax_df[position].nlargest(2)\n",
    "\n",
    "        if top_variants.iloc[1] < 0.1:\n",
    "            top_combinations.append([top_variants.index[0]])\n",
    "        \n",
    "        else:\n",
    "            top_combinations.append(top_variants.index.tolist())\n",
    "\n",
    "        potential_combinations = list(itertools.product(*top_combinations))\n",
    "\n",
    "    \n",
    "    variants = {\"Variant\" : [], \"Probability\" : []}\n",
    "    \n",
    "    for combination in potential_combinations:\n",
    "        final_variant = []\n",
    "        for i, pos in enumerate(positions):\n",
    "\n",
    "            if combination[i] == ref_seq[pos - 1]:\n",
    "                continue\n",
    "\n",
    "            elif combination[i] == \"-\":\n",
    "                var = f\"{ref_seq[pos - 1]}{pos}DEL\"\n",
    "                final_variant.append(var)\n",
    "            else:\n",
    "                var = f\"{ref_seq[pos - 1]}{pos}{combination[i]}\"\n",
    "                final_variant.append(var)\n",
    "\n",
    "        final_variant = '_'.join(final_variant)\n",
    "        if final_variant == \"\":\n",
    "            final_variant = \"#PARENT#\"\n",
    "\n",
    "        joint_prob = np.prod([softmax_df.at[combination[i], positions[i]] for i in range(len(positions))])\n",
    "    \n",
    "        variants[\"Variant\"].append(final_variant)\n",
    "        variants[\"Probability\"].append(joint_prob)\n",
    "\n",
    "    return variants\n",
    "\n",
    "def get_variant_soft(bam_file, template_seq, ref_name, padding = 50):\n",
    "\n",
    "    variants = {\"Variant\" : [], \"Position\" : [], \"Alignment Probability\" : [], \"Alignment Count\" : []}\n",
    "\n",
    "    alignment_count = int(subprocess.run(f\"samtools view -c {bam_file}\", shell=True, capture_output=True).stdout.decode(\"utf-8\").strip())\n",
    "\n",
    "    if alignment_count < 5:\n",
    "        print(\"Not enough alignments\")\n",
    "        return None\n",
    "\n",
    "\n",
    "    template = analyser.get_template_sequence(template_seq)\n",
    "\n",
    "    padding_start, padding_end = padding, padding\n",
    "    range_positions = range(padding_start + 1, len(template) - padding_end + 1) \n",
    "\n",
    "    freq_dist = pd.DataFrame(analyser.get_highest_non_ref_base_freq_2(bam_file, ref_name, range_positions, template, qualities=False)[0]).T.rename(columns={0:\"Base\", 1:\"Frequency\"})\n",
    "\n",
    "    nb_positions = analyser.get_nb_positions(freq_dist, 0.3)\n",
    "\n",
    "    available_positions = [pos for pos in range_positions if pos not in nb_positions]\n",
    "\n",
    "    if len(nb_positions) == 0:\n",
    "        # Select random 3 positions\n",
    "        nb_positions = np.random.choice(available_positions, 3, replace=False)\n",
    "\n",
    "    elif len(nb_positions) == 1:\n",
    "        add_pos  = np.random.choice(available_positions, 2, replace=False)\n",
    "        nb_positions = np.append(nb_positions, add_pos)\n",
    "\n",
    "    elif len(nb_positions) > 15:\n",
    "        print(\"Too many positions, either contaminated or sequencing error\")\n",
    "        #nb_positions = np.random.choice(range_positions, 3, replace=False)\n",
    "\n",
    "    bases_df, qual_df = get_bases_from_pileup(bam_file, ref_name, add_neighbouring_positions(nb_positions, 2, len(template)))\n",
    "    bases_df, qual_df = calculate_mean_quality_for_reads(bases_df, qual_df, nb_positions, 2)\n",
    "    softmax_df = get_softmax_count_df(bases_df, qual_df, nb_positions)\n",
    "    print(softmax_df)\n",
    "    variant_df = pd.DataFrame(call_potential_populations(softmax_df, template)).sort_values(by=\"Probability\", ascending=False)\n",
    "\n",
    "    # Take top variant\n",
    "    variants[\"Variant\"] = variant_df[\"Variant\"].iloc[0]\n",
    "    variants[\"Position\"] = nb_positions\n",
    "    variants[\"Alignment Probability\"] = variant_df[\"Probability\"].iloc[0]\n",
    "    variants[\"Alignment Count\"] = alignment_count\n",
    "\n",
    "    return variants\n",
    "\n",
    "\n",
    "def get_variant_df_soft(demultiplex_folder: Path, ref_seq : Path, ref_name : str, barcode_dicts : dict = None, merge = True, min_depth= 5, padding=50):\n",
    "\n",
    "\n",
    "    if barcode_dicts is None:\n",
    "        barcode_dicts = get_barcode_dict(demultiplex_folder)\n",
    "    \n",
    "    variant_template_df = analyser.template_df(barcode_dicts, rowwise=False)\n",
    "\n",
    "    variants = {\"RBC\": [], \"FBC\": [], \"Position\": [], \"Variant\": [], \"Alignment Probability\": [], \"Alignment Count\": []}\n",
    "\n",
    "    template = analyser.get_template_sequence(ref_seq) # Reference sequence\n",
    "\n",
    "    summary = analyser.read_summary_file(demultiplex_folder)\n",
    "    n_counts = summary.groupby([\"RBC\",\"FBC\"])[\"FBC\"].value_counts().reset_index() \n",
    "\n",
    "\n",
    "\n",
    "    for barcode_id, barcode_dict in barcode_dicts.items():\n",
    "\n",
    "        rbc = os.path.basename(barcode_id)\n",
    "\n",
    "        for front_barcode in barcode_dict:\n",
    "\n",
    "            fbc = os.path.basename(front_barcode)\n",
    "            print(\"Processing\", rbc, fbc)\n",
    "            count = n_counts[(n_counts[\"RBC\"] == rbc) & (n_counts[\"FBC\"] == fbc)][\"count\"].values[0]\n",
    "\n",
    "            # # If alignment file exist continue\n",
    "            if not os.path.exists(os.path.join(front_barcode, \"alignment_minimap.bam\")):\n",
    "                print(f\"Alignment file in {front_barcode} does not exist, running alignment and indexing\")\n",
    "                analyser.run_alignment_and_indexing(ref_seq, front_barcode)\n",
    "            \n",
    "            else: \n",
    "                print(\"Alignment file already exists, skipping alignment and indexing\")\n",
    "\n",
    "\n",
    "            bam_file = front_barcode / \"alignment_minimap.bam\"\n",
    "\n",
    "\n",
    "            if not bam_file.exists() or count < min_depth:\n",
    "                print(f\"{bam_file} does not exist.\")\n",
    "                variants[\"RBC\"].append(rbc)\n",
    "                variants[\"FBC\"].append(fbc)\n",
    "                variants[\"Position\"].append([\"NA\"])\n",
    "                variants[\"Variant\"].append([\"NA\"])\n",
    "                variants[\"Alignment Count\"].append([\"NA\"])\n",
    "                variants[\"Alignment Probability\"].append([\"NA\"])\n",
    "                print(f\"Skipping Variant: {fbc}/{rbc}\")\n",
    "                continue\n",
    "\n",
    "            # try:\n",
    "            if padding == 0:\n",
    "                print(\"Padding is 0. Implementing soft alignment\")\n",
    "            \n",
    "            else: \n",
    "                nn_variants = get_variant_soft(bam_file, ref_seq, ref_name, padding = padding)\n",
    "                print(nn_variants)\n",
    "\n",
    "            if nn_variants[\"Variant\"] is None:\n",
    "                print(\"Empty variant list\")\n",
    "                variants[\"RBC\"].append(rbc)\n",
    "                variants[\"FBC\"].append(fbc)\n",
    "                variants[\"Position\"].append([\"NA\"])\n",
    "                variants[\"Variant\"].append([\"NA\"])\n",
    "                variants[\"Alignment Count\"].append([\"NA\"])\n",
    "                variants[\"Alignment Probability\"].append([\"NA\"])\n",
    "                print(f\"Skipping Variant: {fbc}/{rbc}\")\n",
    "                continue\n",
    "\n",
    "            \n",
    "            variants[\"RBC\"].append(rbc)\n",
    "            variants[\"FBC\"].append(fbc)\n",
    "            variants[\"Position\"].append(nn_variants[\"Position\"])\n",
    "            variants[\"Variant\"].append(nn_variants[\"Variant\"])\n",
    "            #variants[\"Reads\"].append(count)\n",
    "            # Check if Alignment count is a number\n",
    "            if isinstance(nn_variants[\"Alignment Count\"], int) & (isinstance(nn_variants[\"Alignment Probability\"], float) or nn_variants[\"Alignment Probability\"] == \"-\"):\n",
    "                variants[\"Alignment Count\"].append(nn_variants[\"Alignment Count\"])\n",
    "                variants[\"Alignment Probability\"].append(nn_variants[\"Alignment Probability\"])\n",
    "\n",
    "\n",
    "            else:\n",
    "                print(f\"Skipping {rbc}/{fbc} due incomplete data\")\n",
    "                variants[\"Alignment Count\"].append(\"NA\")\n",
    "                variants[\"Alignment Probability\"].append(\"NA\")\n",
    "\n",
    "            print(f\"Variant: {fbc}/{rbc} {nn_variants['Alignment Count']} {nn_variants['Alignment Probability']}\")\n",
    "        \n",
    "        # except Exception as e:\n",
    "        #     # Append 'NA' in case of an exception\n",
    "        #     print(f\"Error processing {rbc}/{fbc}: {e}\")\n",
    "        #     variants[\"RBC\"].append(rbc)\n",
    "        #     variants[\"FBC\"].append(fbc)\n",
    "        #     variants[\"Position\"].append(\"NA\")\n",
    "        #     variants[\"Variant\"].append(\"NA\")\n",
    "        #     variants[\"Alignment Count\"].append(\"NA\")\n",
    "        #     variants[\"Alignment Frequency\"].append(\"NA\")\n",
    "\n",
    "\n",
    "\n",
    "    if merge:\n",
    "        variant_df = analyser.rename_barcode(pd.DataFrame(variants).merge(n_counts, on=[\"RBC\",\"FBC\"] , how=\"left\"))\n",
    "        variant_df[\"Variant\"] = variant_df[\"Variant\"].apply(analyser.format_variant_list)\n",
    "        variant_df[\"Variant\"] = variant_df[\"Variant\"].apply(lambda x: analyser.adjust_variant(x, padding))\n",
    "\n",
    "        return variant_df.merge(variant_template_df, on=[\"Plate\", \"Well\"], how=\"right\")\n",
    "    else:\n",
    "        return variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Plate</th>\n",
       "      <th>Well</th>\n",
       "      <th>Position</th>\n",
       "      <th>Variant</th>\n",
       "      <th>Alignment Count</th>\n",
       "      <th>Alignment Frequency</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>A5</td>\n",
       "      <td>[NA]</td>\n",
       "      <td>NA</td>\n",
       "      <td>[NA]</td>\n",
       "      <td>[NA]</td>\n",
       "      <td>135.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1</td>\n",
       "      <td>A8</td>\n",
       "      <td>-</td>\n",
       "      <td>NA</td>\n",
       "      <td>2</td>\n",
       "      <td>-</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1</td>\n",
       "      <td>A9</td>\n",
       "      <td>-</td>\n",
       "      <td>NA</td>\n",
       "      <td>2</td>\n",
       "      <td>-</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>1</td>\n",
       "      <td>B9</td>\n",
       "      <td>-</td>\n",
       "      <td>NA</td>\n",
       "      <td>3</td>\n",
       "      <td>-</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>1</td>\n",
       "      <td>B12</td>\n",
       "      <td>-</td>\n",
       "      <td>NA</td>\n",
       "      <td>2</td>\n",
       "      <td>-</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>1</td>\n",
       "      <td>C10</td>\n",
       "      <td>-</td>\n",
       "      <td>NA</td>\n",
       "      <td>4</td>\n",
       "      <td>-</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>1</td>\n",
       "      <td>C12</td>\n",
       "      <td>-</td>\n",
       "      <td>NA</td>\n",
       "      <td>7</td>\n",
       "      <td>-</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>1</td>\n",
       "      <td>E11</td>\n",
       "      <td>-</td>\n",
       "      <td>NA</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>1</td>\n",
       "      <td>F11</td>\n",
       "      <td>-</td>\n",
       "      <td>NA</td>\n",
       "      <td>2</td>\n",
       "      <td>-</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>1</td>\n",
       "      <td>F12</td>\n",
       "      <td>-</td>\n",
       "      <td>NA</td>\n",
       "      <td>8</td>\n",
       "      <td>-</td>\n",
       "      <td>8.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>1</td>\n",
       "      <td>G7</td>\n",
       "      <td>-</td>\n",
       "      <td>NA</td>\n",
       "      <td>8</td>\n",
       "      <td>-</td>\n",
       "      <td>8.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>1</td>\n",
       "      <td>G9</td>\n",
       "      <td>-</td>\n",
       "      <td>NA</td>\n",
       "      <td>3</td>\n",
       "      <td>-</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>2</td>\n",
       "      <td>A12</td>\n",
       "      <td>-</td>\n",
       "      <td>NA</td>\n",
       "      <td>5</td>\n",
       "      <td>-</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>115</th>\n",
       "      <td>2</td>\n",
       "      <td>B6</td>\n",
       "      <td>-</td>\n",
       "      <td>NA</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117</th>\n",
       "      <td>2</td>\n",
       "      <td>B8</td>\n",
       "      <td>[NA]</td>\n",
       "      <td>NA</td>\n",
       "      <td>[NA]</td>\n",
       "      <td>[NA]</td>\n",
       "      <td>12.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>2</td>\n",
       "      <td>B10</td>\n",
       "      <td>[NA]</td>\n",
       "      <td>NA</td>\n",
       "      <td>[NA]</td>\n",
       "      <td>[NA]</td>\n",
       "      <td>19.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>2</td>\n",
       "      <td>C7</td>\n",
       "      <td>-</td>\n",
       "      <td>NA</td>\n",
       "      <td>4</td>\n",
       "      <td>-</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130</th>\n",
       "      <td>2</td>\n",
       "      <td>C9</td>\n",
       "      <td>-</td>\n",
       "      <td>NA</td>\n",
       "      <td>4</td>\n",
       "      <td>-</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136</th>\n",
       "      <td>2</td>\n",
       "      <td>D3</td>\n",
       "      <td>-</td>\n",
       "      <td>NA</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>137</th>\n",
       "      <td>2</td>\n",
       "      <td>D4</td>\n",
       "      <td>-</td>\n",
       "      <td>NA</td>\n",
       "      <td>4</td>\n",
       "      <td>-</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161</th>\n",
       "      <td>2</td>\n",
       "      <td>F4</td>\n",
       "      <td>-</td>\n",
       "      <td>NA</td>\n",
       "      <td>6</td>\n",
       "      <td>-</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>180</th>\n",
       "      <td>2</td>\n",
       "      <td>G11</td>\n",
       "      <td>-</td>\n",
       "      <td>NA</td>\n",
       "      <td>5</td>\n",
       "      <td>-</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>194</th>\n",
       "      <td>3</td>\n",
       "      <td>A1</td>\n",
       "      <td>-</td>\n",
       "      <td>NA</td>\n",
       "      <td>2</td>\n",
       "      <td>-</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205</th>\n",
       "      <td>3</td>\n",
       "      <td>A12</td>\n",
       "      <td>-</td>\n",
       "      <td>NA</td>\n",
       "      <td>5</td>\n",
       "      <td>-</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>218</th>\n",
       "      <td>3</td>\n",
       "      <td>C1</td>\n",
       "      <td>[NA]</td>\n",
       "      <td>NA</td>\n",
       "      <td>[NA]</td>\n",
       "      <td>[NA]</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>219</th>\n",
       "      <td>3</td>\n",
       "      <td>C2</td>\n",
       "      <td>-</td>\n",
       "      <td>NA</td>\n",
       "      <td>7</td>\n",
       "      <td>-</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>267</th>\n",
       "      <td>3</td>\n",
       "      <td>G2</td>\n",
       "      <td>-</td>\n",
       "      <td>NA</td>\n",
       "      <td>4</td>\n",
       "      <td>-</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>273</th>\n",
       "      <td>3</td>\n",
       "      <td>G8</td>\n",
       "      <td>-</td>\n",
       "      <td>NA</td>\n",
       "      <td>7</td>\n",
       "      <td>-</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>278</th>\n",
       "      <td>3</td>\n",
       "      <td>H1</td>\n",
       "      <td>-</td>\n",
       "      <td>NA</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Plate Well Position Variant Alignment Count Alignment Frequency  count\n",
       "4        1   A5     [NA]      NA            [NA]                [NA]  135.0\n",
       "7        1   A8        -      NA               2                   -    2.0\n",
       "8        1   A9        -      NA               2                   -    2.0\n",
       "20       1   B9        -      NA               3                   -    3.0\n",
       "23       1  B12        -      NA               2                   -    2.0\n",
       "33       1  C10        -      NA               4                   -    4.0\n",
       "35       1  C12        -      NA               7                   -    7.0\n",
       "58       1  E11        -      NA               1                   -    1.0\n",
       "70       1  F11        -      NA               2                   -    2.0\n",
       "71       1  F12        -      NA               8                   -    8.0\n",
       "78       1   G7        -      NA               8                   -    8.0\n",
       "80       1   G9        -      NA               3                   -    3.0\n",
       "109      2  A12        -      NA               5                   -    5.0\n",
       "115      2   B6        -      NA               1                   -    1.0\n",
       "117      2   B8     [NA]      NA            [NA]                [NA]   12.0\n",
       "119      2  B10     [NA]      NA            [NA]                [NA]   19.0\n",
       "128      2   C7        -      NA               4                   -    4.0\n",
       "130      2   C9        -      NA               4                   -    4.0\n",
       "136      2   D3        -      NA               1                   -    1.0\n",
       "137      2   D4        -      NA               4                   -    4.0\n",
       "161      2   F4        -      NA               6                   -    6.0\n",
       "180      2  G11        -      NA               5                   -    5.0\n",
       "194      3   A1        -      NA               2                   -    2.0\n",
       "205      3  A12        -      NA               5                   -    5.0\n",
       "218      3   C1     [NA]      NA            [NA]                [NA]   15.0\n",
       "219      3   C2        -      NA               7                   -    7.0\n",
       "267      3   G2        -      NA               4                   -    4.0\n",
       "273      3   G8        -      NA               7                   -    7.0\n",
       "278      3   H1        -      NA               1                   -    1.0"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get alignment which were difficult to align \n",
    "variant_df = pd.read_pickle('/home/emre/github_repo/MinION/results/2_hetcpiii_minion_errorprone/local/variants_SW_40k.pkl')\n",
    "variant_df_guppy = pd.read_pickle('/home/emre/github_repo/MinION/results/2_hetcpiii_minion_errorprone/local/variants_SW_BF_40k.pkl')\n",
    "variant_df[variant_df[\"Variant\"] == \"NA\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "        80   146  647\n",
      "A  0.914990  0.0  0.0\n",
      "C  0.000000  0.0  1.0\n",
      "T  0.000000  0.0  0.0\n",
      "G  0.029121  1.0  0.0\n",
      "-  0.055889  0.0  0.0\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'Variant': 'T80A',\n",
       " 'Position': array([ 80, 146, 647]),\n",
       " 'Alignment Probability': 0.9149902666562933,\n",
       " 'Alignment Count': 34}"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bam_file = Path(\"/home/emre/minION_results/MinION_RBC_0902723_sup/Demultiplex_cpp_70_40k_reads/RB02/NB04/alignment_minimap.bam\")\n",
    "template_seq = Path(\"/home/emre/github_repo/MinION/minION/refseq/hetcpiii_padded.fasta\")\n",
    "ref_name = \"HetCPIII\"\n",
    "\n",
    "\n",
    "\n",
    "get_variant_soft(bam_file, template_seq, ref_name, padding = 50)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing RB03 NB87\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "       84        164       178       211\n",
      "A  0.00000  0.412497  0.627552  0.413806\n",
      "C  0.00000  0.000000  0.000000  0.017949\n",
      "T  0.00000  0.000000  0.000000  0.000000\n",
      "G  0.58565  0.550829  0.372448  0.546233\n",
      "-  0.41435  0.036674  0.000000  0.022012\n",
      "{'Variant': 'A164G_A211G', 'Position': [84, 164, 178, 211], 'Alignment Probability': 0.11058156794895055, 'Alignment Count': 46}\n",
      "Variant: NB87/RB03 46 0.11058156794895055\n",
      "Processing RB03 NB03\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        87        116\n",
      "A  0.000000  0.010879\n",
      "C  0.016721  0.937923\n",
      "T  0.916892  0.040256\n",
      "G  0.040408  0.000000\n",
      "-  0.025979  0.010943\n",
      "{'Variant': 'C87T_T116C', 'Position': [87, 116], 'Alignment Probability': 0.8599735671700774, 'Alignment Count': 94}\n",
      "Variant: NB03/RB03 94 0.8599735671700774\n",
      "Processing RB03 NB20\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        191  520       630\n",
      "A  0.000000  NaN  0.000000\n",
      "C  0.907562  NaN  0.045196\n",
      "T  0.080157  NaN  0.000000\n",
      "G  0.000000  NaN  0.954804\n",
      "-  0.012281  NaN  0.000000\n",
      "{'Variant': 'T191C_T520A', 'Position': array([191, 520, 630]), 'Alignment Probability': nan, 'Alignment Count': 56}\n",
      "Variant: NB20/RB03 56 nan\n",
      "Processing RB03 NB48\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        74        143       176       185       202\n",
      "A  0.220049  0.000000  0.245783  0.005951  0.217440\n",
      "C  0.008568  0.221219  0.009359  0.709545  0.009711\n",
      "T  0.000000  0.778781  0.000000  0.259366  0.740451\n",
      "G  0.714494  0.000000  0.735224  0.000000  0.007839\n",
      "-  0.056889  0.000000  0.009634  0.025138  0.024559\n",
      "{'Variant': 'A74G_C143T_A176G_T185C_A202T', 'Position': [74, 143, 176, 185, 202], 'Alignment Probability': 0.21493622012910957, 'Alignment Count': 105}\n",
      "Variant: NB48/RB03 105 0.21493622012910957\n",
      "Processing RB03 NB70\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        226       84        168\n",
      "A  0.953943  0.000000  0.000000\n",
      "C  0.006863  0.004629  0.005100\n",
      "T  0.000000  0.005662  0.983699\n",
      "G  0.036388  0.953247  0.000000\n",
      "-  0.002806  0.036462  0.011201\n",
      "{'Variant': 'G226A', 'Position': array([226,  84, 168]), 'Alignment Probability': 0.8945201987628429, 'Alignment Count': 147}\n",
      "Variant: NB70/RB03 147 0.8945201987628429\n",
      "Processing RB03 NB04\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        74        101       192\n",
      "A  0.014025  0.062013  0.967026\n",
      "C  0.000000  0.011931  0.010750\n",
      "T  0.000000  0.008524  0.011623\n",
      "G  0.908996  0.904941  0.010601\n",
      "-  0.076979  0.012590  0.000000\n",
      "{'Variant': 'A74G_A101G_C192A', 'Position': [74, 101, 192], 'Alignment Probability': 0.7954643823608245, 'Alignment Count': 82}\n",
      "Variant: NB04/RB03 82 0.7954643823608245\n",
      "Processing RB03 NB37\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "   236  368  345\n",
      "A  0.0  0.0  0.0\n",
      "C  0.0  1.0  0.0\n",
      "T  1.0  0.0  0.0\n",
      "G  0.0  0.0  1.0\n",
      "-  0.0  0.0  0.0\n",
      "{'Variant': 'G236T', 'Position': array([236, 368, 345]), 'Alignment Probability': 1.0, 'Alignment Count': 35}\n",
      "Variant: NB37/RB03 35 1.0\n",
      "Processing RB03 NB88\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        105       112  165       211       229\n",
      "A  0.033734  0.000000  0.0  0.000000  0.000000\n",
      "C  0.000000  0.000000  0.0  0.000000  0.000000\n",
      "T  0.966266  0.000000  0.0  0.000000  0.000000\n",
      "G  0.000000  0.960395  1.0  0.912974  0.793585\n",
      "-  0.000000  0.039605  0.0  0.087026  0.206415\n",
      "{'Variant': 'A105T_A112G_A165G_A211G_A229G', 'Position': [105, 112, 165, 211, 229], 'Alignment Probability': 0.6723541383690932, 'Alignment Count': 23}\n",
      "Variant: NB88/RB03 23 0.6723541383690932\n",
      "Processing RB03 NB14\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        79        135       217\n",
      "A  0.005559  0.020454  0.993453\n",
      "C  0.000000  0.000000  0.000000\n",
      "T  0.000000  0.000000  0.006547\n",
      "G  0.984742  0.979546  0.000000\n",
      "-  0.009699  0.000000  0.000000\n",
      "{'Variant': 'A79G_A135G_T217A', 'Position': [79, 135, 217], 'Alignment Probability': 0.9582839254506399, 'Alignment Count': 84}\n",
      "Variant: NB14/RB03 84 0.9582839254506399\n",
      "Processing RB03 NB82\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "   108       344       487\n",
      "A  1.0  0.000000  0.000000\n",
      "C  0.0  0.000000  0.000000\n",
      "T  0.0  0.000000  0.986669\n",
      "G  0.0  0.953279  0.000000\n",
      "-  0.0  0.046721  0.013331\n",
      "{'Variant': '#PARENT#', 'Position': array([108, 344, 487]), 'Alignment Probability': 0.9405708003381947, 'Alignment Count': 61}\n",
      "Variant: NB82/RB03 61 0.9405708003381947\n",
      "Processing RB03 NB54\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "   330      243       158\n",
      "A  0.0  0.01064  0.009112\n",
      "C  0.0  0.00000  0.000000\n",
      "T  0.0  0.98936  0.000000\n",
      "G  1.0  0.00000  0.990888\n",
      "-  0.0  0.00000  0.000000\n",
      "{'Variant': '#PARENT#', 'Position': array([330, 243, 158]), 'Alignment Probability': 0.9803449412104964, 'Alignment Count': 83}\n",
      "Variant: NB54/RB03 83 0.9803449412104964\n",
      "Processing RB03 NB44\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        83        94        209\n",
      "A  0.014733  0.027952  0.985208\n",
      "C  0.850198  0.951463  0.000000\n",
      "T  0.056275  0.009651  0.000000\n",
      "G  0.013687  0.000000  0.000000\n",
      "-  0.065106  0.010934  0.014792\n",
      "{'Variant': 'T83C_A94C_T209A', 'Position': [83, 94, 209], 'Alignment Probability': 0.7969662899678907, 'Alignment Count': 68}\n",
      "Variant: NB44/RB03 68 0.7969662899678907\n",
      "Processing RB03 NB57\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "       339  585  176\n",
      "A  0.97993  NaN  1.0\n",
      "C  0.00000  NaN  0.0\n",
      "T  0.02007  NaN  0.0\n",
      "G  0.00000  NaN  0.0\n",
      "-  0.00000  NaN  0.0\n",
      "{'Variant': 'G585A', 'Position': array([339, 585, 176]), 'Alignment Probability': nan, 'Alignment Count': 84}\n",
      "Variant: NB57/RB03 84 nan\n",
      "Processing RB03 NB28\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "       222  73        385\n",
      "A  0.00000  NaN  0.977837\n",
      "C  0.00000  NaN  0.011448\n",
      "T  0.98988  NaN  0.004370\n",
      "G  0.00000  NaN  0.000000\n",
      "-  0.01012  NaN  0.006346\n",
      "{'Variant': 'G73A', 'Position': array([222,  73, 385]), 'Alignment Probability': nan, 'Alignment Count': 151}\n",
      "Variant: NB28/RB03 151 nan\n",
      "Processing RB03 NB40\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        153       105       589\n",
      "A  0.829772  0.976032  0.000000\n",
      "C  0.008670  0.000000  0.014196\n",
      "T  0.000000  0.000000  0.949397\n",
      "G  0.034866  0.023968  0.008302\n",
      "-  0.126693  0.000000  0.028105\n",
      "{'Variant': 'G153A', 'Position': array([153, 105, 589]), 'Alignment Probability': 0.7689016423433737, 'Alignment Count': 114}\n",
      "Variant: NB40/RB03 114 0.7689016423433737\n",
      "Processing RB03 NB81\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        171  143       572\n",
      "A  0.000000  0.0  0.000000\n",
      "C  0.735804  1.0  0.011641\n",
      "T  0.218442  0.0  0.013519\n",
      "G  0.000000  0.0  0.974840\n",
      "-  0.045754  0.0  0.000000\n",
      "{'Variant': 'T171C', 'Position': array([171, 143, 572]), 'Alignment Probability': 0.7172912729113928, 'Alignment Count': 64}\n",
      "Variant: NB81/RB03 64 0.7172912729113928\n",
      "Processing RB03 NB13\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        90        157       160       176\n",
      "A  0.000000  0.659734  0.671543  0.697854\n",
      "C  0.021120  0.000000  0.000000  0.000000\n",
      "T  0.347104  0.329865  0.000000  0.000000\n",
      "G  0.609668  0.000000  0.316840  0.302146\n",
      "-  0.022108  0.010401  0.011616  0.000000\n",
      "{'Variant': '#PARENT#', 'Position': [90, 157, 160, 176], 'Alignment Probability': 0.1884952830960649, 'Alignment Count': 73}\n",
      "Variant: NB13/RB03 73 0.1884952830960649\n",
      "Processing RB03 NB53\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "   66        102       235\n",
      "A  NaN  0.000000  0.918660\n",
      "C  NaN  0.008340  0.027038\n",
      "T  NaN  0.984712  0.000000\n",
      "G  NaN  0.000000  0.027560\n",
      "-  NaN  0.006948  0.026742\n",
      "{'Variant': 'C102T_G235A', 'Position': [66, 102, 235], 'Alignment Probability': nan, 'Alignment Count': 107}\n",
      "Variant: NB53/RB03 107 nan\n",
      "Processing RB03 NB80\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "   495  254  475\n",
      "A  1.0  0.0  0.0\n",
      "C  0.0  0.0  0.0\n",
      "T  0.0  1.0  1.0\n",
      "G  0.0  0.0  0.0\n",
      "-  0.0  0.0  0.0\n",
      "{'Variant': '#PARENT#', 'Position': array([495, 254, 475]), 'Alignment Probability': 1.0, 'Alignment Count': 7}\n",
      "Variant: NB80/RB03 7 1.0\n",
      "Processing RB03 NB47\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "   253       454       97 \n",
      "A  1.0  0.027914  0.954672\n",
      "C  0.0  0.941899  0.012236\n",
      "T  0.0  0.000000  0.000000\n",
      "G  0.0  0.000000  0.033092\n",
      "-  0.0  0.030187  0.000000\n",
      "{'Variant': '#PARENT#', 'Position': array([253, 454,  97]), 'Alignment Probability': 0.8992048335935534, 'Alignment Count': 54}\n",
      "Variant: NB47/RB03 54 0.8992048335935534\n",
      "Processing RB03 NB84\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "   158       619       372\n",
      "A  0.0  0.000000  0.000000\n",
      "C  0.0  0.981759  0.017564\n",
      "T  0.0  0.000000  0.000000\n",
      "G  1.0  0.000000  0.982436\n",
      "-  0.0  0.018241  0.000000\n",
      "{'Variant': '#PARENT#', 'Position': array([158, 619, 372]), 'Alignment Probability': 0.9645145373806984, 'Alignment Count': 42}\n",
      "Variant: NB84/RB03 42 0.9645145373806984\n",
      "Processing RB03 NB72\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        197       414       396\n",
      "A  0.000000  0.006789  0.000000\n",
      "C  0.000000  0.985122  0.006324\n",
      "T  0.000000  0.008089  0.000000\n",
      "G  0.992891  0.000000  0.993676\n",
      "-  0.007109  0.000000  0.000000\n",
      "{'Variant': 'A197G', 'Position': array([197, 414, 396]), 'Alignment Probability': 0.9719327699494851, 'Alignment Count': 117}\n",
      "Variant: NB72/RB03 117 0.9719327699494851\n",
      "Processing RB03 NB10\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "   151  542  570\n",
      "A  1.0  0.0  0.0\n",
      "C  0.0  0.0  0.0\n",
      "T  0.0  0.0  1.0\n",
      "G  0.0  1.0  0.0\n",
      "-  0.0  0.0  0.0\n",
      "{'Variant': 'G151A', 'Position': array([151, 542, 570]), 'Alignment Probability': 1.0, 'Alignment Count': 83}\n",
      "Variant: NB10/RB03 83 1.0\n",
      "Processing RB03 NB43\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "         77        89\n",
      "A  0.919360  0.000000\n",
      "C  0.013200  0.898419\n",
      "T  0.018542  0.066483\n",
      "G  0.030993  0.012926\n",
      "-  0.017905  0.022171\n",
      "{'Variant': 'T77A_T89C', 'Position': [77, 89], 'Alignment Probability': 0.8259710094479471, 'Alignment Count': 122}\n",
      "Variant: NB43/RB03 122 0.8259710094479471\n",
      "Processing RB03 NB58\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        108  402       542\n",
      "A  0.240224  0.0  0.017797\n",
      "C  0.000000  1.0  0.000000\n",
      "T  0.000000  0.0  0.000000\n",
      "G  0.744691  0.0  0.931158\n",
      "-  0.015085  0.0  0.051045\n",
      "{'Variant': 'A108G', 'Position': array([108, 402, 542]), 'Alignment Probability': 0.6934251032577231, 'Alignment Count': 58}\n",
      "Variant: NB58/RB03 58 0.6934251032577231\n",
      "Processing RB03 NB79\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "   105       176\n",
      "A  0.0  0.000000\n",
      "C  0.0  0.000000\n",
      "T  0.0  0.000000\n",
      "G  1.0  0.985857\n",
      "-  0.0  0.014143\n",
      "{'Variant': 'A105G_A176G', 'Position': [105, 176], 'Alignment Probability': 0.9858569860729203, 'Alignment Count': 64}\n",
      "Variant: NB79/RB03 64 0.9858569860729203\n",
      "Processing RB03 NB36\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        77        147       220\n",
      "A  0.959132  0.972481  0.000000\n",
      "C  0.020534  0.000000  0.000000\n",
      "T  0.006244  0.014971  0.000000\n",
      "G  0.008273  0.007451  0.995454\n",
      "-  0.005817  0.005097  0.004546\n",
      "{'Variant': 'T77A_T147A_A220G', 'Position': [77, 147, 220], 'Alignment Probability': 0.9284968818031699, 'Alignment Count': 120}\n",
      "Variant: NB36/RB03 120 0.9284968818031699\n",
      "Processing RB03 NB30\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        116       191       75 \n",
      "A  0.000000  0.000000  0.021421\n",
      "C  0.351285  0.000000  0.000000\n",
      "T  0.648715  0.984345  0.932130\n",
      "G  0.000000  0.015655  0.046448\n",
      "-  0.000000  0.000000  0.000000\n",
      "{'Variant': '#PARENT#', 'Position': array([116, 191,  75]), 'Alignment Probability': 0.5952203008623982, 'Alignment Count': 40}\n",
      "Variant: NB30/RB03 40 0.5952203008623982\n",
      "Processing RB03 NB60\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        122       157\n",
      "A  0.000000  0.041922\n",
      "C  0.956032  0.000000\n",
      "T  0.031387  0.000000\n",
      "G  0.012581  0.958078\n",
      "-  0.000000  0.000000\n",
      "{'Variant': 'T122C_A157G', 'Position': [122, 157], 'Alignment Probability': 0.9159530324808215, 'Alignment Count': 136}\n",
      "Variant: NB60/RB03 136 0.9159530324808215\n",
      "Processing RB03 NB45\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "   568  523  358\n",
      "A  0.0  0.0  0.0\n",
      "C  1.0  1.0  0.0\n",
      "T  0.0  0.0  0.0\n",
      "G  0.0  0.0  1.0\n",
      "-  0.0  0.0  0.0\n",
      "{'Variant': '#PARENT#', 'Position': array([568, 523, 358]), 'Alignment Probability': 1.0, 'Alignment Count': 37}\n",
      "Variant: NB45/RB03 37 1.0\n",
      "Processing RB03 NB93\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        633       296       268\n",
      "A  0.000000  0.000000  0.000000\n",
      "C  0.010895  0.011075  0.011881\n",
      "T  0.978268  0.988925  0.966540\n",
      "G  0.010838  0.000000  0.009729\n",
      "-  0.000000  0.000000  0.011850\n",
      "{'Variant': '#PARENT#', 'Position': array([633, 296, 268]), 'Alignment Probability': 0.9350639914979719, 'Alignment Count': 86}\n",
      "Variant: NB93/RB03 86 0.9350639914979719\n",
      "Processing RB03 NB31\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        191  183       198\n",
      "A  0.000000  0.0  0.036473\n",
      "C  0.887171  1.0  0.000000\n",
      "T  0.112829  0.0  0.000000\n",
      "G  0.000000  0.0  0.938726\n",
      "-  0.000000  0.0  0.024801\n",
      "{'Variant': 'T191C', 'Position': array([191, 183, 198]), 'Alignment Probability': 0.8328108650597349, 'Alignment Count': 25}\n",
      "Variant: NB31/RB03 25 0.8328108650597349\n",
      "Processing RB03 NB75\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        96        106\n",
      "A  0.014772  0.000000\n",
      "C  0.000000  0.984522\n",
      "T  0.926384  0.015478\n",
      "G  0.015762  0.000000\n",
      "-  0.043082  0.000000\n",
      "{'Variant': 'A96T_T106C', 'Position': [96, 106], 'Alignment Probability': 0.9120457676502498, 'Alignment Count': 58}\n",
      "Variant: NB75/RB03 58 0.9120457676502498\n",
      "Processing RB03 NB21\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        51        54   106\n",
      "A  0.970065  0.986061  0.0\n",
      "C  0.009194  0.000000  0.0\n",
      "T  0.000000  0.000000  1.0\n",
      "G  0.000000  0.005706  0.0\n",
      "-  0.020741  0.008234  0.0\n",
      "{'Variant': '#PARENT#', 'Position': array([ 51,  54, 106]), 'Alignment Probability': 0.956542988606225, 'Alignment Count': 107}\n",
      "Variant: NB21/RB03 107 0.956542988606225\n",
      "Processing RB03 NB86\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        89       110\n",
      "A  0.000000  0.00000\n",
      "C  0.886996  0.00000\n",
      "T  0.058503  0.97721\n",
      "G  0.026213  0.02279\n",
      "-  0.028288  0.00000\n",
      "{'Variant': 'T89C_G110T', 'Position': [89, 110], 'Alignment Probability': 0.8667815585877442, 'Alignment Count': 35}\n",
      "Variant: NB86/RB03 35 0.8667815585877442\n",
      "Processing RB03 NB07\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        197       233\n",
      "A  0.034873  0.000000\n",
      "C  0.000000  0.901556\n",
      "T  0.000000  0.073270\n",
      "G  0.965127  0.000000\n",
      "-  0.000000  0.025174\n",
      "{'Variant': 'A197G_T233C', 'Position': [197, 233], 'Alignment Probability': 0.8701158930358783, 'Alignment Count': 71}\n",
      "Variant: NB07/RB03 71 0.8701158930358783\n",
      "Processing RB03 NB34\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "   554  417       185\n",
      "A  1.0  0.0  0.000000\n",
      "C  0.0  0.0  0.000000\n",
      "T  0.0  1.0  0.990459\n",
      "G  0.0  0.0  0.000000\n",
      "-  0.0  0.0  0.009541\n",
      "{'Variant': '#PARENT#', 'Position': array([554, 417, 185]), 'Alignment Probability': 0.9904592096691877, 'Alignment Count': 77}\n",
      "Variant: NB34/RB03 77 0.9904592096691877\n",
      "Processing RB03 NB62\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        142       107       408\n",
      "A  0.000000  0.000000  0.011551\n",
      "C  0.991583  0.976278  0.985004\n",
      "T  0.008417  0.017447  0.000000\n",
      "G  0.000000  0.000000  0.000000\n",
      "-  0.000000  0.006275  0.003445\n",
      "{'Variant': 'T142C', 'Position': array([142, 107, 408]), 'Alignment Probability': 0.9535430489508924, 'Alignment Count': 148}\n",
      "Variant: NB62/RB03 148 0.9535430489508924\n",
      "Processing RB03 NB23\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        187  225\n",
      "A  0.019163  NaN\n",
      "C  0.000000  NaN\n",
      "T  0.000000  NaN\n",
      "G  0.960139  NaN\n",
      "-  0.020698  NaN\n",
      "{'Variant': 'A187G_T225A', 'Position': [187, 225], 'Alignment Probability': nan, 'Alignment Count': 92}\n",
      "Variant: NB23/RB03 92 nan\n",
      "Processing RB03 NB92\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        225  56        126\n",
      "A  0.000000  0.0  0.000000\n",
      "C  0.939146  0.0  0.980041\n",
      "T  0.000000  1.0  0.019959\n",
      "G  0.000000  0.0  0.000000\n",
      "-  0.060854  0.0  0.000000\n",
      "{'Variant': 'T225C', 'Position': array([225,  56, 126]), 'Alignment Probability': 0.9204010879254421, 'Alignment Count': 33}\n",
      "Variant: NB92/RB03 33 0.9204010879254421\n",
      "Processing RB03 NB73\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        97        106\n",
      "A  0.027350  0.000000\n",
      "C  0.000000  0.943177\n",
      "T  0.033529  0.056823\n",
      "G  0.669414  0.000000\n",
      "-  0.269707  0.000000\n",
      "{'Variant': 'A97G_T106C', 'Position': [97, 106], 'Alignment Probability': 0.6313755614590287, 'Alignment Count': 30}\n",
      "Variant: NB73/RB03 30 0.6313755614590287\n",
      "Processing RB03 NB67\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "       176       288       363\n",
      "A  0.97972  0.000000  0.000000\n",
      "C  0.00000  0.000000  0.000000\n",
      "T  0.00000  0.009123  0.974877\n",
      "G  0.02028  0.984481  0.025123\n",
      "-  0.00000  0.006396  0.000000\n",
      "{'Variant': '#PARENT#', 'Position': array([176, 288, 363]), 'Alignment Probability': 0.9402843896382779, 'Alignment Count': 99}\n",
      "Variant: NB67/RB03 99 0.9402843896382779\n",
      "Processing RB03 NB29\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "   203       285       210\n",
      "A  NaN  0.014217  0.000000\n",
      "C  NaN  0.007277  0.017086\n",
      "T  NaN  0.000000  0.000000\n",
      "G  NaN  0.978506  0.982914\n",
      "-  NaN  0.000000  0.000000\n",
      "{'Variant': '#PARENT#', 'Position': array([203, 285, 210]), 'Alignment Probability': nan, 'Alignment Count': 114}\n",
      "Variant: NB29/RB03 114 nan\n",
      "Processing RB03 NB33\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        207       268  147\n",
      "A  0.017731  0.000000  0.0\n",
      "C  0.006427  0.000000  0.0\n",
      "T  0.000000  0.989489  1.0\n",
      "G  0.975841  0.000000  0.0\n",
      "-  0.000000  0.010511  0.0\n",
      "{'Variant': '#PARENT#', 'Position': array([207, 268, 147]), 'Alignment Probability': 0.9655846614843899, 'Alignment Count': 114}\n",
      "Variant: NB33/RB03 114 0.9655846614843899\n",
      "Processing RB03 NB06\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        136       202\n",
      "A  0.040493  0.052497\n",
      "C  0.010419  0.005628\n",
      "T  0.000000  0.925027\n",
      "G  0.949087  0.009133\n",
      "-  0.000000  0.007715\n",
      "{'Variant': 'A136G_A202T', 'Position': [136, 202], 'Alignment Probability': 0.8779313365438444, 'Alignment Count': 91}\n",
      "Variant: NB06/RB03 91 0.8779313365438444\n",
      "Processing RB03 NB78\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        148       191       337\n",
      "A  0.008970  0.000000  0.000000\n",
      "C  0.944606  0.009590  0.054783\n",
      "T  0.029389  0.980792  0.000000\n",
      "G  0.009644  0.000000  0.920229\n",
      "-  0.007391  0.009617  0.024989\n",
      "{'Variant': 'T148C', 'Position': array([148, 191, 337]), 'Alignment Probability': 0.8525568090701154, 'Alignment Count': 102}\n",
      "Variant: NB78/RB03 102 0.8525568090701154\n",
      "Processing RB03 NB95\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        157       233       259\n",
      "A  0.042317  0.000000  0.013826\n",
      "C  0.000000  0.958873  0.960950\n",
      "T  0.000000  0.014607  0.011494\n",
      "G  0.957683  0.000000  0.000000\n",
      "-  0.000000  0.026520  0.013729\n",
      "{'Variant': 'A157G_T233C_T259C', 'Position': [157, 233, 259], 'Alignment Probability': 0.88243708393114, 'Alignment Count': 69}\n",
      "Variant: NB95/RB03 69 0.88243708393114\n",
      "Processing RB03 NB26\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        96        167       185       226\n",
      "A  0.287761  0.000000  0.761089  0.000000\n",
      "C  0.000000  0.106859  0.000000  0.000000\n",
      "T  0.000000  0.768043  0.000000  0.000000\n",
      "G  0.425734  0.125097  0.000000  0.580868\n",
      "-  0.286506  0.000000  0.238911  0.419132\n",
      "{'Variant': 'A96G_G167T_T185A', 'Position': [96, 167, 185, 226], 'Alignment Probability': 0.14455607448995075, 'Alignment Count': 7}\n",
      "Variant: NB26/RB03 7 0.14455607448995075\n",
      "Processing RB03 NB16\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        183       193\n",
      "A  0.000000  0.000000\n",
      "C  0.035607  0.033670\n",
      "T  0.964393  0.032701\n",
      "G  0.000000  0.914968\n",
      "-  0.000000  0.018661\n",
      "{'Variant': 'C183T_T193G', 'Position': [183, 193], 'Alignment Probability': 0.882388857943057, 'Alignment Count': 28}\n",
      "Variant: NB16/RB03 28 0.882388857943057\n",
      "Processing RB03 NB76\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        62        102       142       222\n",
      "A  0.824901  0.008029  0.000000  0.000000\n",
      "C  0.008201  0.006365  0.960226  0.980333\n",
      "T  0.016480  0.979960  0.022868  0.012856\n",
      "G  0.099557  0.000000  0.000000  0.000000\n",
      "-  0.050861  0.005646  0.016906  0.006811\n",
      "{'Variant': 'G62A_C102T_T142C_T222C', 'Position': [62, 102, 142, 222], 'Alignment Probability': 0.7609515896095644, 'Alignment Count': 125}\n",
      "Variant: NB76/RB03 125 0.7609515896095644\n",
      "Processing RB03 NB59\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        147       171      214       234\n",
      "A  0.000000  0.000000  0.00000  0.000000\n",
      "C  0.953439  0.977217  0.97908  0.712083\n",
      "T  0.046561  0.022783  0.00000  0.195433\n",
      "G  0.000000  0.000000  0.00000  0.000000\n",
      "-  0.000000  0.000000  0.02092  0.092484\n",
      "{'Variant': 'T147C_T171C_T214C_T234C', 'Position': [147, 171, 214, 234], 'Alignment Probability': 0.6495801692944537, 'Alignment Count': 40}\n",
      "Variant: NB59/RB03 40 0.6495801692944537\n",
      "Processing RB03 NB68\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        154       228\n",
      "A  0.008200  0.000000\n",
      "C  0.000000  0.013312\n",
      "T  0.000000  0.931161\n",
      "G  0.970257  0.050086\n",
      "-  0.021543  0.005442\n",
      "{'Variant': 'A154G_G228T', 'Position': [154, 228], 'Alignment Probability': 0.9034648364390424, 'Alignment Count': 113}\n",
      "Variant: NB68/RB03 113 0.9034648364390424\n",
      "Processing RB03 NB55\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        219       337     283\n",
      "A  0.000000  0.000000  0.0000\n",
      "C  0.000000  0.025705  0.9556\n",
      "T  0.045749  0.000000  0.0000\n",
      "G  0.954251  0.974295  0.0000\n",
      "-  0.000000  0.000000  0.0444\n",
      "{'Variant': '#PARENT#', 'Position': array([219, 337, 283]), 'Alignment Probability': 0.888442763500454, 'Alignment Count': 40}\n",
      "Variant: NB55/RB03 40 0.888442763500454\n",
      "Processing RB03 NB74\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "/home/emre/minION_results/MinION_RBC_0902723_sup/Demultiplex_cpp_70_40k_reads/RB03/NB74/alignment_minimap.bam does not exist.\n",
      "Skipping Variant: NB74/RB03\n",
      "Processing RB03 NB89\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "   64   655  142\n",
      "A  1.0  1.0  0.0\n",
      "C  0.0  0.0  0.0\n",
      "T  0.0  0.0  1.0\n",
      "G  0.0  0.0  0.0\n",
      "-  0.0  0.0  0.0\n",
      "{'Variant': '#PARENT#', 'Position': array([ 64, 655, 142]), 'Alignment Probability': 1.0, 'Alignment Count': 31}\n",
      "Variant: NB89/RB03 31 1.0\n",
      "Processing RB03 NB83\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        73        248       432\n",
      "A  0.015516  0.000000  0.991368\n",
      "C  0.000000  0.029120  0.000000\n",
      "T  0.000000  0.965403  0.000000\n",
      "G  0.984484  0.005477  0.000000\n",
      "-  0.000000  0.000000  0.008632\n",
      "{'Variant': '#PARENT#', 'Position': array([ 73, 248, 432]), 'Alignment Probability': 0.9422187836636094, 'Alignment Count': 118}\n",
      "Variant: NB83/RB03 118 0.9422187836636094\n",
      "Processing RB03 NB22\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        91        107       139  156\n",
      "A  0.000000  0.000000  0.000000  0.0\n",
      "C  0.921012  0.035620  0.005236  0.0\n",
      "T  0.015299  0.887773  0.010526  0.0\n",
      "G  0.008550  0.008435  0.972309  1.0\n",
      "-  0.055139  0.068172  0.011929  0.0\n",
      "{'Variant': 'T91C_C107T_C139G_A156G', 'Position': [91, 107, 139, 156], 'Alignment Probability': 0.795007925082894, 'Alignment Count': 85}\n",
      "Variant: NB22/RB03 85 0.795007925082894\n",
      "Processing RB03 NB96\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "       157  233       259\n",
      "A  0.27249  NaN  0.015244\n",
      "C  0.00000  NaN  0.796480\n",
      "T  0.00000  NaN  0.173584\n",
      "G  0.72751  NaN  0.014693\n",
      "-  0.00000  NaN  0.000000\n",
      "{'Variant': 'A157G_T233A_T259C', 'Position': [157, 233, 259], 'Alignment Probability': nan, 'Alignment Count': 53}\n",
      "Variant: NB96/RB03 53 nan\n",
      "Processing RB03 NB32\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        180       517       163\n",
      "A  0.000000  0.000000  0.000000\n",
      "C  0.000000  0.000000  0.000000\n",
      "T  0.004917  0.919435  0.958053\n",
      "G  0.995083  0.007409  0.000000\n",
      "-  0.000000  0.073155  0.041947\n",
      "{'Variant': '#PARENT#', 'Position': array([180, 517, 163]), 'Alignment Probability': 0.8765367192064489, 'Alignment Count': 98}\n",
      "Variant: NB32/RB03 98 0.8765367192064489\n",
      "Processing RB03 NB41\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "   137       362       372\n",
      "A  NaN  0.000000  0.000000\n",
      "C  NaN  0.993096  0.004276\n",
      "T  NaN  0.000000  0.000000\n",
      "G  NaN  0.000000  0.995724\n",
      "-  NaN  0.006904  0.000000\n",
      "{'Variant': 'G137A', 'Position': array([137, 362, 372]), 'Alignment Probability': nan, 'Alignment Count': 130}\n",
      "Variant: NB41/RB03 130 nan\n",
      "Processing RB03 NB17\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        251      259  179\n",
      "A  0.020280  0.00000  NaN\n",
      "C  0.000000  0.00557  NaN\n",
      "T  0.005813  0.96986  NaN\n",
      "G  0.957279  0.00000  NaN\n",
      "-  0.016628  0.02457  NaN\n",
      "{'Variant': 'G179A', 'Position': array([251, 259, 179]), 'Alignment Probability': nan, 'Alignment Count': 175}\n",
      "Variant: NB17/RB03 175 nan\n",
      "Processing RB03 NB66\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        71        66        226\n",
      "A  0.826105  0.984484  0.000000\n",
      "C  0.008312  0.000000  0.016905\n",
      "T  0.000000  0.000000  0.000000\n",
      "G  0.150331  0.015516  0.919224\n",
      "-  0.015253  0.000000  0.063870\n",
      "{'Variant': 'G71A', 'Position': array([ 71,  66, 226]), 'Alignment Probability': 0.7475931232616039, 'Alignment Count': 115}\n",
      "Variant: NB66/RB03 115 0.7475931232616039\n",
      "Processing RB03 NB85\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "/home/emre/minION_results/MinION_RBC_0902723_sup/Demultiplex_cpp_70_40k_reads/RB03/NB85/alignment_minimap.bam does not exist.\n",
      "Skipping Variant: NB85/RB03\n",
      "Processing RB03 NB61\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        404       626       334\n",
      "A  0.010762  0.000000  0.000000\n",
      "C  0.030089  0.000000  0.965910\n",
      "T  0.959150  0.013763  0.017254\n",
      "G  0.000000  0.986237  0.000000\n",
      "-  0.000000  0.000000  0.016836\n",
      "{'Variant': '#PARENT#', 'Position': array([404, 626, 334]), 'Alignment Probability': 0.913701641636998, 'Alignment Count': 59}\n",
      "Variant: NB61/RB03 59 0.913701641636998\n",
      "Processing RB03 NB56\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "   69        511       646\n",
      "A  0.0  0.975361  0.956954\n",
      "C  1.0  0.024639  0.000000\n",
      "T  0.0  0.000000  0.000000\n",
      "G  0.0  0.000000  0.022343\n",
      "-  0.0  0.000000  0.020703\n",
      "{'Variant': '#PARENT#', 'Position': array([ 69, 511, 646]), 'Alignment Probability': 0.9333758236333095, 'Alignment Count': 40}\n",
      "Variant: NB56/RB03 40 0.9333758236333095\n",
      "Processing RB03 NB46\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "   125  270       523\n",
      "A  0.0  0.0  0.000000\n",
      "C  1.0  0.0  0.977623\n",
      "T  0.0  1.0  0.000000\n",
      "G  0.0  0.0  0.000000\n",
      "-  0.0  0.0  0.022377\n",
      "{'Variant': '#PARENT#', 'Position': array([125, 270, 523]), 'Alignment Probability': 0.9776233274810667, 'Alignment Count': 35}\n",
      "Variant: NB46/RB03 35 0.9776233274810667\n",
      "Processing RB03 NB49\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        634       581  657\n",
      "A  0.000000  0.000000  0.0\n",
      "C  0.040126  0.012677  1.0\n",
      "T  0.013442  0.961389  0.0\n",
      "G  0.946432  0.000000  0.0\n",
      "-  0.000000  0.025934  0.0\n",
      "{'Variant': '#PARENT#', 'Position': array([634, 581, 657]), 'Alignment Probability': 0.9098895166913491, 'Alignment Count': 60}\n",
      "Variant: NB49/RB03 60 0.9098895166913491\n",
      "Processing RB03 NB77\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        101       166       168       177       182       187       206  \\\n",
      "A  0.635628  0.000000  0.000000  0.601163  0.409052  0.580539  0.007552   \n",
      "C  0.008367  0.432788  0.432947  0.000000  0.000000  0.000000  0.432617   \n",
      "T  0.000000  0.567212  0.561790  0.008671  0.000000  0.000000  0.544660   \n",
      "G  0.356005  0.000000  0.000000  0.390166  0.583593  0.411450  0.008320   \n",
      "-  0.000000  0.000000  0.005263  0.000000  0.007355  0.008011  0.006852   \n",
      "\n",
      "        223       230  \n",
      "A  0.407109  0.008630  \n",
      "C  0.016183  0.328108  \n",
      "T  0.569672  0.638366  \n",
      "G  0.007037  0.024897  \n",
      "-  0.000000  0.000000  \n",
      "{'Variant': '#PARENT#', 'Position': [101, 166, 168, 177, 182, 187, 206, 223, 230], 'Alignment Probability': 0.008171007706587661, 'Alignment Count': 116}\n",
      "Variant: NB77/RB03 116 0.008171007706587661\n",
      "Processing RB03 NB12\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        94        97        178       211       257       554       634  \\\n",
      "A  0.629789  0.636612  0.199648  0.216016  0.382172  0.622359  0.000000   \n",
      "C  0.000000  0.363388  0.000000  0.000000  0.000000  0.000000  0.395107   \n",
      "T  0.000000  0.000000  0.800352  0.000000  0.000000  0.000000  0.201808   \n",
      "G  0.000000  0.000000  0.000000  0.783984  0.617828  0.377641  0.403085   \n",
      "-  0.370211  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   \n",
      "\n",
      "        641  \n",
      "A  0.000000  \n",
      "C  0.361171  \n",
      "T  0.000000  \n",
      "G  0.638829  \n",
      "-  0.000000  \n",
      "{'Variant': 'A178T_A211G', 'Position': [94, 97, 178, 211, 257, 554, 634, 641], 'Alignment Probability': 0.02490846869733819, 'Alignment Count': 5}\n",
      "Variant: NB12/RB03 5 0.02490846869733819\n",
      "Processing RB03 NB38\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "   573       581       252\n",
      "A  0.0  0.000000  0.029163\n",
      "C  0.0  0.000000  0.970837\n",
      "T  1.0  0.866667  0.000000\n",
      "G  0.0  0.000000  0.000000\n",
      "-  0.0  0.133333  0.000000\n",
      "{'Variant': '#PARENT#', 'Position': array([573, 581, 252]), 'Alignment Probability': 0.8413921435861572, 'Alignment Count': 20}\n",
      "Variant: NB38/RB03 20 0.8413921435861572\n",
      "Processing RB03 NB64\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        163       498       389\n",
      "A  0.910647  0.980037  0.011774\n",
      "C  0.000000  0.000000  0.000000\n",
      "T  0.000000  0.000000  0.010881\n",
      "G  0.000000  0.000000  0.977345\n",
      "-  0.089353  0.019963  0.000000\n",
      "{'Variant': 'T163A', 'Position': array([163, 498, 389]), 'Alignment Probability': 0.8722489236877397, 'Alignment Count': 86}\n",
      "Variant: NB64/RB03 86 0.8722489236877397\n",
      "Processing RB03 NB71\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        126       130      142       227       242\n",
      "A  0.000000  0.014698  0.00000  0.000000  0.038283\n",
      "C  0.066479  0.940756  0.97988  0.030152  0.006458\n",
      "T  0.933521  0.000000  0.02012  0.962476  0.005582\n",
      "G  0.000000  0.000000  0.00000  0.000000  0.934404\n",
      "-  0.000000  0.044546  0.00000  0.007372  0.015274\n",
      "{'Variant': 'C126T_T130C_T142C_G227T_A242G', 'Position': [126, 130, 142, 227, 242], 'Alignment Probability': 0.7739249310933318, 'Alignment Count': 128}\n",
      "Variant: NB71/RB03 128 0.7739249310933318\n",
      "Processing RB03 NB24\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        178       211\n",
      "A  0.115093  0.115486\n",
      "C  0.016010  0.000000\n",
      "T  0.868897  0.000000\n",
      "G  0.000000  0.850997\n",
      "-  0.000000  0.033517\n",
      "{'Variant': 'A178T_A211G', 'Position': [178, 211], 'Alignment Probability': 0.7394291772441519, 'Alignment Count': 111}\n",
      "Variant: NB24/RB03 111 0.7394291772441519\n",
      "Processing RB03 NB94\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        77        185      200\n",
      "A  0.946300  0.906360  0.00000\n",
      "C  0.027598  0.044602  0.93305\n",
      "T  0.000000  0.016263  0.00000\n",
      "G  0.000000  0.000000  0.00908\n",
      "-  0.026102  0.032775  0.05787\n",
      "{'Variant': 'T77A_T185A_T200C', 'Position': [77, 185, 200], 'Alignment Probability': 0.8002663237858948, 'Alignment Count': 102}\n",
      "Variant: NB94/RB03 102 0.8002663237858948\n",
      "Processing RB03 NB50\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        80       176\n",
      "A  0.000000  0.04790\n",
      "C  0.958327  0.01796\n",
      "T  0.021566  0.93414\n",
      "G  0.009301  0.00000\n",
      "-  0.010806  0.00000\n",
      "{'Variant': 'T80C_A176T', 'Position': [80, 176], 'Alignment Probability': 0.8952113639302017, 'Alignment Count': 93}\n",
      "Variant: NB50/RB03 93 0.8952113639302017\n",
      "Processing RB03 NB65\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        97   147  549\n",
      "A  0.047246  0.0  1.0\n",
      "C  0.000000  0.0  0.0\n",
      "T  0.000000  1.0  0.0\n",
      "G  0.739297  0.0  0.0\n",
      "-  0.213457  0.0  0.0\n",
      "{'Variant': 'A97G', 'Position': array([ 97, 147, 549]), 'Alignment Probability': 0.7392972646775715, 'Alignment Count': 22}\n",
      "Variant: NB65/RB03 22 0.7392972646775715\n",
      "Processing RB03 NB15\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        456       441  503\n",
      "A  0.005783  0.085618  0.0\n",
      "C  0.981071  0.000000  1.0\n",
      "T  0.000000  0.011707  0.0\n",
      "G  0.013146  0.891587  0.0\n",
      "-  0.000000  0.011088  0.0\n",
      "{'Variant': '#PARENT#', 'Position': array([456, 441, 503]), 'Alignment Probability': 0.8747102050875002, 'Alignment Count': 151}\n",
      "Variant: NB15/RB03 151 0.8747102050875002\n",
      "Processing RB03 NB02\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        643  74        391\n",
      "A  0.926855  1.0  0.969443\n",
      "C  0.000000  0.0  0.012310\n",
      "T  0.000000  0.0  0.000000\n",
      "G  0.000000  0.0  0.000000\n",
      "-  0.073145  0.0  0.018247\n",
      "{'Variant': '#PARENT#', 'Position': array([643,  74, 391]), 'Alignment Probability': 0.8985335898360839, 'Alignment Count': 55}\n",
      "Variant: NB02/RB03 55 0.8985335898360839\n",
      "Processing RB03 NB01\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "/home/emre/minION_results/MinION_RBC_0902723_sup/Demultiplex_cpp_70_40k_reads/RB03/NB01/alignment_minimap.bam does not exist.\n",
      "Skipping Variant: NB01/RB03\n",
      "Processing RB03 NB27\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        166       478       429\n",
      "A  0.000000  0.000000  0.008028\n",
      "C  0.954987  0.008528  0.000000\n",
      "T  0.038845  0.983900  0.006344\n",
      "G  0.000000  0.000000  0.985628\n",
      "-  0.006168  0.007572  0.000000\n",
      "{'Variant': 'T166C', 'Position': array([166, 478, 429]), 'Alignment Probability': 0.9261081979200018, 'Alignment Count': 120}\n",
      "Variant: NB27/RB03 120 0.9261081979200018\n",
      "Processing RB03 NB08\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        270  313  194\n",
      "A  0.000000  NaN  0.0\n",
      "C  0.366806  NaN  0.0\n",
      "T  0.512023  NaN  0.0\n",
      "G  0.000000  NaN  1.0\n",
      "-  0.121171  NaN  0.0\n",
      "{'Variant': 'G313A', 'Position': array([270, 313, 194]), 'Alignment Probability': nan, 'Alignment Count': 89}\n",
      "Variant: NB08/RB03 89 nan\n",
      "Processing RB03 NB91\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        171       327       410\n",
      "A  0.919947  0.000000  0.000000\n",
      "C  0.000000  0.000000  0.017429\n",
      "T  0.062612  0.923125  0.982571\n",
      "G  0.017441  0.000000  0.000000\n",
      "-  0.000000  0.076875  0.000000\n",
      "{'Variant': 'T171A', 'Position': array([171, 327, 410]), 'Alignment Probability': 0.8344255764590741, 'Alignment Count': 58}\n",
      "Variant: NB91/RB03 58 0.8344255764590741\n",
      "Processing RB03 NB19\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        286       371       324\n",
      "A  0.000000  0.000000  0.000000\n",
      "C  0.000000  0.013776  0.944302\n",
      "T  0.007325  0.981408  0.013008\n",
      "G  0.978987  0.000000  0.031217\n",
      "-  0.013688  0.004816  0.011472\n",
      "{'Variant': '#PARENT#', 'Position': array([286, 371, 324]), 'Alignment Probability': 0.9072721110669357, 'Alignment Count': 132}\n",
      "Variant: NB19/RB03 132 0.9072721110669357\n",
      "Processing RB03 NB51\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        89        171       223\n",
      "A  0.000000  0.792482  0.000000\n",
      "C  0.721097  0.000000  0.715423\n",
      "T  0.268616  0.194259  0.193952\n",
      "G  0.000000  0.000000  0.053512\n",
      "-  0.010288  0.013259  0.037114\n",
      "{'Variant': 'T89C_T171A_T223C', 'Position': [89, 171, 223], 'Alignment Probability': 0.40883245567606347, 'Alignment Count': 72}\n",
      "Variant: NB51/RB03 72 0.40883245567606347\n",
      "Processing RB03 NB52\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        127       227\n",
      "A  0.000000  0.954018\n",
      "C  0.930473  0.000000\n",
      "T  0.027529  0.000000\n",
      "G  0.000000  0.015581\n",
      "-  0.041998  0.030401\n",
      "{'Variant': 'T127C_G227A', 'Position': [127, 227], 'Alignment Probability': 0.8876879075667243, 'Alignment Count': 94}\n",
      "Variant: NB52/RB03 94 0.8876879075667243\n",
      "Processing RB03 NB35\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "   138       193       217\n",
      "A  0.0  0.000000  0.000000\n",
      "C  0.0  0.969375  0.880573\n",
      "T  0.0  0.010902  0.029500\n",
      "G  1.0  0.000000  0.000000\n",
      "-  0.0  0.019723  0.089927\n",
      "{'Variant': 'A138G_T193C_T217C', 'Position': [138, 193, 217], 'Alignment Probability': 0.8536055471133784, 'Alignment Count': 68}\n",
      "Variant: NB35/RB03 68 0.8536055471133784\n",
      "Processing RB03 NB09\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        193       211\n",
      "A  0.927718  0.014838\n",
      "C  0.000000  0.962190\n",
      "T  0.000000  0.000000\n",
      "G  0.008552  0.016033\n",
      "-  0.063730  0.006940\n",
      "{'Variant': 'T193A_A211C', 'Position': [193, 211], 'Alignment Probability': 0.8926404558434934, 'Alignment Count': 112}\n",
      "Variant: NB09/RB03 112 0.8926404558434934\n",
      "Processing RB03 NB11\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "   220       514  600\n",
      "A  0.0  0.000000  1.0\n",
      "C  0.0  0.977556  0.0\n",
      "T  0.0  0.011203  0.0\n",
      "G  1.0  0.011241  0.0\n",
      "-  0.0  0.000000  0.0\n",
      "{'Variant': 'A220G', 'Position': array([220, 514, 600]), 'Alignment Probability': 0.9775564022734319, 'Alignment Count': 89}\n",
      "Variant: NB11/RB03 89 0.9775564022734319\n",
      "Processing RB03 NB63\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        174       63   167\n",
      "A  0.088852  0.000000  0.0\n",
      "C  0.000000  0.014516  0.0\n",
      "T  0.000000  0.000000  0.0\n",
      "G  0.843834  0.985484  1.0\n",
      "-  0.067314  0.000000  0.0\n",
      "{'Variant': 'A174G', 'Position': array([174,  63, 167]), 'Alignment Probability': 0.8315840713152657, 'Alignment Count': 68}\n",
      "Variant: NB63/RB03 68 0.8315840713152657\n",
      "Processing RB03 NB42\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        211       108       196\n",
      "A  0.020497  0.981773  0.979577\n",
      "C  0.000000  0.000000  0.000000\n",
      "T  0.000000  0.000000  0.000000\n",
      "G  0.901844  0.018227  0.020423\n",
      "-  0.077658  0.000000  0.000000\n",
      "{'Variant': 'A211G', 'Position': array([211, 108, 196]), 'Alignment Probability': 0.8673234134917646, 'Alignment Count': 146}\n",
      "Variant: NB42/RB03 146 0.8673234134917646\n",
      "Processing RB03 NB18\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        66        154       162       222\n",
      "A  0.529880  0.471008  0.479806  0.000000\n",
      "C  0.445524  0.000000  0.000000  0.488393\n",
      "T  0.024596  0.000000  0.000000  0.500368\n",
      "G  0.000000  0.502055  0.497985  0.000000\n",
      "-  0.000000  0.026937  0.022209  0.011239\n",
      "{'Variant': 'A154G', 'Position': [66, 154, 162, 222], 'Alignment Probability': 0.06628782016943746, 'Alignment Count': 70}\n",
      "Variant: NB18/RB03 70 0.06628782016943746\n",
      "Processing RB03 NB90\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "   654       226       354\n",
      "A  0.0  0.000000  0.000000\n",
      "C  1.0  0.000000  0.037197\n",
      "T  0.0  0.000000  0.946721\n",
      "G  0.0  0.902604  0.000000\n",
      "-  0.0  0.097396  0.016082\n",
      "{'Variant': '#PARENT#', 'Position': array([654, 226, 354]), 'Alignment Probability': 0.8545136782209047, 'Alignment Count': 50}\n",
      "Variant: NB90/RB03 50 0.8545136782209047\n",
      "Processing RB03 NB05\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "   303       551       516\n",
      "A  0.0  0.034688  0.012314\n",
      "C  0.0  0.913515  0.000000\n",
      "T  1.0  0.032557  0.923331\n",
      "G  0.0  0.000000  0.000000\n",
      "-  0.0  0.019240  0.064355\n",
      "{'Variant': '#PARENT#', 'Position': array([303, 551, 516]), 'Alignment Probability': 0.8434769569339193, 'Alignment Count': 82}\n",
      "Variant: NB05/RB03 82 0.8434769569339193\n",
      "Processing RB03 NB39\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        157       166\n",
      "A  0.623967  0.000000\n",
      "C  0.000000  0.400588\n",
      "T  0.000000  0.572998\n",
      "G  0.376033  0.000000\n",
      "-  0.000000  0.026414\n",
      "{'Variant': '#PARENT#', 'Position': [157, 166], 'Alignment Probability': 0.3575320600613673, 'Alignment Count': 102}\n",
      "Variant: NB39/RB03 102 0.3575320600613673\n",
      "Processing RB03 NB69\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        297       465       197\n",
      "A  0.983108  0.020497  0.978389\n",
      "C  0.016892  0.979503  0.000000\n",
      "T  0.000000  0.000000  0.000000\n",
      "G  0.000000  0.000000  0.021611\n",
      "-  0.000000  0.000000  0.000000\n",
      "{'Variant': '#PARENT#', 'Position': array([297, 465, 197]), 'Alignment Probability': 0.9421471532394462, 'Alignment Count': 42}\n",
      "Variant: NB69/RB03 42 0.9421471532394462\n",
      "Processing RB03 NB25\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "       176  196\n",
      "A  0.12042  0.0\n",
      "C  0.00000  0.0\n",
      "T  0.87958  0.0\n",
      "G  0.00000  1.0\n",
      "-  0.00000  0.0\n",
      "{'Variant': 'A176T_A196G', 'Position': [176, 196], 'Alignment Probability': 0.8795802627120864, 'Alignment Count': 15}\n",
      "Variant: NB25/RB03 15 0.8795802627120864\n",
      "Processing RB01 NB87\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        570       555  536\n",
      "A  0.004545  0.000000  NaN\n",
      "C  0.005607  0.000000  NaN\n",
      "T  0.989848  0.000000  NaN\n",
      "G  0.000000  0.979627  NaN\n",
      "-  0.000000  0.020373  NaN\n",
      "{'Variant': 'C536A', 'Position': array([570, 555, 536]), 'Alignment Probability': nan, 'Alignment Count': 177}\n",
      "Variant: NB87/RB01 177 nan\n",
      "Processing RB01 NB03\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        174       521       197\n",
      "A  0.534337  0.005233  0.978468\n",
      "C  0.000000  0.014627  0.000000\n",
      "T  0.000000  0.000000  0.000000\n",
      "G  0.431361  0.974898  0.021532\n",
      "-  0.034302  0.005241  0.000000\n",
      "{'Variant': '#PARENT#', 'Position': array([174, 521, 197]), 'Alignment Probability': 0.509708189686276, 'Alignment Count': 172}\n",
      "Variant: NB03/RB01 172 0.509708189686276\n",
      "Processing RB01 NB48\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "   656  570       64 \n",
      "A  0.0  0.0  0.935659\n",
      "C  1.0  0.0  0.000000\n",
      "T  0.0  1.0  0.000000\n",
      "G  0.0  0.0  0.064341\n",
      "-  0.0  0.0  0.000000\n",
      "{'Variant': '#PARENT#', 'Position': array([656, 570,  64]), 'Alignment Probability': 0.935659350000092, 'Alignment Count': 16}\n",
      "Variant: NB48/RB01 16 0.935659350000092\n",
      "Processing RB01 NB70\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "   661       382  545\n",
      "A  0.0  0.935817  1.0\n",
      "C  0.0  0.064183  0.0\n",
      "T  0.0  0.000000  0.0\n",
      "G  1.0  0.000000  0.0\n",
      "-  0.0  0.000000  0.0\n",
      "{'Variant': '#PARENT#', 'Position': array([661, 382, 545]), 'Alignment Probability': 0.9358169958667577, 'Alignment Count': 15}\n",
      "Variant: NB70/RB01 15 0.9358169958667577\n",
      "Processing RB01 NB04\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        130       566       401\n",
      "A  0.000000  0.000000  0.000000\n",
      "C  0.361463  0.970209  0.000000\n",
      "T  0.615524  0.011697  0.000000\n",
      "G  0.000000  0.000000  0.990513\n",
      "-  0.023013  0.018094  0.009487\n",
      "{'Variant': '#PARENT#', 'Position': array([130, 566, 401]), 'Alignment Probability': 0.5915215843027077, 'Alignment Count': 87}\n",
      "Variant: NB04/RB01 87 0.5915215843027077\n",
      "Processing RB01 NB37\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        169       581       273\n",
      "A  0.018902  0.000000  0.982328\n",
      "C  0.000000  0.000000  0.000000\n",
      "T  0.922520  0.925217  0.000000\n",
      "G  0.000000  0.004627  0.017672\n",
      "-  0.058578  0.070157  0.000000\n",
      "{'Variant': 'A169T', 'Position': array([169, 581, 273]), 'Alignment Probability': 0.8384470013412799, 'Alignment Count': 108}\n",
      "Variant: NB37/RB01 108 0.8384470013412799\n",
      "Processing RB01 NB88\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        193       261       398\n",
      "A  0.000000  0.000000  0.000000\n",
      "C  0.044595  0.015732  0.033821\n",
      "T  0.942173  0.984268  0.966179\n",
      "G  0.000000  0.000000  0.000000\n",
      "-  0.013232  0.000000  0.000000\n",
      "{'Variant': '#PARENT#', 'Position': array([193, 261, 398]), 'Alignment Probability': 0.8959864640121868, 'Alignment Count': 56}\n",
      "Variant: NB88/RB01 56 0.8959864640121868\n",
      "Processing RB01 NB14\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        105       171       156\n",
      "A  0.987067  0.000000  0.990580\n",
      "C  0.000000  0.007556  0.000000\n",
      "T  0.000000  0.959019  0.000000\n",
      "G  0.012933  0.006126  0.003541\n",
      "-  0.000000  0.027300  0.005879\n",
      "{'Variant': '#PARENT#', 'Position': array([105, 171, 156]), 'Alignment Probability': 0.9376979848198573, 'Alignment Count': 166}\n",
      "Variant: NB14/RB01 166 0.9376979848198573\n",
      "Processing RB01 NB82\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "   363  494  132\n",
      "A  0.0  0.0  1.0\n",
      "C  0.0  0.0  0.0\n",
      "T  1.0  1.0  0.0\n",
      "G  0.0  0.0  0.0\n",
      "-  0.0  0.0  0.0\n",
      "{'Variant': '#PARENT#', 'Position': array([363, 494, 132]), 'Alignment Probability': 1.0, 'Alignment Count': 13}\n",
      "Variant: NB82/RB01 13 1.0\n",
      "Processing RB01 NB54\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        178       215       634\n",
      "A  0.978991  0.000000  0.000000\n",
      "C  0.000000  0.953498  0.028454\n",
      "T  0.000000  0.022834  0.000000\n",
      "G  0.000000  0.023667  0.971546\n",
      "-  0.021009  0.000000  0.000000\n",
      "{'Variant': '#PARENT#', 'Position': array([178, 215, 634]), 'Alignment Probability': 0.906904652079735, 'Alignment Count': 34}\n",
      "Variant: NB54/RB01 34 0.906904652079735\n",
      "Processing RB01 NB44\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        398       550       140\n",
      "A  0.008744  0.000000  0.004006\n",
      "C  0.005696  0.005735  0.002727\n",
      "T  0.974569  0.994265  0.989611\n",
      "G  0.000000  0.000000  0.003656\n",
      "-  0.010991  0.000000  0.000000\n",
      "{'Variant': '#PARENT#', 'Position': array([398, 550, 140]), 'Alignment Probability': 0.9589130184920908, 'Alignment Count': 178}\n",
      "Variant: NB44/RB01 178 0.9589130184920908\n",
      "Processing RB01 NB57\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        491  61        355\n",
      "A  0.000000  0.0  0.020501\n",
      "C  0.015961  1.0  0.000000\n",
      "T  0.967779  0.0  0.000000\n",
      "G  0.000000  0.0  0.960961\n",
      "-  0.016260  0.0  0.018538\n",
      "{'Variant': '#PARENT#', 'Position': array([491,  61, 355]), 'Alignment Probability': 0.9299972652012701, 'Alignment Count': 49}\n",
      "Variant: NB57/RB01 49 0.9299972652012701\n",
      "Processing RB01 NB40\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "   111       64        242\n",
      "A  0.0  0.983487  0.959533\n",
      "C  0.0  0.000000  0.000000\n",
      "T  0.0  0.000000  0.019978\n",
      "G  1.0  0.010308  0.020489\n",
      "-  0.0  0.006205  0.000000\n",
      "{'Variant': '#PARENT#', 'Position': array([111,  64, 242]), 'Alignment Probability': 0.9436886176188919, 'Alignment Count': 99}\n",
      "Variant: NB40/RB01 99 0.9436886176188919\n",
      "Processing RB01 NB81\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "/home/emre/minION_results/MinION_RBC_0902723_sup/Demultiplex_cpp_70_40k_reads/RB01/NB81/alignment_minimap.bam does not exist.\n",
      "Skipping Variant: NB81/RB01\n",
      "Processing RB01 NB13\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        55        70        154\n",
      "A  0.000000  0.011534  0.943175\n",
      "C  0.271609  0.974346  0.004402\n",
      "T  0.689394  0.003248  0.000000\n",
      "G  0.000000  0.004213  0.000000\n",
      "-  0.038997  0.006660  0.052423\n",
      "{'Variant': 'C55T', 'Position': array([ 55,  70, 154]), 'Alignment Probability': 0.6335380417082075, 'Alignment Count': 230}\n",
      "Variant: NB13/RB01 230 0.6335380417082075\n",
      "Processing RB01 NB53\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        613       519       547\n",
      "A  0.603874  0.000000  0.922936\n",
      "C  0.387519  0.015189  0.000000\n",
      "T  0.000000  0.984811  0.000000\n",
      "G  0.008607  0.000000  0.000000\n",
      "-  0.000000  0.000000  0.077064\n",
      "{'Variant': 'C613A', 'Position': array([613, 519, 547]), 'Alignment Probability': 0.5488719123360869, 'Alignment Count': 84}\n",
      "Variant: NB53/RB01 84 0.5488719123360869\n",
      "Processing RB01 NB80\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "   93        198  508\n",
      "A  NaN  0.005779  0.0\n",
      "C  NaN  0.000000  0.0\n",
      "T  NaN  0.003842  1.0\n",
      "G  NaN  0.986335  0.0\n",
      "-  NaN  0.004043  0.0\n",
      "{'Variant': 'G93A', 'Position': array([ 93, 198, 508]), 'Alignment Probability': nan, 'Alignment Count': 189}\n",
      "Variant: NB80/RB01 189 nan\n",
      "Processing RB01 NB47\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        290  107  170\n",
      "A  0.000000  0.0  0.0\n",
      "C  0.024531  1.0  1.0\n",
      "T  0.000000  0.0  0.0\n",
      "G  0.947298  0.0  0.0\n",
      "-  0.028172  0.0  0.0\n",
      "{'Variant': '#PARENT#', 'Position': array([290, 107, 170]), 'Alignment Probability': 0.9472978478003493, 'Alignment Count': 30}\n",
      "Variant: NB47/RB01 30 0.9472978478003493\n",
      "Processing RB01 NB84\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        120       401       237\n",
      "A  0.000000  0.004994  0.000000\n",
      "C  0.942321  0.007998  0.008646\n",
      "T  0.028222  0.000000  0.000000\n",
      "G  0.016985  0.979803  0.955652\n",
      "-  0.012472  0.007206  0.035702\n",
      "{'Variant': 'T120C', 'Position': array([120, 401, 237]), 'Alignment Probability': 0.882342548446408, 'Alignment Count': 120}\n",
      "Variant: NB84/RB01 120 0.882342548446408\n",
      "Processing RB01 NB72\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "   603  318       387\n",
      "A  0.0  0.0  0.000000\n",
      "C  0.0  1.0  0.903228\n",
      "T  1.0  0.0  0.096772\n",
      "G  0.0  0.0  0.000000\n",
      "-  0.0  0.0  0.000000\n",
      "{'Variant': '#PARENT#', 'Position': array([603, 318, 387]), 'Alignment Probability': 0.9032284228470342, 'Alignment Count': 8}\n",
      "Variant: NB72/RB01 8 0.9032284228470342\n",
      "Processing RB01 NB43\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        629  585       497\n",
      "A  0.003872  NaN  0.000000\n",
      "C  0.014375  NaN  0.993018\n",
      "T  0.969948  NaN  0.006982\n",
      "G  0.000000  NaN  0.000000\n",
      "-  0.011805  NaN  0.000000\n",
      "{'Variant': 'G585A', 'Position': array([629, 585, 497]), 'Alignment Probability': nan, 'Alignment Count': 241}\n",
      "Variant: NB43/RB01 241 nan\n",
      "Processing RB01 NB79\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        135  407  390\n",
      "A  0.250931  0.0  0.0\n",
      "C  0.000000  0.0  0.0\n",
      "T  0.000000  1.0  1.0\n",
      "G  0.749069  0.0  0.0\n",
      "-  0.000000  0.0  0.0\n",
      "{'Variant': 'A135G', 'Position': array([135, 407, 390]), 'Alignment Probability': 0.7490688046885854, 'Alignment Count': 8}\n",
      "Variant: NB79/RB01 8 0.7490688046885854\n",
      "Processing RB01 NB36\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "   573  124       149\n",
      "A  0.0  1.0  0.897675\n",
      "C  0.0  0.0  0.000000\n",
      "T  1.0  0.0  0.000000\n",
      "G  0.0  0.0  0.000000\n",
      "-  0.0  0.0  0.102325\n",
      "{'Variant': '#PARENT#', 'Position': array([573, 124, 149]), 'Alignment Probability': 0.8976747016832044, 'Alignment Count': 7}\n",
      "Variant: NB36/RB01 7 0.8976747016832044\n",
      "Processing RB01 NB30\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        155       217\n",
      "A  0.023253  0.000000\n",
      "C  0.020539  0.833778\n",
      "T  0.000000  0.093016\n",
      "G  0.945541  0.000000\n",
      "-  0.010667  0.073206\n",
      "{'Variant': 'A155G_T217C', 'Position': [155, 217], 'Alignment Probability': 0.7883713374183797, 'Alignment Count': 84}\n",
      "Variant: NB30/RB01 84 0.7883713374183797\n",
      "Processing RB01 NB60\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        453  416  346\n",
      "A  0.000000  0.0  1.0\n",
      "C  0.720189  0.0  0.0\n",
      "T  0.000000  1.0  0.0\n",
      "G  0.000000  0.0  0.0\n",
      "-  0.279811  0.0  0.0\n",
      "{'Variant': '#PARENT#', 'Position': array([453, 416, 346]), 'Alignment Probability': 0.7201888647643352, 'Alignment Count': 22}\n",
      "Variant: NB60/RB01 22 0.7201888647643352\n",
      "Processing RB01 NB45\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "   175       346       78 \n",
      "A  1.0  0.977054  0.029876\n",
      "C  0.0  0.000000  0.000000\n",
      "T  0.0  0.000000  0.027919\n",
      "G  0.0  0.000000  0.942204\n",
      "-  0.0  0.022946  0.000000\n",
      "{'Variant': '#PARENT#', 'Position': array([175, 346,  78]), 'Alignment Probability': 0.9205848617501011, 'Alignment Count': 34}\n",
      "Variant: NB45/RB01 34 0.9205848617501011\n",
      "Processing RB01 NB93\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        525       458  573\n",
      "A  0.868534  0.000000  0.0\n",
      "C  0.037105  0.024327  0.0\n",
      "T  0.000000  0.975673  1.0\n",
      "G  0.000000  0.000000  0.0\n",
      "-  0.094360  0.000000  0.0\n",
      "{'Variant': '#PARENT#', 'Position': array([525, 458, 573]), 'Alignment Probability': 0.8474053729958967, 'Alignment Count': 72}\n",
      "Variant: NB93/RB01 72 0.8474053729958967\n",
      "Processing RB01 NB31\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        447       313  196\n",
      "A  0.977953  0.040828  1.0\n",
      "C  0.000000  0.000000  0.0\n",
      "T  0.000000  0.000000  0.0\n",
      "G  0.000000  0.959172  0.0\n",
      "-  0.022047  0.000000  0.0\n",
      "{'Variant': '#PARENT#', 'Position': array([447, 313, 196]), 'Alignment Probability': 0.9380251751125096, 'Alignment Count': 43}\n",
      "Variant: NB31/RB01 43 0.9380251751125096\n",
      "Processing RB01 NB75\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "   568  264  90 \n",
      "A  0.0  0.0  0.0\n",
      "C  1.0  0.0  0.0\n",
      "T  0.0  1.0  0.0\n",
      "G  0.0  0.0  1.0\n",
      "-  0.0  0.0  0.0\n",
      "{'Variant': '#PARENT#', 'Position': array([568, 264,  90]), 'Alignment Probability': 1.0, 'Alignment Count': 17}\n",
      "Variant: NB75/RB01 17 1.0\n",
      "Processing RB01 NB21\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "/home/emre/minION_results/MinION_RBC_0902723_sup/Demultiplex_cpp_70_40k_reads/RB01/NB21/alignment_minimap.bam does not exist.\n",
      "Skipping Variant: NB21/RB01\n",
      "Processing RB01 NB86\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        620       413  355\n",
      "A  0.000000  0.000000  0.0\n",
      "C  0.000000  0.995038  0.0\n",
      "T  0.977508  0.000000  0.0\n",
      "G  0.016101  0.000000  1.0\n",
      "-  0.006391  0.004962  0.0\n",
      "{'Variant': '#PARENT#', 'Position': array([620, 413, 355]), 'Alignment Probability': 0.9726569260117381, 'Alignment Count': 124}\n",
      "Variant: NB86/RB01 124 0.9726569260117381\n",
      "Processing RB01 NB07\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        241  229  151\n",
      "A  0.000000  1.0  0.0\n",
      "C  0.924151  0.0  0.0\n",
      "T  0.039017  0.0  0.0\n",
      "G  0.000000  0.0  1.0\n",
      "-  0.036831  0.0  0.0\n",
      "{'Variant': '#PARENT#', 'Position': array([241, 229, 151]), 'Alignment Probability': 0.9241514241726286, 'Alignment Count': 76}\n",
      "Variant: NB07/RB01 76 0.9241514241726286\n",
      "Processing RB01 NB34\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "/home/emre/minION_results/MinION_RBC_0902723_sup/Demultiplex_cpp_70_40k_reads/RB01/NB34/alignment_minimap.bam does not exist.\n",
      "Skipping Variant: NB34/RB01\n",
      "Processing RB01 NB62\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        91   237  371\n",
      "A  0.692776  0.0  0.0\n",
      "C  0.019296  0.0  0.0\n",
      "T  0.280816  0.0  1.0\n",
      "G  0.000000  1.0  0.0\n",
      "-  0.007112  0.0  0.0\n",
      "{'Variant': 'T91A', 'Position': array([ 91, 237, 371]), 'Alignment Probability': 0.6927758656449341, 'Alignment Count': 103}\n",
      "Variant: NB62/RB01 103 0.6927758656449341\n",
      "Processing RB01 NB23\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        175  377  103\n",
      "A  0.045670  0.0  NaN\n",
      "C  0.000000  0.0  NaN\n",
      "T  0.000000  0.0  NaN\n",
      "G  0.940239  1.0  NaN\n",
      "-  0.014091  0.0  NaN\n",
      "{'Variant': 'A175G_C103A', 'Position': array([175, 377, 103]), 'Alignment Probability': nan, 'Alignment Count': 127}\n",
      "Variant: NB23/RB01 127 nan\n",
      "Processing RB01 NB92\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        354       467  531\n",
      "A  0.000000  0.000000  0.0\n",
      "C  0.021799  0.019155  0.0\n",
      "T  0.978201  0.980845  0.0\n",
      "G  0.000000  0.000000  1.0\n",
      "-  0.000000  0.000000  0.0\n",
      "{'Variant': '#PARENT#', 'Position': array([354, 467, 531]), 'Alignment Probability': 0.9594627847811734, 'Alignment Count': 46}\n",
      "Variant: NB92/RB01 46 0.9594627847811734\n",
      "Processing RB01 NB73\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        120       148\n",
      "A  0.000000  0.000000\n",
      "C  0.399794  0.562542\n",
      "T  0.559598  0.437458\n",
      "G  0.040608  0.000000\n",
      "-  0.000000  0.000000\n",
      "{'Variant': 'T148C', 'Position': [120, 148], 'Alignment Probability': 0.3147972153069414, 'Alignment Count': 25}\n",
      "Variant: NB73/RB01 25 0.3147972153069414\n",
      "Processing RB01 NB67\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        324       119       298\n",
      "A  0.000000  0.022383  0.977897\n",
      "C  0.982248  0.000000  0.000000\n",
      "T  0.017752  0.000000  0.000000\n",
      "G  0.000000  0.958442  0.022103\n",
      "-  0.000000  0.019175  0.000000\n",
      "{'Variant': '#PARENT#', 'Position': array([324, 119, 298]), 'Alignment Probability': 0.9206191414718135, 'Alignment Count': 40}\n",
      "Variant: NB67/RB01 40 0.9206191414718135\n",
      "Processing RB01 NB29\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        249    587  345\n",
      "A  0.014636  0.000  0.0\n",
      "C  0.068381  0.000  0.0\n",
      "T  0.014390  0.984  0.0\n",
      "G  0.883294  0.000  1.0\n",
      "-  0.019300  0.016  0.0\n",
      "{'Variant': '#PARENT#', 'Position': array([249, 587, 345]), 'Alignment Probability': 0.8691612799862994, 'Alignment Count': 53}\n",
      "Variant: NB29/RB01 53 0.8691612799862994\n",
      "Processing RB01 NB06\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "   94   109       381\n",
      "A  NaN  0.0  0.018709\n",
      "C  NaN  1.0  0.022635\n",
      "T  NaN  0.0  0.005465\n",
      "G  NaN  0.0  0.940959\n",
      "-  NaN  0.0  0.012232\n",
      "{'Variant': '#PARENT#', 'Position': array([ 94, 109, 381]), 'Alignment Probability': nan, 'Alignment Count': 129}\n",
      "Variant: NB06/RB01 129 nan\n",
      "Processing RB01 NB78\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        194  126       449\n",
      "A  0.018724  0.0  0.010717\n",
      "C  0.010819  1.0  0.926714\n",
      "T  0.000000  0.0  0.034545\n",
      "G  0.970457  0.0  0.022054\n",
      "-  0.000000  0.0  0.005971\n",
      "{'Variant': '#PARENT#', 'Position': array([194, 126, 449]), 'Alignment Probability': 0.8993360252019663, 'Alignment Count': 90}\n",
      "Variant: NB78/RB01 90 0.8993360252019663\n",
      "Processing RB01 NB95\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "       120       633  646\n",
      "A  0.00000  0.000000  1.0\n",
      "C  0.00000  0.000000  0.0\n",
      "T  0.69109  0.948999  0.0\n",
      "G  0.00000  0.000000  0.0\n",
      "-  0.30891  0.051001  0.0\n",
      "{'Variant': '#PARENT#', 'Position': array([120, 633, 646]), 'Alignment Probability': 0.6558441794341854, 'Alignment Count': 12}\n",
      "Variant: NB95/RB01 12 0.6558441794341854\n",
      "Processing RB01 NB26\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "   495  487  189\n",
      "A  1.0  0.0  0.0\n",
      "C  0.0  0.0  0.0\n",
      "T  0.0  1.0  0.0\n",
      "G  0.0  0.0  1.0\n",
      "-  0.0  0.0  0.0\n",
      "{'Variant': '#PARENT#', 'Position': array([495, 487, 189]), 'Alignment Probability': 1.0, 'Alignment Count': 12}\n",
      "Variant: NB26/RB01 12 1.0\n",
      "Processing RB01 NB16\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        597       61   346\n",
      "A  0.939844  0.000000  1.0\n",
      "C  0.000000  0.890301  0.0\n",
      "T  0.060156  0.050233  0.0\n",
      "G  0.000000  0.000000  0.0\n",
      "-  0.000000  0.059466  0.0\n",
      "{'Variant': 'G597A', 'Position': array([597,  61, 346]), 'Alignment Probability': 0.8367438616371589, 'Alignment Count': 17}\n",
      "Variant: NB16/RB01 17 0.8367438616371589\n",
      "Processing RB01 NB76\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        196  395  500\n",
      "A  0.024352  1.0  0.0\n",
      "C  0.000000  0.0  1.0\n",
      "T  0.000000  0.0  0.0\n",
      "G  0.975648  0.0  0.0\n",
      "-  0.000000  0.0  0.0\n",
      "{'Variant': 'A196G', 'Position': array([196, 395, 500]), 'Alignment Probability': 0.9756477510776299, 'Alignment Count': 39}\n",
      "Variant: NB76/RB01 39 0.9756477510776299\n",
      "Processing RB01 NB59\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "/home/emre/minION_results/MinION_RBC_0902723_sup/Demultiplex_cpp_70_40k_reads/RB01/NB59/alignment_minimap.bam does not exist.\n",
      "Skipping Variant: NB59/RB01\n",
      "Processing RB01 NB68\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        652  52        459\n",
      "A  0.962965  0.0  0.910867\n",
      "C  0.000000  0.0  0.026806\n",
      "T  0.000000  1.0  0.028757\n",
      "G  0.037035  0.0  0.000000\n",
      "-  0.000000  0.0  0.033569\n",
      "{'Variant': '#PARENT#', 'Position': array([652,  52, 459]), 'Alignment Probability': 0.8771330249100381, 'Alignment Count': 27}\n",
      "Variant: NB68/RB01 27 0.8771330249100381\n",
      "Processing RB01 NB55\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "   659       361       526\n",
      "A  0.0  0.980963  0.968179\n",
      "C  1.0  0.000000  0.007189\n",
      "T  0.0  0.000000  0.000000\n",
      "G  0.0  0.011011  0.007879\n",
      "-  0.0  0.008025  0.016753\n",
      "{'Variant': '#PARENT#', 'Position': array([659, 361, 526]), 'Alignment Probability': 0.9497482585623085, 'Alignment Count': 92}\n",
      "Variant: NB55/RB01 92 0.9497482585623085\n",
      "Processing RB01 NB74\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        155  652       101\n",
      "A  0.066484  1.0  0.965914\n",
      "C  0.016079  0.0  0.000000\n",
      "T  0.018496  0.0  0.015038\n",
      "G  0.881601  0.0  0.019048\n",
      "-  0.017340  0.0  0.000000\n",
      "{'Variant': 'A155G', 'Position': array([155, 652, 101]), 'Alignment Probability': 0.8515500342633082, 'Alignment Count': 55}\n",
      "Variant: NB74/RB01 55 0.8515500342633082\n",
      "Processing RB01 NB89\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "   569  164  313\n",
      "A  0.0  1.0  0.0\n",
      "C  0.0  0.0  0.0\n",
      "T  1.0  0.0  0.0\n",
      "G  0.0  0.0  1.0\n",
      "-  0.0  0.0  0.0\n",
      "{'Variant': '#PARENT#', 'Position': array([569, 164, 313]), 'Alignment Probability': 1.0, 'Alignment Count': 18}\n",
      "Variant: NB89/RB01 18 1.0\n",
      "Processing RB01 NB83\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        319       408       75 \n",
      "A  0.010867  0.010947  0.000000\n",
      "C  0.000000  0.981010  0.017023\n",
      "T  0.003898  0.008044  0.957668\n",
      "G  0.978619  0.000000  0.004616\n",
      "-  0.006616  0.000000  0.020693\n",
      "{'Variant': '#PARENT#', 'Position': array([319, 408,  75]), 'Alignment Probability': 0.919394648447092, 'Alignment Count': 165}\n",
      "Variant: NB83/RB01 165 0.919394648447092\n",
      "Processing RB01 NB96\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        144       165\n",
      "A  0.504870  0.495961\n",
      "C  0.007485  0.000000\n",
      "T  0.000000  0.490238\n",
      "G  0.487645  0.013801\n",
      "-  0.000000  0.000000\n",
      "{'Variant': '#PARENT#', 'Position': [144, 165], 'Alignment Probability': 0.25039572156314177, 'Alignment Count': 101}\n",
      "Variant: NB96/RB01 101 0.25039572156314177\n",
      "Processing RB01 NB32\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        88   246  54 \n",
      "A  0.020904  1.0  1.0\n",
      "C  0.000000  0.0  0.0\n",
      "T  0.000000  0.0  0.0\n",
      "G  0.979096  0.0  0.0\n",
      "-  0.000000  0.0  0.0\n",
      "{'Variant': '#PARENT#', 'Position': array([ 88, 246,  54]), 'Alignment Probability': 0.9790957905209281, 'Alignment Count': 43}\n",
      "Variant: NB32/RB01 43 0.9790957905209281\n",
      "Processing RB01 NB41\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        270      633       152\n",
      "A  0.000000  0.00000  0.000000\n",
      "C  0.007426  0.02844  0.007874\n",
      "T  0.992574  0.97156  0.992126\n",
      "G  0.000000  0.00000  0.000000\n",
      "-  0.000000  0.00000  0.000000\n",
      "{'Variant': '#PARENT#', 'Position': array([270, 633, 152]), 'Alignment Probability': 0.9567520745278204, 'Alignment Count': 129}\n",
      "Variant: NB41/RB01 129 0.9567520745278204\n",
      "Processing RB01 NB17\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "   655       478      252\n",
      "A  1.0  0.000000  0.00000\n",
      "C  0.0  0.011038  0.99017\n",
      "T  0.0  0.980000  0.00983\n",
      "G  0.0  0.000000  0.00000\n",
      "-  0.0  0.008962  0.00000\n",
      "{'Variant': '#PARENT#', 'Position': array([655, 478, 252]), 'Alignment Probability': 0.9703670608066378, 'Alignment Count': 92}\n",
      "Variant: NB17/RB01 92 0.9703670608066378\n",
      "Processing RB01 NB66\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        82        99        102       128\n",
      "A  0.000000  0.000000  0.000000  0.000000\n",
      "C  0.000000  0.990168  0.015150  0.943199\n",
      "T  0.000000  0.000000  0.000000  0.043581\n",
      "G  0.986021  0.009832  0.015517  0.013220\n",
      "-  0.013979  0.000000  0.969333  0.000000\n",
      "{'Variant': 'A82G_T99C_C102DEL_T128C', 'Position': [82, 99, 102, 128], 'Alignment Probability': 0.8926288597912746, 'Alignment Count': 59}\n",
      "Variant: NB66/RB01 59 0.8926288597912746\n",
      "Processing RB01 NB61\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        76        168       223\n",
      "A  0.533100  0.005223  0.000000\n",
      "C  0.004706  0.517097  0.427267\n",
      "T  0.000000  0.464178  0.548670\n",
      "G  0.458975  0.000000  0.024064\n",
      "-  0.003219  0.013502  0.000000\n",
      "{'Variant': 'T168C', 'Position': [76, 168, 223], 'Alignment Probability': 0.15124885671417465, 'Alignment Count': 191}\n",
      "Variant: NB61/RB01 191 0.15124885671417465\n",
      "Processing RB01 NB56\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        420       185       120\n",
      "A  0.985138  0.004056  0.000000\n",
      "C  0.000000  0.011364  0.016372\n",
      "T  0.000000  0.966706  0.928469\n",
      "G  0.000000  0.000000  0.010893\n",
      "-  0.014862  0.017874  0.044266\n",
      "{'Variant': '#PARENT#', 'Position': array([420, 185, 120]), 'Alignment Probability': 0.8842177650547729, 'Alignment Count': 165}\n",
      "Variant: NB56/RB01 165 0.8842177650547729\n",
      "Processing RB01 NB46\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        129       121       544\n",
      "A  0.009205  0.018074  0.970124\n",
      "C  0.937450  0.008075  0.000000\n",
      "T  0.018706  0.973851  0.006103\n",
      "G  0.009596  0.000000  0.000000\n",
      "-  0.025043  0.000000  0.023773\n",
      "{'Variant': '#PARENT#', 'Position': array([129, 121, 544]), 'Alignment Probability': 0.8856621099740113, 'Alignment Count': 103}\n",
      "Variant: NB46/RB01 103 0.8856621099740113\n",
      "Processing RB01 NB49\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        215       57        582\n",
      "A  0.822832  0.006620  0.000000\n",
      "C  0.163105  0.903288  0.060675\n",
      "T  0.000000  0.015730  0.006659\n",
      "G  0.008202  0.000000  0.920098\n",
      "-  0.005860  0.074362  0.012568\n",
      "{'Variant': 'C215A', 'Position': array([215,  57, 582]), 'Alignment Probability': 0.6838665244752651, 'Alignment Count': 117}\n",
      "Variant: NB49/RB01 117 0.6838665244752651\n",
      "Processing RB01 NB77\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "   505       411  644\n",
      "A  0.0  0.000000  NaN\n",
      "C  1.0  0.985471  NaN\n",
      "T  0.0  0.005800  NaN\n",
      "G  0.0  0.008729  NaN\n",
      "-  0.0  0.000000  NaN\n",
      "{'Variant': 'C644A', 'Position': array([505, 411, 644]), 'Alignment Probability': nan, 'Alignment Count': 133}\n",
      "Variant: NB77/RB01 133 nan\n",
      "Processing RB01 NB12\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        225       482       505\n",
      "A  0.028396  0.970007  0.000000\n",
      "C  0.883732  0.000000  0.972066\n",
      "T  0.059178  0.000000  0.027934\n",
      "G  0.000000  0.000000  0.000000\n",
      "-  0.028694  0.029993  0.000000\n",
      "{'Variant': 'T225C', 'Position': array([225, 482, 505]), 'Alignment Probability': 0.8332805414040451, 'Alignment Count': 34}\n",
      "Variant: NB12/RB01 34 0.8332805414040451\n",
      "Processing RB01 NB38\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        202       247     365\n",
      "A  0.456264  0.965953  0.0000\n",
      "C  0.005249  0.005334  0.0000\n",
      "T  0.011946  0.019038  0.9947\n",
      "G  0.484921  0.006455  0.0000\n",
      "-  0.041620  0.003220  0.0053\n",
      "{'Variant': 'A202G', 'Position': array([202, 247, 365]), 'Alignment Probability': 0.4659285482574037, 'Alignment Count': 160}\n",
      "Variant: NB38/RB01 160 0.4659285482574037\n",
      "Processing RB01 NB64\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        108       133\n",
      "A  0.076279  0.040363\n",
      "C  0.000000  0.000000\n",
      "T  0.000000  0.000000\n",
      "G  0.923721  0.959637\n",
      "-  0.000000  0.000000\n",
      "{'Variant': 'A108G_A133G', 'Position': [108, 133], 'Alignment Probability': 0.8864365457790696, 'Alignment Count': 44}\n",
      "Variant: NB64/RB01 44 0.8864365457790696\n",
      "Processing RB01 NB71\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "/home/emre/minION_results/MinION_RBC_0902723_sup/Demultiplex_cpp_70_40k_reads/RB01/NB71/alignment_minimap.bam does not exist.\n",
      "Skipping Variant: NB71/RB01\n",
      "Processing RB01 NB24\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "/home/emre/minION_results/MinION_RBC_0902723_sup/Demultiplex_cpp_70_40k_reads/RB01/NB24/alignment_minimap.bam does not exist.\n",
      "Skipping Variant: NB24/RB01\n",
      "Processing RB01 NB94\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "   245  109  237\n",
      "A  1.0  0.0  0.0\n",
      "C  0.0  1.0  0.0\n",
      "T  0.0  0.0  0.0\n",
      "G  0.0  0.0  1.0\n",
      "-  0.0  0.0  0.0\n",
      "{'Variant': '#PARENT#', 'Position': array([245, 109, 237]), 'Alignment Probability': 1.0, 'Alignment Count': 11}\n",
      "Variant: NB94/RB01 11 1.0\n",
      "Processing RB01 NB50\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        106       461       239\n",
      "A  0.942110  0.000000  0.015384\n",
      "C  0.000000  0.935609  0.000000\n",
      "T  0.020712  0.006290  0.019373\n",
      "G  0.007719  0.000000  0.965243\n",
      "-  0.029459  0.058101  0.000000\n",
      "{'Variant': 'T106A', 'Position': array([106, 461, 239]), 'Alignment Probability': 0.8508104819515392, 'Alignment Count': 124}\n",
      "Variant: NB50/RB01 124 0.8508104819515392\n",
      "Processing RB01 NB65\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        229  439  394\n",
      "A  0.564590  0.0  1.0\n",
      "C  0.000000  0.0  0.0\n",
      "T  0.000000  0.0  0.0\n",
      "G  0.363427  1.0  0.0\n",
      "-  0.071982  0.0  0.0\n",
      "{'Variant': '#PARENT#', 'Position': array([229, 439, 394]), 'Alignment Probability': 0.5645903100972095, 'Alignment Count': 25}\n",
      "Variant: NB65/RB01 25 0.5645903100972095\n",
      "Processing RB01 NB15\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        104  89        520\n",
      "A  0.009974  NaN  0.000000\n",
      "C  0.944144  NaN  0.003213\n",
      "T  0.004957  NaN  0.991246\n",
      "G  0.036784  NaN  0.000000\n",
      "-  0.004141  NaN  0.005541\n",
      "{'Variant': 'T89A', 'Position': array([104,  89, 520]), 'Alignment Probability': nan, 'Alignment Count': 199}\n",
      "Variant: NB15/RB01 199 nan\n",
      "Processing RB01 NB02\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        193       293       271\n",
      "A  0.006075  0.000000  0.000000\n",
      "C  0.000000  0.000000  0.960905\n",
      "T  0.985745  0.967997  0.000000\n",
      "G  0.000000  0.008555  0.007602\n",
      "-  0.008179  0.023447  0.031493\n",
      "{'Variant': '#PARENT#', 'Position': array([193, 293, 271]), 'Alignment Probability': 0.9168948579121544, 'Alignment Count': 115}\n",
      "Variant: NB02/RB01 115 0.9168948579121544\n",
      "Processing RB01 NB27\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        202       173       150\n",
      "A  0.929183  0.007453  0.013928\n",
      "C  0.000000  0.000000  0.000000\n",
      "T  0.000000  0.000000  0.000000\n",
      "G  0.034840  0.992547  0.978877\n",
      "-  0.035977  0.000000  0.007195\n",
      "{'Variant': '#PARENT#', 'Position': array([202, 173, 150]), 'Alignment Probability': 0.9027772748837862, 'Alignment Count': 137}\n",
      "Variant: NB27/RB01 137 0.9027772748837862\n",
      "Processing RB01 NB08\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "/home/emre/minION_results/MinION_RBC_0902723_sup/Demultiplex_cpp_70_40k_reads/RB01/NB08/alignment_minimap.bam does not exist.\n",
      "Skipping Variant: NB08/RB01\n",
      "Processing RB01 NB91\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "   561  91   370\n",
      "A  0.0  0.0  0.0\n",
      "C  0.0  0.0  0.0\n",
      "T  1.0  1.0  0.0\n",
      "G  0.0  0.0  1.0\n",
      "-  0.0  0.0  0.0\n",
      "{'Variant': '#PARENT#', 'Position': array([561,  91, 370]), 'Alignment Probability': 1.0, 'Alignment Count': 24}\n",
      "Variant: NB91/RB01 24 1.0\n",
      "Processing RB01 NB19\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "   368  377  502\n",
      "A  0.0  0.0  0.0\n",
      "C  1.0  0.0  1.0\n",
      "T  0.0  0.0  0.0\n",
      "G  0.0  1.0  0.0\n",
      "-  0.0  0.0  0.0\n",
      "{'Variant': '#PARENT#', 'Position': array([368, 377, 502]), 'Alignment Probability': 1.0, 'Alignment Count': 33}\n",
      "Variant: NB19/RB01 33 1.0\n",
      "Processing RB01 NB51\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        478       638       518\n",
      "A  0.000000  0.008530  0.000000\n",
      "C  0.000000  0.930456  0.009413\n",
      "T  0.992188  0.028053  0.968596\n",
      "G  0.000000  0.000000  0.008721\n",
      "-  0.007812  0.032962  0.013270\n",
      "{'Variant': '#PARENT#', 'Position': array([478, 638, 518]), 'Alignment Probability': 0.8941951558875545, 'Alignment Count': 108}\n",
      "Variant: NB51/RB01 108 0.8941951558875545\n",
      "Processing RB01 NB52\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        249       427       312\n",
      "A  0.007515  0.000000  0.000000\n",
      "C  0.028094  0.000000  0.979443\n",
      "T  0.007507  0.011041  0.020557\n",
      "G  0.956884  0.988959  0.000000\n",
      "-  0.000000  0.000000  0.000000\n",
      "{'Variant': '#PARENT#', 'Position': array([249, 427, 312]), 'Alignment Probability': 0.9268648403667199, 'Alignment Count': 133}\n",
      "Variant: NB52/RB01 133 0.9268648403667199\n",
      "Processing RB01 NB35\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        152       545  459\n",
      "A  0.000000  0.918091  1.0\n",
      "C  0.677812  0.000000  0.0\n",
      "T  0.292637  0.000000  0.0\n",
      "G  0.000000  0.000000  0.0\n",
      "-  0.029551  0.081909  0.0\n",
      "{'Variant': 'T152C', 'Position': array([152, 545, 459]), 'Alignment Probability': 0.6222933679995788, 'Alignment Count': 62}\n",
      "Variant: NB35/RB01 62 0.6222933679995788\n",
      "Processing RB01 NB09\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "/home/emre/minION_results/MinION_RBC_0902723_sup/Demultiplex_cpp_70_40k_reads/RB01/NB09/alignment_minimap.bam does not exist.\n",
      "Skipping Variant: NB09/RB01\n",
      "Processing RB01 NB11\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        225       276      166\n",
      "A  0.000000  0.022404  0.00000\n",
      "C  0.890354  0.947908  0.00814\n",
      "T  0.028245  0.000000  0.98307\n",
      "G  0.000000  0.007894  0.00000\n",
      "-  0.081402  0.021794  0.00879\n",
      "{'Variant': 'T225C', 'Position': array([225, 276, 166]), 'Alignment Probability': 0.82968464333635, 'Alignment Count': 113}\n",
      "Variant: NB11/RB01 113 0.82968464333635\n",
      "Processing RB01 NB63\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        548       122  124\n",
      "A  0.000000  0.000000  1.0\n",
      "C  0.987224  0.015411  0.0\n",
      "T  0.000000  0.984589  0.0\n",
      "G  0.000000  0.000000  0.0\n",
      "-  0.012776  0.000000  0.0\n",
      "{'Variant': '#PARENT#', 'Position': array([548, 122, 124]), 'Alignment Probability': 0.9720102966164333, 'Alignment Count': 64}\n",
      "Variant: NB63/RB01 64 0.9720102966164333\n",
      "Processing RB01 NB42\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        107      222  353\n",
      "A  0.007404  0.00000  0.0\n",
      "C  0.978173  0.00000  1.0\n",
      "T  0.007223  0.97866  0.0\n",
      "G  0.000000  0.00000  0.0\n",
      "-  0.007200  0.02134  0.0\n",
      "{'Variant': '#PARENT#', 'Position': array([107, 222, 353]), 'Alignment Probability': 0.9572985923494784, 'Alignment Count': 135}\n",
      "Variant: NB42/RB01 135 0.9572985923494784\n",
      "Processing RB01 NB18\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        93        259       201\n",
      "A  0.021951  0.016590  0.008841\n",
      "C  0.003489  0.004559  0.938549\n",
      "T  0.008549  0.933415  0.013523\n",
      "G  0.899876  0.003603  0.000000\n",
      "-  0.066135  0.041833  0.039086\n",
      "{'Variant': '#PARENT#', 'Position': array([ 93, 259, 201]), 'Alignment Probability': 0.7883423136059394, 'Alignment Count': 214}\n",
      "Variant: NB18/RB01 214 0.7883423136059394\n",
      "Processing RB01 NB90\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        404       187  469\n",
      "A  0.000000  0.965341  0.0\n",
      "C  0.000000  0.000000  0.0\n",
      "T  0.947688  0.000000  0.0\n",
      "G  0.000000  0.000000  1.0\n",
      "-  0.052312  0.034659  0.0\n",
      "{'Variant': '#PARENT#', 'Position': array([404, 187, 469]), 'Alignment Probability': 0.9148419252001637, 'Alignment Count': 16}\n",
      "Variant: NB90/RB01 16 0.9148419252001637\n",
      "Processing RB01 NB05\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "       157       176       206  642  643  644\n",
      "A  0.48537  0.533262  0.000000  NaN  NaN  NaN\n",
      "C  0.00000  0.014375  0.423998  NaN  NaN  NaN\n",
      "T  0.00000  0.000000  0.570018  NaN  NaN  NaN\n",
      "G  0.51463  0.420232  0.005984  NaN  NaN  NaN\n",
      "-  0.00000  0.032130  0.000000  NaN  NaN  NaN\n",
      "{'Variant': 'A157G_C642A_C644A', 'Position': [157, 176, 206, 642, 643, 644], 'Alignment Probability': nan, 'Alignment Count': 135}\n",
      "Variant: NB05/RB01 135 nan\n",
      "Processing RB01 NB39\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "   233  235       251\n",
      "A  NaN  NaN  0.009473\n",
      "C  NaN  NaN  0.000000\n",
      "T  NaN  NaN  0.000000\n",
      "G  NaN  NaN  0.974505\n",
      "-  NaN  NaN  0.016023\n",
      "{'Variant': 'T233A_G235A', 'Position': array([233, 235, 251]), 'Alignment Probability': nan, 'Alignment Count': 108}\n",
      "Variant: NB39/RB01 108 nan\n",
      "Processing RB01 NB69\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "   529       91   648\n",
      "A  NaN  0.000000  0.0\n",
      "C  NaN  0.011937  1.0\n",
      "T  NaN  0.973996  0.0\n",
      "G  NaN  0.002643  0.0\n",
      "-  NaN  0.011425  0.0\n",
      "{'Variant': '#PARENT#', 'Position': array([529,  91, 648]), 'Alignment Probability': nan, 'Alignment Count': 205}\n",
      "Variant: NB69/RB01 205 nan\n",
      "Processing RB01 NB25\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        513  423       542\n",
      "A  0.000000  1.0  0.043274\n",
      "C  0.981028  0.0  0.000000\n",
      "T  0.010021  0.0  0.000000\n",
      "G  0.000000  0.0  0.945601\n",
      "-  0.008950  0.0  0.011126\n",
      "{'Variant': '#PARENT#', 'Position': array([513, 423, 542]), 'Alignment Probability': 0.927661136762727, 'Alignment Count': 93}\n",
      "Variant: NB25/RB01 93 0.927661136762727\n",
      "Processing RB02 NB87\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        94        171\n",
      "A  0.028645  0.000000\n",
      "C  0.000000  0.928471\n",
      "T  0.000000  0.043793\n",
      "G  0.926857  0.000000\n",
      "-  0.044499  0.027736\n",
      "{'Variant': 'A94G_T171C', 'Position': [94, 171], 'Alignment Probability': 0.8605598260020175, 'Alignment Count': 127}\n",
      "Variant: NB87/RB02 127 0.8605598260020175\n",
      "Processing RB02 NB03\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        76        144\n",
      "A  0.005517  0.000000\n",
      "C  0.000000  0.000000\n",
      "T  0.000000  0.947616\n",
      "G  0.994483  0.000000\n",
      "-  0.000000  0.052384\n",
      "{'Variant': 'A76G_A144T', 'Position': [76, 144], 'Alignment Probability': 0.94238782083488, 'Alignment Count': 114}\n",
      "Variant: NB03/RB02 114 0.94238782083488\n",
      "Processing RB02 NB20\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        130  187\n",
      "A  0.000000  0.0\n",
      "C  0.931137  0.0\n",
      "T  0.000000  0.0\n",
      "G  0.000000  1.0\n",
      "-  0.068863  0.0\n",
      "{'Variant': 'T130C_A187G', 'Position': [130, 187], 'Alignment Probability': 0.9311372067032416, 'Alignment Count': 12}\n",
      "Variant: NB20/RB02 12 0.9311372067032416\n",
      "Processing RB02 NB48\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        444       572       583\n",
      "A  0.000000  0.000000  0.008040\n",
      "C  0.922339  0.000000  0.004371\n",
      "T  0.025814  0.008309  0.937611\n",
      "G  0.015055  0.991691  0.021220\n",
      "-  0.036792  0.000000  0.028758\n",
      "{'Variant': '#PARENT#', 'Position': array([444, 572, 583]), 'Alignment Probability': 0.857609288262638, 'Alignment Count': 117}\n",
      "Variant: NB48/RB02 117 0.857609288262638\n",
      "Processing RB02 NB70\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        245       555       253\n",
      "A  0.034870  0.000000  0.995999\n",
      "C  0.005071  0.000000  0.000000\n",
      "T  0.025541  0.000000  0.000000\n",
      "G  0.007879  0.976935  0.000000\n",
      "-  0.926639  0.023065  0.004001\n",
      "{'Variant': 'A245DEL', 'Position': array([245, 555, 253]), 'Alignment Probability': 0.9016441736731773, 'Alignment Count': 127}\n",
      "Variant: NB70/RB02 127 0.9016441736731773\n",
      "Processing RB02 NB04\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        80        373       285\n",
      "A  0.914990  0.980812  0.000000\n",
      "C  0.000000  0.000000  0.000000\n",
      "T  0.000000  0.019188  0.000000\n",
      "G  0.029121  0.000000  0.970496\n",
      "-  0.055889  0.000000  0.029504\n",
      "{'Variant': 'T80A', 'Position': array([ 80, 373, 285]), 'Alignment Probability': 0.8709551623222926, 'Alignment Count': 34}\n",
      "Variant: NB04/RB02 34 0.8709551623222926\n",
      "Processing RB02 NB37\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        217       233\n",
      "A  0.000000  0.370030\n",
      "C  0.009073  0.000000\n",
      "T  0.591902  0.594534\n",
      "G  0.008469  0.009107\n",
      "-  0.390556  0.026329\n",
      "{'Variant': '#PARENT#', 'Position': [217, 233], 'Alignment Probability': 0.3519055277233461, 'Alignment Count': 113}\n",
      "Variant: NB37/RB02 113 0.3519055277233461\n",
      "Processing RB02 NB88\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        111       628       235\n",
      "A  0.382114  0.986137  0.000000\n",
      "C  0.000000  0.000000  0.013331\n",
      "T  0.000000  0.000000  0.028908\n",
      "G  0.617886  0.013863  0.809587\n",
      "-  0.000000  0.000000  0.148174\n",
      "{'Variant': '#PARENT#', 'Position': array([111, 628, 235]), 'Alignment Probability': 0.4932978330447302, 'Alignment Count': 72}\n",
      "Variant: NB88/RB02 72 0.4932978330447302\n",
      "Processing RB02 NB14\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        97        121       196\n",
      "A  0.280544  0.029141  0.321497\n",
      "C  0.010468  0.628204  0.000000\n",
      "T  0.692570  0.342655  0.000000\n",
      "G  0.008235  0.000000  0.678503\n",
      "-  0.008184  0.000000  0.000000\n",
      "{'Variant': 'A97T_T121C_A196G', 'Position': [97, 121, 196], 'Alignment Probability': 0.29519967993887286, 'Alignment Count': 89}\n",
      "Variant: NB14/RB02 89 0.29519967993887286\n",
      "Processing RB02 NB82\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        517  661       481\n",
      "A  0.000000  NaN  0.010984\n",
      "C  0.015093  NaN  0.989016\n",
      "T  0.970115  NaN  0.000000\n",
      "G  0.000000  NaN  0.000000\n",
      "-  0.014792  NaN  0.000000\n",
      "{'Variant': 'G661A', 'Position': array([517, 661, 481]), 'Alignment Probability': nan, 'Alignment Count': 67}\n",
      "Variant: NB82/RB02 67 nan\n",
      "Processing RB02 NB54\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        211       212  55 \n",
      "A  0.983153  0.017362  NaN\n",
      "C  0.000000  0.000000  NaN\n",
      "T  0.007325  0.000000  NaN\n",
      "G  0.000000  0.982638  NaN\n",
      "-  0.009522  0.000000  NaN\n",
      "{'Variant': 'C55A', 'Position': array([211, 212,  55]), 'Alignment Probability': nan, 'Alignment Count': 97}\n",
      "Variant: NB54/RB02 97 nan\n",
      "Processing RB02 NB57\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        116       403  352\n",
      "A  0.000000  0.000000  0.0\n",
      "C  0.888091  0.038115  1.0\n",
      "T  0.075353  0.961885  0.0\n",
      "G  0.000000  0.000000  0.0\n",
      "-  0.036556  0.000000  0.0\n",
      "{'Variant': 'T116C', 'Position': array([116, 403, 352]), 'Alignment Probability': 0.8542415426383895, 'Alignment Count': 27}\n",
      "Variant: NB57/RB02 27 0.8542415426383895\n",
      "Processing RB02 NB28\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        341       492       441\n",
      "A  0.000000  0.000000  0.065192\n",
      "C  0.000000  0.990522  0.011294\n",
      "T  0.991882  0.009478  0.034669\n",
      "G  0.000000  0.000000  0.888845\n",
      "-  0.008118  0.000000  0.000000\n",
      "{'Variant': '#PARENT#', 'Position': array([341, 492, 441]), 'Alignment Probability': 0.8732736788744832, 'Alignment Count': 90}\n",
      "Variant: NB28/RB02 90 0.8732736788744832\n",
      "Processing RB02 NB40\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "/home/emre/minION_results/MinION_RBC_0902723_sup/Demultiplex_cpp_70_40k_reads/RB02/NB40/alignment_minimap.bam does not exist.\n",
      "Skipping Variant: NB40/RB02\n",
      "Processing RB02 NB81\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        110       446  229\n",
      "A  0.000000  0.005836  1.0\n",
      "C  0.019305  0.929371  0.0\n",
      "T  0.967826  0.020615  0.0\n",
      "G  0.006417  0.011790  0.0\n",
      "-  0.006452  0.032389  0.0\n",
      "{'Variant': 'G110T', 'Position': array([110, 446, 229]), 'Alignment Probability': 0.899469000590912, 'Alignment Count': 147}\n",
      "Variant: NB81/RB02 147 0.899469000590912\n",
      "Processing RB02 NB13\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        218  306       660\n",
      "A  0.000000  0.0  0.000000\n",
      "C  0.637416  1.0  0.014282\n",
      "T  0.348227  0.0  0.985718\n",
      "G  0.014357  0.0  0.000000\n",
      "-  0.000000  0.0  0.000000\n",
      "{'Variant': 'T218C', 'Position': array([218, 306, 660]), 'Alignment Probability': 0.6283123605259014, 'Alignment Count': 70}\n",
      "Variant: NB13/RB02 70 0.6283123605259014\n",
      "Processing RB02 NB53\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        418       360       286\n",
      "A  0.016828  0.000000  0.000000\n",
      "C  0.956207  0.000000  0.008564\n",
      "T  0.007191  0.000000  0.000000\n",
      "G  0.019774  0.994794  0.991436\n",
      "-  0.000000  0.005206  0.000000\n",
      "{'Variant': '#PARENT#', 'Position': array([418, 360, 286]), 'Alignment Probability': 0.9430828664591961, 'Alignment Count': 104}\n",
      "Variant: NB53/RB02 104 0.9430828664591961\n",
      "Processing RB02 NB47\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        234       194       633\n",
      "A  0.005735  0.008717  0.000000\n",
      "C  0.023469  0.000000  0.021058\n",
      "T  0.544101  0.000000  0.971432\n",
      "G  0.331213  0.991283  0.000000\n",
      "-  0.095482  0.000000  0.007511\n",
      "{'Variant': '#PARENT#', 'Position': array([234, 194, 633]), 'Alignment Probability': 0.5239493562902139, 'Alignment Count': 117}\n",
      "Variant: NB47/RB02 117 0.5239493562902139\n",
      "Processing RB02 NB84\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        89        193\n",
      "A  0.000000  0.000000\n",
      "C  0.642769  0.677922\n",
      "T  0.309526  0.302150\n",
      "G  0.012576  0.000000\n",
      "-  0.035129  0.019928\n",
      "{'Variant': 'T89C_T193C', 'Position': [89, 193], 'Alignment Probability': 0.4357473378229533, 'Alignment Count': 80}\n",
      "Variant: NB84/RB02 80 0.4357473378229533\n",
      "Processing RB02 NB72\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        227       460  124\n",
      "A  0.952806  0.000000  1.0\n",
      "C  0.000000  0.000000  0.0\n",
      "T  0.014026  0.974162  0.0\n",
      "G  0.009981  0.000000  0.0\n",
      "-  0.023187  0.025838  0.0\n",
      "{'Variant': 'G227A', 'Position': array([227, 460, 124]), 'Alignment Probability': 0.9281871315222563, 'Alignment Count': 73}\n",
      "Variant: NB72/RB02 73 0.9281871315222563\n",
      "Processing RB02 NB10\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        126       338       614\n",
      "A  0.000000  0.035434  0.000000\n",
      "C  0.038360  0.000000  0.000000\n",
      "T  0.949585  0.012513  0.988039\n",
      "G  0.000000  0.952053  0.011961\n",
      "-  0.012055  0.000000  0.000000\n",
      "{'Variant': 'C126T', 'Position': array([126, 338, 614]), 'Alignment Probability': 0.8932416075630207, 'Alignment Count': 77}\n",
      "Variant: NB10/RB02 77 0.8932416075630207\n",
      "Processing RB02 NB58\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        83   473       168\n",
      "A  0.000000  0.0  0.000000\n",
      "C  0.420332  0.0  0.000000\n",
      "T  0.564536  1.0  0.943522\n",
      "G  0.000000  0.0  0.000000\n",
      "-  0.015132  0.0  0.056478\n",
      "{'Variant': '#PARENT#', 'Position': array([ 83, 473, 168]), 'Alignment Probability': 0.5326516731798581, 'Alignment Count': 65}\n",
      "Variant: NB58/RB02 65 0.5326516731798581\n",
      "Processing RB02 NB79\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "   242  205  465\n",
      "A  1.0  0.0  0.0\n",
      "C  0.0  0.0  1.0\n",
      "T  0.0  1.0  0.0\n",
      "G  0.0  0.0  0.0\n",
      "-  0.0  0.0  0.0\n",
      "{'Variant': '#PARENT#', 'Position': array([242, 205, 465]), 'Alignment Probability': 1.0, 'Alignment Count': 46}\n",
      "Variant: NB79/RB02 46 1.0\n",
      "Processing RB02 NB36\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        308       599  251\n",
      "A  0.000000  0.000000  NaN\n",
      "C  0.000000  0.968644  NaN\n",
      "T  0.000000  0.004819  NaN\n",
      "G  0.976597  0.015300  NaN\n",
      "-  0.023403  0.011237  NaN\n",
      "{'Variant': 'G251A', 'Position': array([308, 599, 251]), 'Alignment Probability': nan, 'Alignment Count': 174}\n",
      "Variant: NB36/RB02 174 nan\n",
      "Processing RB02 NB30\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        87        228\n",
      "A  0.000000  0.000000\n",
      "C  0.000000  0.949823\n",
      "T  0.922231  0.000000\n",
      "G  0.000000  0.050177\n",
      "-  0.077769  0.000000\n",
      "{'Variant': 'C87T_G228C', 'Position': [87, 228], 'Alignment Probability': 0.8759557633631465, 'Alignment Count': 18}\n",
      "Variant: NB30/RB02 18 0.8759557633631465\n",
      "Processing RB02 NB60\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        388       542       363\n",
      "A  0.956902  0.025964  0.000000\n",
      "C  0.007745  0.000000  0.004506\n",
      "T  0.000000  0.000000  0.988445\n",
      "G  0.035354  0.929849  0.000000\n",
      "-  0.000000  0.044187  0.007049\n",
      "{'Variant': '#PARENT#', 'Position': array([388, 542, 363]), 'Alignment Probability': 0.8794928547653156, 'Alignment Count': 225}\n",
      "Variant: NB60/RB02 225 0.8794928547653156\n",
      "Processing RB02 NB45\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "   175       590       516\n",
      "A  1.0  0.939262  0.000000\n",
      "C  0.0  0.000000  0.000000\n",
      "T  0.0  0.000000  0.948207\n",
      "G  0.0  0.000000  0.022182\n",
      "-  0.0  0.060738  0.029610\n",
      "{'Variant': '#PARENT#', 'Position': array([175, 590, 516]), 'Alignment Probability': 0.8906154549472924, 'Alignment Count': 33}\n",
      "Variant: NB45/RB02 33 0.8906154549472924\n",
      "Processing RB02 NB93\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        319       225       150\n",
      "A  0.000000  0.006525  0.000000\n",
      "C  0.000000  0.008388  0.000000\n",
      "T  0.006680  0.976508  0.000000\n",
      "G  0.969663  0.000000  0.983777\n",
      "-  0.023657  0.008579  0.016223\n",
      "{'Variant': '#PARENT#', 'Position': array([319, 225, 150]), 'Alignment Probability': 0.9315227282684947, 'Alignment Count': 118}\n",
      "Variant: NB93/RB02 118 0.9315227282684947\n",
      "Processing RB02 NB31\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "/home/emre/minION_results/MinION_RBC_0902723_sup/Demultiplex_cpp_70_40k_reads/RB02/NB31/alignment_minimap.bam does not exist.\n",
      "Skipping Variant: NB31/RB02\n",
      "Processing RB02 NB75\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        207       116     437\n",
      "A  0.000000  0.000000  0.0097\n",
      "C  0.000000  0.014657  0.9903\n",
      "T  0.699346  0.975145  0.0000\n",
      "G  0.259465  0.000000  0.0000\n",
      "-  0.041189  0.010198  0.0000\n",
      "{'Variant': 'G207T', 'Position': array([207, 116, 437]), 'Alignment Probability': 0.6753488917240199, 'Alignment Count': 97}\n",
      "Variant: NB75/RB02 97 0.6753488917240199\n",
      "Processing RB02 NB21\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        175       211\n",
      "A  0.029526  0.048494\n",
      "C  0.000000  0.025437\n",
      "T  0.000000  0.000000\n",
      "G  0.970474  0.926069\n",
      "-  0.000000  0.000000\n",
      "{'Variant': 'A175G_A211G', 'Position': [175, 211], 'Alignment Probability': 0.8987259837618204, 'Alignment Count': 32}\n",
      "Variant: NB21/RB02 32 0.8987259837618204\n",
      "Processing RB02 NB86\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        389       293       615\n",
      "A  0.006459  0.000000  0.000000\n",
      "C  0.000000  0.000000  0.004454\n",
      "T  0.000000  0.960725  0.984072\n",
      "G  0.993541  0.013332  0.000000\n",
      "-  0.000000  0.025943  0.011474\n",
      "{'Variant': '#PARENT#', 'Position': array([389, 293, 615]), 'Alignment Probability': 0.9393159515532891, 'Alignment Count': 147}\n",
      "Variant: NB86/RB02 147 0.9393159515532891\n",
      "Processing RB02 NB07\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        79        230\n",
      "A  0.423158  0.000000\n",
      "C  0.000000  0.440079\n",
      "T  0.000000  0.510022\n",
      "G  0.576842  0.034766\n",
      "-  0.000000  0.015133\n",
      "{'Variant': 'A79G', 'Position': [79, 230], 'Alignment Probability': 0.29420204736580297, 'Alignment Count': 54}\n",
      "Variant: NB07/RB02 54 0.29420204736580297\n",
      "Processing RB02 NB34\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "   202  467       284\n",
      "A  NaN  NaN  0.967586\n",
      "C  NaN  NaN  0.000000\n",
      "T  NaN  NaN  0.005952\n",
      "G  NaN  NaN  0.010247\n",
      "-  NaN  NaN  0.016215\n",
      "{'Variant': 'T467A', 'Position': array([202, 467, 284]), 'Alignment Probability': nan, 'Alignment Count': 97}\n",
      "Variant: NB34/RB02 97 nan\n",
      "Processing RB02 NB62\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        116       55   110\n",
      "A  0.000000  0.000000  0.0\n",
      "C  0.884286  0.979706  0.0\n",
      "T  0.091800  0.010732  0.0\n",
      "G  0.000000  0.000000  1.0\n",
      "-  0.023914  0.009562  0.0\n",
      "{'Variant': 'T116C', 'Position': array([116,  55, 110]), 'Alignment Probability': 0.8663402625813055, 'Alignment Count': 85}\n",
      "Variant: NB62/RB02 85 0.8663402625813055\n",
      "Processing RB02 NB23\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        529      188  101\n",
      "A  0.986149  0.00000  1.0\n",
      "C  0.000000  0.97526  0.0\n",
      "T  0.000000  0.02474  0.0\n",
      "G  0.007995  0.00000  0.0\n",
      "-  0.005856  0.00000  0.0\n",
      "{'Variant': '#PARENT#', 'Position': array([529, 188, 101]), 'Alignment Probability': 0.961751583054566, 'Alignment Count': 115}\n",
      "Variant: NB23/RB02 115 0.961751583054566\n",
      "Processing RB02 NB92\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        194  287       347\n",
      "A  0.016708  1.0  0.000000\n",
      "C  0.000000  0.0  0.977113\n",
      "T  0.000000  0.0  0.022887\n",
      "G  0.968252  0.0  0.000000\n",
      "-  0.015040  0.0  0.000000\n",
      "{'Variant': '#PARENT#', 'Position': array([194, 287, 347]), 'Alignment Probability': 0.9460908095505614, 'Alignment Count': 61}\n",
      "Variant: NB92/RB02 61 0.9460908095505614\n",
      "Processing RB02 NB73\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        145       446  78 \n",
      "A  0.000000  0.000000  0.0\n",
      "C  0.000000  0.980715  0.0\n",
      "T  0.980315  0.000000  0.0\n",
      "G  0.000000  0.000000  1.0\n",
      "-  0.019685  0.019285  0.0\n",
      "{'Variant': '#PARENT#', 'Position': array([145, 446,  78]), 'Alignment Probability': 0.9614097655363141, 'Alignment Count': 31}\n",
      "Variant: NB73/RB02 31 0.9614097655363141\n",
      "Processing RB02 NB67\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        479  655  505\n",
      "A  0.000000  1.0  0.0\n",
      "C  0.989761  0.0  1.0\n",
      "T  0.010239  0.0  0.0\n",
      "G  0.000000  0.0  0.0\n",
      "-  0.000000  0.0  0.0\n",
      "{'Variant': '#PARENT#', 'Position': array([479, 655, 505]), 'Alignment Probability': 0.9897607264356527, 'Alignment Count': 77}\n",
      "Variant: NB67/RB02 77 0.9897607264356527\n",
      "Processing RB02 NB29\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        168       438\n",
      "A  0.000000  0.000000\n",
      "C  0.364558  0.000000\n",
      "T  0.635442  0.000000\n",
      "G  0.000000  0.665498\n",
      "-  0.000000  0.334502\n",
      "{'Variant': '#PARENT#', 'Position': [168, 438], 'Alignment Probability': 0.4228854795532493, 'Alignment Count': 11}\n",
      "Variant: NB29/RB02 11 0.4228854795532493\n",
      "Processing RB02 NB33\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "/home/emre/minION_results/MinION_RBC_0902723_sup/Demultiplex_cpp_70_40k_reads/RB02/NB33/alignment_minimap.bam does not exist.\n",
      "Skipping Variant: NB33/RB02\n",
      "Processing RB02 NB06\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        75        118\n",
      "A  0.013307  0.011298\n",
      "C  0.968124  0.000000\n",
      "T  0.010363  0.000000\n",
      "G  0.008206  0.963703\n",
      "-  0.000000  0.024999\n",
      "{'Variant': 'T75C_A118G', 'Position': [75, 118], 'Alignment Probability': 0.93298414609271, 'Alignment Count': 75}\n",
      "Variant: NB06/RB02 75 0.93298414609271\n",
      "Processing RB02 NB78\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        225       532       67 \n",
      "A  0.027264  0.013902  0.000000\n",
      "C  0.856371  0.000000  0.008040\n",
      "T  0.007180  0.000000  0.984721\n",
      "G  0.014835  0.977953  0.000000\n",
      "-  0.094350  0.008145  0.007239\n",
      "{'Variant': 'T225C', 'Position': array([225, 532,  67]), 'Alignment Probability': 0.8246941440679628, 'Alignment Count': 126}\n",
      "Variant: NB78/RB02 126 0.8246941440679628\n",
      "Processing RB02 NB95\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "   613       53        438\n",
      "A  0.0  0.017021  0.041219\n",
      "C  1.0  0.019183  0.017897\n",
      "T  0.0  0.015604  0.009758\n",
      "G  0.0  0.937816  0.843133\n",
      "-  0.0  0.010377  0.087993\n",
      "{'Variant': '#PARENT#', 'Position': array([613,  53, 438]), 'Alignment Probability': 0.790703412249465, 'Alignment Count': 98}\n",
      "Variant: NB95/RB02 98 0.790703412249465\n",
      "Processing RB02 NB26\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "   132       213\n",
      "A  NaN  0.598312\n",
      "C  NaN  0.005531\n",
      "T  NaN  0.374339\n",
      "G  NaN  0.005808\n",
      "-  NaN  0.016009\n",
      "{'Variant': '#PARENT#', 'Position': [132, 213], 'Alignment Probability': nan, 'Alignment Count': 175}\n",
      "Variant: NB26/RB02 175 nan\n",
      "Processing RB02 NB76\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "   524       110      341\n",
      "A  0.0  0.000000  0.00000\n",
      "C  0.0  0.000000  0.00000\n",
      "T  1.0  0.000000  0.96993\n",
      "G  0.0  0.979209  0.03007\n",
      "-  0.0  0.020791  0.00000\n",
      "{'Variant': '#PARENT#', 'Position': array([524, 110, 341]), 'Alignment Probability': 0.9497640043958447, 'Alignment Count': 32}\n",
      "Variant: NB76/RB02 32 0.9497640043958447\n",
      "Processing RB02 NB59\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "   508       71        220\n",
      "A  0.0  0.000000  0.985959\n",
      "C  0.0  0.006824  0.000000\n",
      "T  1.0  0.014613  0.000000\n",
      "G  0.0  0.940199  0.008029\n",
      "-  0.0  0.038364  0.006012\n",
      "{'Variant': '#PARENT#', 'Position': array([508,  71, 220]), 'Alignment Probability': 0.9269973728673533, 'Alignment Count': 127}\n",
      "Variant: NB59/RB02 127 0.9269973728673533\n",
      "Processing RB02 NB68\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "   305       117  299\n",
      "A  0.0  0.000000  0.0\n",
      "C  1.0  0.000000  0.0\n",
      "T  0.0  0.000000  0.0\n",
      "G  0.0  0.968294  1.0\n",
      "-  0.0  0.031706  0.0\n",
      "{'Variant': '#PARENT#', 'Position': array([305, 117, 299]), 'Alignment Probability': 0.9682939719287291, 'Alignment Count': 30}\n",
      "Variant: NB68/RB02 30 0.9682939719287291\n",
      "Processing RB02 NB55\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        177       214\n",
      "A  0.072096  0.000000\n",
      "C  0.006986  0.966213\n",
      "T  0.000000  0.012401\n",
      "G  0.912817  0.000000\n",
      "-  0.008101  0.021386\n",
      "{'Variant': 'A177G_T214C', 'Position': [177, 214], 'Alignment Probability': 0.8819756093223535, 'Alignment Count': 123}\n",
      "Variant: NB55/RB02 123 0.8819756093223535\n",
      "Processing RB02 NB74\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "   584  193  89 \n",
      "A  1.0  0.0  0.0\n",
      "C  0.0  0.0  0.0\n",
      "T  0.0  1.0  1.0\n",
      "G  0.0  0.0  0.0\n",
      "-  0.0  0.0  0.0\n",
      "{'Variant': '#PARENT#', 'Position': array([584, 193,  89]), 'Alignment Probability': 1.0, 'Alignment Count': 41}\n",
      "Variant: NB74/RB02 41 1.0\n",
      "Processing RB02 NB89\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        356       139       240\n",
      "A  0.000000  0.000000  0.006231\n",
      "C  0.000000  0.982600  0.000000\n",
      "T  0.005282  0.000000  0.000000\n",
      "G  0.994718  0.004487  0.988280\n",
      "-  0.000000  0.012913  0.005489\n",
      "{'Variant': '#PARENT#', 'Position': array([356, 139, 240]), 'Alignment Probability': 0.9659540892702876, 'Alignment Count': 151}\n",
      "Variant: NB89/RB02 151 0.9659540892702876\n",
      "Processing RB02 NB83\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        56   72       230       417       576\n",
      "A  0.805561  1.0  0.00000  0.000000  0.580703\n",
      "C  0.000000  0.0  0.40518  0.377963  0.000000\n",
      "T  0.000000  0.0  0.59482  0.622037  0.000000\n",
      "G  0.194439  0.0  0.00000  0.000000  0.000000\n",
      "-  0.000000  0.0  0.00000  0.000000  0.419297\n",
      "{'Variant': 'T56A_G72A', 'Position': [56, 72, 230, 417, 576], 'Alignment Probability': 0.17308282863972527, 'Alignment Count': 5}\n",
      "Variant: NB83/RB02 5 0.17308282863972527\n",
      "Processing RB02 NB22\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        97        121       165\n",
      "A  0.209823  0.045963  0.200738\n",
      "C  0.039690  0.731448  0.047130\n",
      "T  0.750487  0.195606  0.000000\n",
      "G  0.000000  0.026983  0.752132\n",
      "-  0.000000  0.000000  0.000000\n",
      "{'Variant': 'A97T_T121C_A165G', 'Position': [97, 121, 165], 'Alignment Probability': 0.4128772195483564, 'Alignment Count': 19}\n",
      "Variant: NB22/RB02 19 0.4128772195483564\n",
      "Processing RB02 NB96\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        126       52        516\n",
      "A  0.000000  0.000000  0.000000\n",
      "C  0.106062  0.029704  0.006177\n",
      "T  0.885947  0.970296  0.965477\n",
      "G  0.000000  0.000000  0.000000\n",
      "-  0.007991  0.000000  0.028346\n",
      "{'Variant': 'C126T', 'Position': array([126,  52, 516]), 'Alignment Probability': 0.8299543992899246, 'Alignment Count': 96}\n",
      "Variant: NB96/RB02 96 0.8299543992899246\n",
      "Processing RB02 NB32\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "   361  156       202\n",
      "A  1.0  1.0  0.943276\n",
      "C  0.0  0.0  0.000000\n",
      "T  0.0  0.0  0.000000\n",
      "G  0.0  0.0  0.056724\n",
      "-  0.0  0.0  0.000000\n",
      "{'Variant': '#PARENT#', 'Position': array([361, 156, 202]), 'Alignment Probability': 0.9432763941659263, 'Alignment Count': 82}\n",
      "Variant: NB32/RB02 82 0.9432763941659263\n",
      "Processing RB02 NB66\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        136       197       223\n",
      "A  0.020088  0.036089  0.965662\n",
      "C  0.000000  0.000000  0.000000\n",
      "T  0.000000  0.008827  0.019375\n",
      "G  0.969862  0.955084  0.008017\n",
      "-  0.010050  0.000000  0.006946\n",
      "{'Variant': 'A136G_A197G_T223A', 'Position': [136, 197, 223], 'Alignment Probability': 0.8944917083782877, 'Alignment Count': 97}\n",
      "Variant: NB66/RB02 97 0.8944917083782877\n",
      "Processing RB02 NB61\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        291       372       357\n",
      "A  0.000000  0.000000  0.000000\n",
      "C  0.992113  0.015259  0.950809\n",
      "T  0.000000  0.007844  0.027306\n",
      "G  0.000000  0.976897  0.000000\n",
      "-  0.007887  0.000000  0.021885\n",
      "{'Variant': '#PARENT#', 'Position': array([291, 372, 357]), 'Alignment Probability': 0.9215161422895276, 'Alignment Count': 126}\n",
      "Variant: NB61/RB02 126 0.9215161422895276\n",
      "Processing RB02 NB56\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        75        116       178\n",
      "A  0.000000  0.005539  0.035525\n",
      "C  0.955158  0.892542  0.007204\n",
      "T  0.006635  0.077386  0.957271\n",
      "G  0.026422  0.000000  0.000000\n",
      "-  0.011784  0.024533  0.000000\n",
      "{'Variant': 'T75C_T116C_A178T', 'Position': [75, 116, 178], 'Alignment Probability': 0.816091434849398, 'Alignment Count': 128}\n",
      "Variant: NB56/RB02 128 0.816091434849398\n",
      "Processing RB02 NB46\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "   316       605       528\n",
      "A  0.0  0.000000  0.949076\n",
      "C  0.0  0.033403  0.000000\n",
      "T  1.0  0.000000  0.000000\n",
      "G  0.0  0.966597  0.010113\n",
      "-  0.0  0.000000  0.040811\n",
      "{'Variant': '#PARENT#', 'Position': array([316, 605, 528]), 'Alignment Probability': 0.9173742323381847, 'Alignment Count': 78}\n",
      "Variant: NB46/RB02 78 0.9173742323381847\n",
      "Processing RB02 NB49\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        148       185\n",
      "A  0.013852  0.020104\n",
      "C  0.886359  0.854526\n",
      "T  0.091491  0.071047\n",
      "G  0.000000  0.005433\n",
      "-  0.008299  0.048890\n",
      "{'Variant': 'T148C_T185C', 'Position': [148, 185], 'Alignment Probability': 0.7574168619276567, 'Alignment Count': 190}\n",
      "Variant: NB49/RB02 190 0.7574168619276567\n",
      "Processing RB02 NB77\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        183       388       298\n",
      "A  0.962055  0.977484  0.975943\n",
      "C  0.029069  0.022516  0.000000\n",
      "T  0.000000  0.000000  0.000000\n",
      "G  0.000000  0.000000  0.000000\n",
      "-  0.008877  0.000000  0.024057\n",
      "{'Variant': 'C183A', 'Position': array([183, 388, 298]), 'Alignment Probability': 0.9177693125225717, 'Alignment Count': 69}\n",
      "Variant: NB77/RB02 69 0.9177693125225717\n",
      "Processing RB02 NB12\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "       162       183       199  228  229  237       356       453       483  \\\n",
      "A  0.00000  0.000000  0.631856  0.0  0.0  0.0  0.000000  0.000000  0.187756   \n",
      "C  0.00000  0.126363  0.000000  0.0  0.0  1.0  0.808319  0.651068  0.000000   \n",
      "T  0.00000  0.873637  0.000000  0.0  0.0  0.0  0.000000  0.000000  0.812244   \n",
      "G  0.68281  0.000000  0.368144  0.0  0.0  0.0  0.191681  0.000000  0.000000   \n",
      "-  0.31719  0.000000  0.000000  1.0  1.0  0.0  0.000000  0.348932  0.000000   \n",
      "\n",
      "       484       485       526       608  643  \n",
      "A  0.00000  0.000000  0.633348  0.601267  NaN  \n",
      "C  0.00000  0.000000  0.366652  0.000000  NaN  \n",
      "T  0.18716  0.188021  0.000000  0.000000  NaN  \n",
      "G  0.81284  0.811979  0.000000  0.398733  NaN  \n",
      "-  0.00000  0.000000  0.000000  0.000000  NaN  \n",
      "{'Variant': 'C183T_G228DEL_A229DEL_G237C_G356C_A483T_A484G_T485G', 'Position': [162, 183, 199, 228, 229, 237, 356, 453, 483, 484, 485, 526, 608, 643], 'Alignment Probability': nan, 'Alignment Count': 5}\n",
      "Variant: NB12/RB02 5 nan\n",
      "Processing RB02 NB38\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "   650       115       318\n",
      "A  0.0  0.000000  0.000000\n",
      "C  1.0  0.006829  0.979909\n",
      "T  0.0  0.972246  0.000000\n",
      "G  0.0  0.000000  0.000000\n",
      "-  0.0  0.020924  0.020091\n",
      "{'Variant': '#PARENT#', 'Position': array([650, 115, 318]), 'Alignment Probability': 0.9527122440695232, 'Alignment Count': 138}\n",
      "Variant: NB38/RB02 138 0.9527122440695232\n",
      "Processing RB02 NB64\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        92        105       122  152       235       596\n",
      "A  0.000000  0.689255  0.000000  0.0  0.000000  0.000000\n",
      "C  0.653474  0.000000  0.318591  1.0  0.155209  0.331919\n",
      "T  0.000000  0.000000  0.681409  0.0  0.000000  0.668081\n",
      "G  0.000000  0.310745  0.000000  0.0  0.506928  0.000000\n",
      "-  0.346526  0.000000  0.000000  0.0  0.337863  0.000000\n",
      "{'Variant': 'T152C', 'Position': [92, 105, 122, 152, 235, 596], 'Alignment Probability': 0.10394223221723166, 'Alignment Count': 6}\n",
      "Variant: NB64/RB02 6 0.10394223221723166\n",
      "Processing RB02 NB71\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        124  647       282\n",
      "A  0.012206  0.0  0.946094\n",
      "C  0.000000  1.0  0.030899\n",
      "T  0.915410  0.0  0.006633\n",
      "G  0.000000  0.0  0.011844\n",
      "-  0.072384  0.0  0.004531\n",
      "{'Variant': 'A124T', 'Position': array([124, 647, 282]), 'Alignment Probability': 0.8660635795742998, 'Alignment Count': 136}\n",
      "Variant: NB71/RB02 136 0.8660635795742998\n",
      "Processing RB02 NB24\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        156       160       176       223\n",
      "A  0.078691  0.017307  0.000000  0.966665\n",
      "C  0.000000  0.000000  0.018150  0.000000\n",
      "T  0.000000  0.967733  0.000000  0.015017\n",
      "G  0.921309  0.000000  0.960654  0.018318\n",
      "-  0.000000  0.014960  0.021196  0.000000\n",
      "{'Variant': 'A156G_A160T_A176G_T223A', 'Position': [156, 160, 176, 223], 'Alignment Probability': 0.8279498090329627, 'Alignment Count': 47}\n",
      "Variant: NB24/RB02 47 0.8279498090329627\n",
      "Processing RB02 NB94\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        148       131       211\n",
      "A  0.000000  0.012647  0.976868\n",
      "C  0.871159  0.013346  0.000000\n",
      "T  0.128841  0.000000  0.000000\n",
      "G  0.000000  0.950793  0.000000\n",
      "-  0.000000  0.023214  0.023132\n",
      "{'Variant': 'T148C', 'Position': array([148, 131, 211]), 'Alignment Probability': 0.8091317454172705, 'Alignment Count': 77}\n",
      "Variant: NB94/RB02 77 0.8091317454172705\n",
      "Processing RB02 NB50\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        454  378       432\n",
      "A  0.009524  0.0  0.983459\n",
      "C  0.934356  0.0  0.000000\n",
      "T  0.005493  1.0  0.000000\n",
      "G  0.006447  0.0  0.000000\n",
      "-  0.044180  0.0  0.016541\n",
      "{'Variant': '#PARENT#', 'Position': array([454, 378, 432]), 'Alignment Probability': 0.9189012894133333, 'Alignment Count': 156}\n",
      "Variant: NB50/RB02 156 0.9189012894133333\n",
      "Processing RB02 NB65\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        203       523  628\n",
      "A  0.237367  0.000000  1.0\n",
      "C  0.000000  0.975171  0.0\n",
      "T  0.000000  0.024829  0.0\n",
      "G  0.762633  0.000000  0.0\n",
      "-  0.000000  0.000000  0.0\n",
      "{'Variant': 'A203G', 'Position': array([203, 523, 628]), 'Alignment Probability': 0.7436981799089777, 'Alignment Count': 38}\n",
      "Variant: NB65/RB02 38 0.7436981799089777\n",
      "Processing RB02 NB15\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "   469       580  296\n",
      "A  0.0  0.000000  0.0\n",
      "C  0.0  0.972917  0.0\n",
      "T  0.0  0.000000  1.0\n",
      "G  1.0  0.027083  0.0\n",
      "-  0.0  0.000000  0.0\n",
      "{'Variant': '#PARENT#', 'Position': array([469, 580, 296]), 'Alignment Probability': 0.9729170923344866, 'Alignment Count': 37}\n",
      "Variant: NB15/RB02 37 0.9729170923344866\n",
      "Processing RB02 NB02\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        408       489  241\n",
      "A  0.062507  0.000000  0.0\n",
      "C  0.937493  0.000000  1.0\n",
      "T  0.000000  0.953389  0.0\n",
      "G  0.000000  0.000000  0.0\n",
      "-  0.000000  0.046611  0.0\n",
      "{'Variant': '#PARENT#', 'Position': array([408, 489, 241]), 'Alignment Probability': 0.8937953904648039, 'Alignment Count': 15}\n",
      "Variant: NB02/RB02 15 0.8937953904648039\n",
      "Processing RB02 NB27\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        159       84        531\n",
      "A  0.000000  0.000000  0.022085\n",
      "C  0.961873  0.014055  0.000000\n",
      "T  0.007852  0.004763  0.000000\n",
      "G  0.030275  0.973735  0.972264\n",
      "-  0.000000  0.007447  0.005651\n",
      "{'Variant': 'G159C', 'Position': array([159,  84, 531]), 'Alignment Probability': 0.9106316873761948, 'Alignment Count': 121}\n",
      "Variant: NB27/RB02 121 0.9106316873761948\n",
      "Processing RB02 NB08\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        89        208\n",
      "A  0.000000  0.688331\n",
      "C  0.630643  0.000000\n",
      "T  0.369357  0.000000\n",
      "G  0.000000  0.311669\n",
      "-  0.000000  0.000000\n",
      "{'Variant': 'T89C', 'Position': [89, 208], 'Alignment Probability': 0.43409096562251887, 'Alignment Count': 16}\n",
      "Variant: NB08/RB02 16 0.43409096562251887\n",
      "Processing RB02 NB91\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "   532  274       73 \n",
      "A  0.0  1.0  0.011048\n",
      "C  0.0  0.0  0.000000\n",
      "T  0.0  0.0  0.000000\n",
      "G  1.0  0.0  0.975400\n",
      "-  0.0  0.0  0.013552\n",
      "{'Variant': '#PARENT#', 'Position': array([532, 274,  73]), 'Alignment Probability': 0.9754000826957329, 'Alignment Count': 84}\n",
      "Variant: NB91/RB02 84 0.9754000826957329\n",
      "Processing RB02 NB19\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        235  432  309\n",
      "A  0.000000  1.0  0.0\n",
      "C  0.000000  0.0  0.0\n",
      "T  0.000000  0.0  0.0\n",
      "G  0.688921  0.0  1.0\n",
      "-  0.311079  0.0  0.0\n",
      "{'Variant': '#PARENT#', 'Position': array([235, 432, 309]), 'Alignment Probability': 0.6889210273007099, 'Alignment Count': 12}\n",
      "Variant: NB19/RB02 12 0.6889210273007099\n",
      "Processing RB02 NB51\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        203       359       94 \n",
      "A  0.017718  0.000000  0.945276\n",
      "C  0.000000  0.984658  0.000000\n",
      "T  0.000000  0.000000  0.000000\n",
      "G  0.982282  0.007236  0.023774\n",
      "-  0.000000  0.008106  0.030951\n",
      "{'Variant': 'A203G', 'Position': array([203, 359,  94]), 'Alignment Probability': 0.9142815297660871, 'Alignment Count': 80}\n",
      "Variant: NB51/RB02 80 0.9142815297660871\n",
      "Processing RB02 NB52\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "   147  375       423\n",
      "A  0.0  0.0  0.985546\n",
      "C  0.0  0.0  0.007381\n",
      "T  1.0  1.0  0.000000\n",
      "G  0.0  0.0  0.000000\n",
      "-  0.0  0.0  0.007073\n",
      "{'Variant': '#PARENT#', 'Position': array([147, 375, 423]), 'Alignment Probability': 0.9855464945612682, 'Alignment Count': 130}\n",
      "Variant: NB52/RB02 130 0.9855464945612682\n",
      "Processing RB02 NB35\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        94        114       161       438\n",
      "A  0.303404  0.000000  0.000000  0.006872\n",
      "C  0.005010  0.659359  0.653859  0.594478\n",
      "T  0.000000  0.340641  0.341015  0.013342\n",
      "G  0.553626  0.000000  0.000000  0.320638\n",
      "-  0.137960  0.000000  0.005126  0.064669\n",
      "{'Variant': 'A94G_G438C', 'Position': [94, 114, 161, 438], 'Alignment Probability': 0.14189209442805867, 'Alignment Count': 154}\n",
      "Variant: NB35/RB02 154 0.14189209442805867\n",
      "Processing RB02 NB09\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        199       231\n",
      "A  0.000000  0.868311\n",
      "C  0.000000  0.030134\n",
      "T  0.028809  0.000000\n",
      "G  0.971191  0.029822\n",
      "-  0.000000  0.071733\n",
      "{'Variant': 'A199G_G231A', 'Position': [199, 231], 'Alignment Probability': 0.8432952147215617, 'Alignment Count': 34}\n",
      "Variant: NB09/RB02 34 0.8432952147215617\n",
      "Processing RB02 NB11\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        151      123       548\n",
      "A  0.455721  0.00000  0.000000\n",
      "C  0.000000  0.00000  0.993404\n",
      "T  0.000000  0.00786  0.006596\n",
      "G  0.535858  0.99214  0.000000\n",
      "-  0.008421  0.00000  0.000000\n",
      "{'Variant': '#PARENT#', 'Position': array([151, 123, 548]), 'Alignment Probability': 0.5281394071974054, 'Alignment Count': 119}\n",
      "Variant: NB11/RB02 119 0.5281394071974054\n",
      "Processing RB02 NB42\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        80   272       161\n",
      "A  0.000000  0.0  0.000000\n",
      "C  0.988738  1.0  0.966832\n",
      "T  0.000000  0.0  0.015339\n",
      "G  0.000000  0.0  0.017830\n",
      "-  0.011262  0.0  0.000000\n",
      "{'Variant': 'T80C', 'Position': array([ 80, 272, 161]), 'Alignment Probability': 0.9559432710243957, 'Alignment Count': 91}\n",
      "Variant: NB42/RB02 91 0.9559432710243957\n",
      "Processing RB02 NB18\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "/home/emre/minION_results/MinION_RBC_0902723_sup/Demultiplex_cpp_70_40k_reads/RB02/NB18/alignment_minimap.bam does not exist.\n",
      "Skipping Variant: NB18/RB02\n",
      "Processing RB02 NB90\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        224  625  533\n",
      "A  0.675399  1.0  0.0\n",
      "C  0.000000  0.0  0.0\n",
      "T  0.000000  0.0  1.0\n",
      "G  0.324601  0.0  0.0\n",
      "-  0.000000  0.0  0.0\n",
      "{'Variant': 'G224A', 'Position': array([224, 625, 533]), 'Alignment Probability': 0.6753990709326787, 'Alignment Count': 71}\n",
      "Variant: NB90/RB02 71 0.6753990709326787\n",
      "Processing RB02 NB05\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        206  368       510\n",
      "A  0.000000  0.0  0.981055\n",
      "C  0.965237  1.0  0.000000\n",
      "T  0.015287  0.0  0.000000\n",
      "G  0.019476  0.0  0.009404\n",
      "-  0.000000  0.0  0.009541\n",
      "{'Variant': 'T206C', 'Position': array([206, 368, 510]), 'Alignment Probability': 0.946949960601093, 'Alignment Count': 104}\n",
      "Variant: NB05/RB02 104 0.946949960601093\n",
      "Processing RB02 NB39\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "/home/emre/minION_results/MinION_RBC_0902723_sup/Demultiplex_cpp_70_40k_reads/RB02/NB39/alignment_minimap.bam does not exist.\n",
      "Skipping Variant: NB39/RB02\n",
      "Processing RB02 NB69\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "   118  187\n",
      "A  0.0  0.0\n",
      "C  0.0  0.0\n",
      "T  0.0  0.0\n",
      "G  1.0  1.0\n",
      "-  0.0  0.0\n",
      "{'Variant': 'A118G_A187G', 'Position': [118, 187], 'Alignment Probability': 1.0, 'Alignment Count': 21}\n",
      "Variant: NB69/RB02 21 1.0\n",
      "Processing RB02 NB25\n",
      "Alignment file already exists, skipping alignment and indexing\n",
      "        176  232\n",
      "A  0.028164  NaN\n",
      "C  0.000000  NaN\n",
      "T  0.005165  NaN\n",
      "G  0.952300  NaN\n",
      "-  0.014371  NaN\n",
      "{'Variant': 'A176G_G232A', 'Position': [176, 232], 'Alignment Probability': nan, 'Alignment Count': 196}\n",
      "Variant: NB25/RB02 196 nan\n"
     ]
    }
   ],
   "source": [
    "demultiplex_folder = Path(\"/home/emre/minION_results/MinION_RBC_0902723_sup/Demultiplex_cpp_70_40k_reads\")\n",
    "template_seq = Path(\"/home/emre/github_repo/MinION/minION/refseq/hetcpiii_padded.fasta\")\n",
    "ref_name = \"HetCPIII\"\n",
    "barcode_dicts = analyser.get_barcode_dict(demultiplex_folder, \"NB\", \"RB\")\n",
    "\n",
    "\n",
    "variant_df_soft = get_variant_df_soft(demultiplex_folder, template_seq, ref_name, barcode_dicts, merge = True, min_depth= 5, padding=50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "variant_df_soft.to_pickle(\"/home/emre/github_repo/MinION/results/2_hetcpiii_minion_errorprone/local/variants_SW_soft_40k.pkl\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Plate</th>\n",
       "      <th>Well</th>\n",
       "      <th>Position</th>\n",
       "      <th>Variant</th>\n",
       "      <th>Alignment Probability</th>\n",
       "      <th>Alignment Count</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>A1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>A2</td>\n",
       "      <td>[193, 293, 271]</td>\n",
       "      <td>#PARENT#</td>\n",
       "      <td>0.916895</td>\n",
       "      <td>115</td>\n",
       "      <td>115.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>A3</td>\n",
       "      <td>[174, 521, 197]</td>\n",
       "      <td>#PARENT#</td>\n",
       "      <td>0.509708</td>\n",
       "      <td>172</td>\n",
       "      <td>172.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>A4</td>\n",
       "      <td>[130, 566, 401]</td>\n",
       "      <td>#PARENT#</td>\n",
       "      <td>0.591522</td>\n",
       "      <td>87</td>\n",
       "      <td>87.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>A5</td>\n",
       "      <td>[157, 176, 206, 642, 643, 644]</td>\n",
       "      <td>A107G_C592A_C594A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>135</td>\n",
       "      <td>135.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>283</th>\n",
       "      <td>3</td>\n",
       "      <td>H8</td>\n",
       "      <td>[225, 56, 126]</td>\n",
       "      <td>T175C</td>\n",
       "      <td>0.920401</td>\n",
       "      <td>33</td>\n",
       "      <td>33.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>284</th>\n",
       "      <td>3</td>\n",
       "      <td>H9</td>\n",
       "      <td>[633, 296, 268]</td>\n",
       "      <td>#PARENT#</td>\n",
       "      <td>0.935064</td>\n",
       "      <td>86</td>\n",
       "      <td>86.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>285</th>\n",
       "      <td>3</td>\n",
       "      <td>H10</td>\n",
       "      <td>[77, 185, 200]</td>\n",
       "      <td>T27A_T135A_T150C</td>\n",
       "      <td>0.800266</td>\n",
       "      <td>102</td>\n",
       "      <td>102.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>286</th>\n",
       "      <td>3</td>\n",
       "      <td>H11</td>\n",
       "      <td>[157, 233, 259]</td>\n",
       "      <td>A107G_T183C_T209C</td>\n",
       "      <td>0.882437</td>\n",
       "      <td>69</td>\n",
       "      <td>69.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>287</th>\n",
       "      <td>3</td>\n",
       "      <td>H12</td>\n",
       "      <td>[157, 233, 259]</td>\n",
       "      <td>A107G_T183A_T209C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>53</td>\n",
       "      <td>53.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>288 rows  7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Plate Well                        Position            Variant  \\\n",
       "0        1   A1                             NaN                NaN   \n",
       "1        1   A2                 [193, 293, 271]           #PARENT#   \n",
       "2        1   A3                 [174, 521, 197]           #PARENT#   \n",
       "3        1   A4                 [130, 566, 401]           #PARENT#   \n",
       "4        1   A5  [157, 176, 206, 642, 643, 644]  A107G_C592A_C594A   \n",
       "..     ...  ...                             ...                ...   \n",
       "283      3   H8                  [225, 56, 126]              T175C   \n",
       "284      3   H9                 [633, 296, 268]           #PARENT#   \n",
       "285      3  H10                  [77, 185, 200]   T27A_T135A_T150C   \n",
       "286      3  H11                 [157, 233, 259]  A107G_T183C_T209C   \n",
       "287      3  H12                 [157, 233, 259]  A107G_T183A_T209C   \n",
       "\n",
       "    Alignment Probability Alignment Count  count  \n",
       "0                     NaN             NaN    NaN  \n",
       "1                0.916895             115  115.0  \n",
       "2                0.509708             172  172.0  \n",
       "3                0.591522              87   87.0  \n",
       "4                     NaN             135  135.0  \n",
       "..                    ...             ...    ...  \n",
       "283              0.920401              33   33.0  \n",
       "284              0.935064              86   86.0  \n",
       "285              0.800266             102  102.0  \n",
       "286              0.882437              69   69.0  \n",
       "287                   NaN              53   53.0  \n",
       "\n",
       "[288 rows x 7 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variant_df_soft"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "medaka",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
