{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Variant calling with bcftools or custom function\n",
    "\n",
    "- Bcftools creates pileup from bam file, which in our case was generated using minimap2. To the variant calling, custom thresholds can be set.\n",
    "\n",
    "### Goal of the analysis\n",
    "\n",
    "- Identify variants based on quality (e.g Phred quality score) & Alignment frequency\n",
    "- If there are multiple variants, check which occur together and which are seperated (mixed population).\n",
    "- Check if the variant is an actual variant or sequencing error\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<module 'minION.consensus' from '/home/emre/github_repo/MinION/minION/consensus.py'>"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import sys\n",
    "import os\n",
    "import importlib\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import pysam \n",
    "import subprocess\n",
    "from pathlib import Path\n",
    "from uncertainties import ufloat\n",
    "from collections import Counter\n",
    "from matplotlib import pyplot as plt\n",
    "from itertools import combinations\n",
    "from scipy.stats import chi2_contingency, fisher_exact\n",
    "from sklearn.metrics import jaccard_score\n",
    "\n",
    "\n",
    "sys.path.append(\"/home/emre/github_repo/MinION\")\n",
    "\n",
    "from minION import analyser\n",
    "from minION import consensus\n",
    "from minION.util import IO_processor\n",
    "importlib.reload(IO_processor)\n",
    "importlib.reload(analyser)\n",
    "importlib.reload(consensus)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def call_and_filter_vcf(input_path, reference, allele_frequency):\n",
    "    \"\"\" Uses bcftools to call variants and filter them based on allele frequency.\n",
    "    Args:\n",
    "        - input_path (str): Path to the input files.\n",
    "        - reference (str): Path to the reference genome.\n",
    "        - allele_frequency (float): Allele frequency threshold.\n",
    "    Returns:\n",
    "        - Subprocess run\n",
    "    \"\"\"\n",
    "    \n",
    "    \n",
    "    prompt_pileup = f\"bcftools mpileup -d 4000 -Ou -f {reference}  {input_path}/alignment.bam > {input_path}/pileup.bcf\"\n",
    "\n",
    "    prompt_call = f\"bcftools call -mv -Ob -o {input_path}/raw_variants.bcf {input_path}/pileup.bcf\"\n",
    "\n",
    "    prompt_view = f\"bcftools view -i 'INFO/AF>{allele_frequency}' -Ob -o {input_path}/filtered_variants.vcf {input_path}/raw_variants.bcf\"\n",
    "\n",
    "\n",
    "    subprocess.run(prompt_pileup, shell=True)\n",
    "\n",
    "    subprocess.run(prompt_call, shell=True)\n",
    "\n",
    "    subprocess.run(prompt_view, shell=True)\n",
    "\n",
    "    print(f\"Variant calling and filtering completed. Output saved to {input_path}/raw_variants_python.bcf\")\n",
    "\n",
    "\n",
    "def extract_positions_from_vcf(vcf_file : str) -> list:\n",
    "    \"\"\" Extracts the positions of the variants from a VCF file.\n",
    "    Args:\n",
    "        - vcf_file (str): Path to the VCF file. \n",
    "    Returns:\n",
    "        - positions (list): List of variant positions.\n",
    "    \"\"\"\n",
    "    positions = []\n",
    " \n",
    "    vcf = pysam.VariantFile(vcf_file)\n",
    "\n",
    "    for record in vcf:\n",
    "        positions.append(record.pos)\n",
    "    \n",
    "    vcf.close()\n",
    "\n",
    "    return positions\n",
    "\n",
    "def extract_mutations_from_vcf(vcf_file):\n",
    "    \"\"\" Extracts the mutations from a VCF file.\n",
    "    Args:\n",
    "        - vcf_file (str): Path to the VCF file.\n",
    "    Returns:\n",
    "        - formatted_mutations (str): Formatted mutations (e.g A100T_G223C).\n",
    "    \"\"\"\n",
    "\n",
    "    mutations = []\n",
    "\n",
    "    vcf = pysam.VariantFile(vcf_file)\n",
    "\n",
    "    for record in vcf:\n",
    "\n",
    "        ref_allele = record.ref\n",
    "        alt_alleles = record.alts\n",
    "\n",
    "        for alt_allele in alt_alleles:\n",
    "            position = record.pos\n",
    "            mutation = f\"{ref_allele}{position}{alt_allele}\"\n",
    "            mutations.append(mutation)\n",
    "\n",
    "    vcf.close()\n",
    "\n",
    "    formatted_mutations = \"_\".join(mutations)\n",
    "\n",
    "    return formatted_mutations\n",
    "\n",
    "\n",
    "def get_base_counts_at_position(bam_file, chrom, position):\n",
    "    \"\"\"\n",
    "    Extract unique bases and gaps (deletions) and their counts from a BAM file \n",
    "    at a specific position using pileup.\n",
    "    \n",
    "    Args:\n",
    "    - bam_file (str): path to the BAM file.\n",
    "    - chrom (str): chromosome or contig name.\n",
    "    - position (int): 1-based position to extract bases from.\n",
    "\n",
    "    Returns:\n",
    "    - dict: unique bases and gaps with their counts at the specified position.\n",
    "    \"\"\"\n",
    "    \n",
    "    bases = []\n",
    "\n",
    "    \n",
    "    with pysam.AlignmentFile(bam_file, 'rb') as bam:\n",
    "        for pileup_column in bam.pileup(chrom, position - 1, position, \n",
    "                                        min_base_quality=0, \n",
    "                                        min_mapping_quality=0, \n",
    "                                        truncate=True):\n",
    "            if pileup_column.pos == position - 1: \n",
    "                for pileup_read in pileup_column.pileups:\n",
    "                    if pileup_read.is_del:\n",
    "                        bases.append(\"-\")  \n",
    "                    elif not pileup_read.is_refskip:\n",
    "                        bases.append(pileup_read.alignment.query_sequence[pileup_read.query_position])\n",
    "\n",
    "    base_counts = Counter(bases)\n",
    "\n",
    "    return base_counts\n",
    "\n",
    "\n",
    "def generate_heatmap_data(bam_file, chrom, positions):\n",
    "    data = []\n",
    "    all_bases = set(['A', 'T', 'C', 'G', '-'])\n",
    "    for position in positions:\n",
    "        base_counts = get_base_counts_at_position(bam_file, chrom, position)\n",
    "        col = [base_counts.get(base, 0) for base in all_bases]\n",
    "        data.append(col)\n",
    "    df = pd.DataFrame(data, index=positions, columns=list(all_bases))\n",
    "    # Order as in all_bases\n",
    "    # TODO: Specify if a character if it is an actual gap or no coverage\n",
    "    df = df[['A', 'T', 'C', 'G', '-']]\n",
    "    return df\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "def get_most_common_base(position, heatmap_data, ref_base):\n",
    "    non_ref_data = heatmap_data.loc[position].drop(ref_base)\n",
    "    return non_ref_data.idxmax(), non_ref_data.max()\n",
    "\n",
    "\n",
    "\n",
    "def get_bases_from_pileup(bam_file, chrom, positions):\n",
    "    bases_dict = {position: {} for position in positions}\n",
    "    \n",
    "    with pysam.AlignmentFile(bam_file, 'rb') as bam:\n",
    "        for pileup_column in bam.pileup(chrom, min(positions) - 1, max(positions) + 1,\n",
    "                                        min_base_quality=0, \n",
    "                                        min_mapping_quality=0, \n",
    "                                        truncate=True):\n",
    "            if pileup_column.pos + 1 in positions:\n",
    "                for pileup_read in pileup_column.pileups:\n",
    "                    if not pileup_read.is_del and not pileup_read.is_refskip:\n",
    "                        base = pileup_read.alignment.query_sequence[pileup_read.query_position]\n",
    "                        read_name = pileup_read.alignment.query_name\n",
    "                        # Check if the read name is already added to the bases_dict for that position\n",
    "                        if read_name not in bases_dict[pileup_column.pos + 1]:\n",
    "                            bases_dict[pileup_column.pos + 1][read_name] = base\n",
    "\n",
    "    # Get unique read names and sort them to have a consistent order\n",
    "    read_names = sorted(set().union(*[bases_dict[pos].keys() for pos in bases_dict]))\n",
    "\n",
    "    # Create DataFrame\n",
    "    df = pd.DataFrame(index=read_names, columns=positions)\n",
    "    \n",
    "    # Populate DataFrame\n",
    "    for pos in positions:\n",
    "        for read_name, base in bases_dict[pos].items():\n",
    "            df.at[read_name, pos] = base\n",
    "    \n",
    "    # Fill NaN with \"-\"\n",
    "    df = df.fillna(\"-\")\n",
    "\n",
    "    return df\n",
    "\n",
    "def get_max_base_at_position(bam_file, chrom, position):\n",
    "    base_counts = get_base_counts_at_position(bam_file, chrom, position)\n",
    "    return base_counts.most_common(1)[0][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "get_base_counts_at_position"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "barcodes = IO_processor.get_barcode_dict(Path(\"/home/emre/minION_results/MinION_RBC_0902723_sup/Demultiplex_cpp_70_short\"))\n",
    "#path_to_bam = \"/home/emre/minION_results/MinION_RBC_0902723_sup/Demultiplex_cpp_70_short/RB03/NB08/consensus\"\n",
    "path_to_bam = \"/home/emre/minION_results/20231119_1502_MN41105/Demultiplex_cpp_70/RB01/NB22/consensus\"\n",
    "#template_fasta = \"/home/emre/github_repo/MinION/minION/refseq/hetcpiii.fasta\"\n",
    "#template_fasta = Path(\"/home/emre/tam-lqv.fasta\")\n",
    "template_fasta = Path(\"/home/emre/PgA9.fasta\")\n",
    "#output_vcf = f\"{path_to_bam}/raw_variants.vcf\"\n",
    "#consensus.get_consensus(Path(path_to_bam), Path(template_fasta))\n",
    "#call_and_filter_vcf(path_to_bam, template_fasta, 0.4)\n",
    "bam_file = f'{path_to_bam}/alignment.bam'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "extract_mutations_from_vcf(f'{path_to_bam}/filtered_variants.vcf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "nb_positions = extract_positions_from_vcf(f'{path_to_bam}/filtered_variants.vcf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Counter({'A': 396})"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#reference = \"HetCPIII\"\n",
    "#reference = \"Tam-LQV\"\n",
    "reference = \"PgA9\"  \n",
    "position_to_check = 1  \n",
    "bases_at_position = get_base_counts_at_position(bam_file, reference, position_to_check)\n",
    "bases_at_position"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#heatmap_data = generate_heatmap_data(bam_file, reference, 1, 100)\n",
    "nb_positions = [23,38,56,98,421]\n",
    "heatmap_data = generate_heatmap_data(bam_file, reference, nb_positions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "heatmap_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "template = analyser.get_template_sequence(template_fasta)\n",
    "\n",
    "# Plotting\n",
    "plt.figure(figsize=(7, 4))\n",
    "sns.heatmap(heatmap_data, \n",
    "            annot=False, \n",
    "            cmap='YlGnBu', \n",
    "            cbar_kws={'label': 'Count'}, \n",
    "            linewidths=0.5, \n",
    "            linecolor='white')\n",
    "plt.title(\"Base Counts per Position\")\n",
    "plt.xlabel(\"Bases\")\n",
    "plt.ylabel(\"Position\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bam_file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get Highest Non Reference Base Frequency\n",
    "\n",
    "- Bam file\n",
    "- Reference sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_seq = analyser.get_template_sequence(template_fasta)\n",
    "freq_dist = pd.DataFrame(analyser.get_highest_non_ref_base_freq(bam_file, reference, range(1,len(ref_seq)), ref_seq)[0]).T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "freq_dist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Replace NaN value with \"T\"\n",
    "freq_dist[0] = freq_dist[0].fillna(\"None\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAABIQAAAJCCAYAAACxsxylAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjcuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8pXeV/AAAACXBIWXMAAA9hAAAPYQGoP6dpAAA40ElEQVR4nO3de5SVdb0/8M8IDDdlJBFMHXFSMZRTKZw8eM08omAttU5RXpCClPKGpC05rvKHXbQbqatALRpFU8mycrlIpVLzWkcFPeY1A0ZhCEXPDODhIrN/f3Bm2LNnzzCzZ89shu/rtdazdJ79fZ7n+9y+e/v2+32eskwmkwkAAAAAkrFLqSsAAAAAQPcSCAEAAAAkRiAEAAAAkBiBEAAAAEBiBEIAAAAAiREIAQAAACRGIAQAAACQmN6lrkB3a2hoiJUrV8Zuu+0WZWVlpa4OAAAAQFFkMplYu3Zt7L333rHLLm33AUouEFq5cmVUVlaWuhoAAAAAXeL111+Pfffdt80yyQVCu+22W0RsPTiDBg0qcW0AAAAAiqO+vj4qKyubso+2JBcINQ4TGzRokEAIAAAA2Om05xE5HioNAAAAkBiBEAAAAEBiBEIAAAAAiREIAQAAACRGIAQAAACQGIEQAAAAQGIEQgAAAACJEQgBAAAAJEYgBAAAAJAYgRAAAABAYgRCAAAAAIkRCAEAJXf11VfHv/7rv8Zuu+0WQ4cOjdNOOy1efvnlZmXWrVsXF1xwQey7777Rv3//GDlyZMydO7dZmfPOOy8OOOCA6N+/f+y5555x6qmnxksvvZR3m+eee2706tUr7rzzzi7bLwCAHZVACAAouYcffjjOP//8ePLJJ2PRokXx3nvvxbhx42L9+vVNZS655JK477774rbbbosXX3wxLrnkkrjwwgvjd7/7XVOZ0aNHR3V1dbz44otx//33RyaTiXHjxsWWLVuabe/dd9+NBQsWxGWXXRbz5s3rtv0EANhRlGUymUypK9Gd6uvro6KiIurq6mLQoEGlrg4AkMebb74ZQ4cOjYcffjiOPfbYiIgYNWpUTJw4Mb7+9a83lRs9enRMmDAhvvnNb+Zdz3PPPRcf/vCH4+9//3sccMABTfNvueWWuOGGG+K+++6L97///fHCCy/E/vvv36X7BADQ1TqSeeghBADscOrq6iIi4n3ve1/TvKOPPjruueeeWLFiRWQymXjwwQfjlVdeiZNOOinvOtavXx/V1dVRVVUVlZWVzT6bN29enHXWWVFRURETJkyI6urqrtsZAIAdkEAIANihZDKZmDFjRhx99NExatSopvnXX399HHLIIbHvvvtGeXl5nHzyyTFnzpw4+uijmy0/Z86c2HXXXWPXXXeN++67LxYtWhTl5eVNn7/66qvx5JNPxsSJEyMi4qyzzorq6upoaGjonh0EANgBCIQAgB3KBRdcEM8991zccccdzeZff/318eSTT8Y999wTTz/9dPzwhz+Mr3zlK/GHP/yhWbkzzzwzFi9eHA8//HAcdNBB8dnPfjY2bNjQ9Pm8efPipJNOiiFDhkRExIQJE2L9+vUt1gMAsDPzDCEAYIdx4YUXxm9/+9v485//HFVVVU3z//d//zcqKiriN7/5TZxyyilN86dOnRpvvPFG3HfffXnXt2nTphg8eHD87Gc/i89//vOxZcuWqKysjFWrVsUuu2z7/2JbtmyJz372s7FgwYKu2zkAgC7WkcyjdzfVCQCgVZlMJi688ML4zW9+Ew899FCzMCgiYvPmzbF58+ZmIU5ERK9evbY71CuTycTGjRsjImLhwoWxdu3aWLx4cfTq1aupzEsvvRRnnnlmrFmzJvbYY48i7RUAwI5LIAQAlNz5558ft99+e/zud7+L3XbbLVatWhURERUVFdG/f/8YNGhQHHfccXHZZZdF//79Y/jw4fHwww/H/PnzY/bs2RER8Y9//CMWLFgQ48aNiz333DNWrFgR3/3ud6N///4xYcKEiNg6XOyUU06JD3/4w822f+ihh8b06dPjtttui4svvrh7dx4AoAQMGQMASq6srCzv/Orq6pg8eXJERKxatSpmzpwZDzzwQLz99tsxfPjwOPfcc+OSSy6JsrKyWLlyZUydOjWefvrpeOedd2LYsGFx7LHHxje+8Y04+OCD45///Gfsu+++cfvtt8dnPvOZFtu66KKL4qGHHornnnuuK3cVAKDLdCTzEAgBAAAA7AQ6knl4yxgAAABAYgRCAECPtayqKpblPIAaAIDtEwgBAAAAJEYgBAAAAJAYgRAAAABAYgRCAAAAAIkRCAEAAAAkRiAEAAAAkBiBEAAAAEBiBEIAAAAAiREIAQAAACRGIAQAAACQGIEQAAAAQGIEQgAAAACJEQgBAAAAJEYgBAAAAJAYgRAAAABAYgRCAAAAAIkRCAEAAAAkRiAEAAAAkBiBEAAAAEBiBEIAAAAAiREIAQAAACRGIAQAAACQGIEQAAAAQGIEQgAAAACJEQgBAAAAJEYgBAAAAJAYgRAAAABAYgRCAAAAAIkRCAEAAAAkRiAEAAAAkBiBEAAAAEBiBEIAAAAAiREIAQAAACRGIAQAAACQGIEQAAAAQGIEQgAAAACJEQgBAAAAJEYgBAAAAJAYgRAAAABAYgRCAAAAAIkRCAEAAAAkRiAEAAAAkBiBEAAAAEBiBEIAAAAAiREIAQAAACRGIAQAAACQGIEQAAAAQGIEQgAAAACJEQgBAAAAJKbkgdCcOXOiqqoq+vXrF6NHj45HHnmkXcs99thj0bt37/jIRz7StRUEAAAA2MmUNBBasGBBTJ8+Pa644opYvHhxHHPMMTF+/Pioqalpc7m6urqYNGlSnHDCCd1UUwAAAICdR0kDodmzZ8eUKVNi6tSpMXLkyLj22mujsrIy5s6d2+Zy5513XpxxxhkxduzYbqopAAAAwM6jZIHQpk2b4umnn45x48Y1mz9u3Lh4/PHHW12uuro6XnvttbjyyivbtZ2NGzdGfX19swkAAAAgZSULhN56663YsmVLDBs2rNn8YcOGxapVq/Iu8+qrr8bll18ev/jFL6J3797t2s7VV18dFRUVTVNlZWWn6w4AAADQk5X8odJlZWXN/s5kMi3mRURs2bIlzjjjjJg1a1aMGDGi3eufOXNm1NXVNU2vv/56p+sMAAAA0JO1r5tNFxgyZEj06tWrRW+g1atXt+g1FBGxdu3aeOqpp2Lx4sVxwQUXREREQ0NDZDKZ6N27dzzwwAPx8Y9/vMVyffv2jb59+3bNTgAAAAD0QCXrIVReXh6jR4+ORYsWNZu/aNGiOPLII1uUHzRoUPz3f/93LFmypGmaNm1aHHzwwbFkyZI44ogjuqvqAAAAAD1ayXoIRUTMmDEjzj777BgzZkyMHTs2brrppqipqYlp06ZFxNbhXitWrIj58+fHLrvsEqNGjWq2/NChQ6Nfv34t5gMAAADQupIGQhMnTow1a9bEVVddFbW1tTFq1KhYuHBhDB8+PCIiamtro6amppRVBAAAANjplGUymUypK9Gd6uvro6KiIurq6mLQoEGlrg4A0AnLqqoiImL/pUtLXBMAgNLrSOZR8reMAQAAANC9BEIAAAAAiREIAQAAACRGIAQAAACQGIEQAAAAQGIEQgAAAACJEQgBAAAAJEYgBAAAAJAYgRAAAABAYgRCAAAAAIkRCAEAAAAkRiAEAAAAkBiBEAAAAEBiBEIAAAAAiREIAQAAACRGIAQAAACQGIEQAAAAQGIEQgAAAACJEQgBAAAAJEYgBAAAAJAYgRAAAABAYgRCAAAAAIkRCAEAAAAkRiAEAAAAkBiBEAAAAEBiBEIAAAAAiREIAQAAACRGIAQAAACQGIEQAAAAQGIEQgAAAACJEQgBAAAAJEYgBAAAAJAYgRAAAABAYgRCAAAAAIkRCAEAAAAkRiAEAAAAkBiBEAAAAEBiBEIAAAAAiREIAQAAACRGIAQAAACQGIEQAAAAQGIEQgAAAACJEQgBAAAAJEYgBAAAAJAYgRAAAABAYgRCAAAAAIkRCAEAAAAkRiAEAAAAkBiBEAAAAEBiBEIAAAAAiREIAQAAACRGIAQAAACQGIEQAAAAQGIEQgAAAACJEQgBAAAAJEYgBAAAAJAYgRAAAABAYgRCAAAAAIkRCAEAAAAkRiAEAAAAkBiBEAAAAEBiBEIAAAAAiREIAQAAACRGIAQAAACQGIEQAAAAQGIEQgAAAACJEQgBAAAAJEYgBAAAAJAYgRAAAABAYgRCAAAAAIkRCAEAAAAkRiAEAAAAkBiBEAAAAEBiBEIAAAAAiREIAQAAACRGIAQAAACQGIEQAAAAQGIEQgAAAACJEQgBAAAAJEYgBAAAAJAYgRAAAABAYgRCAAAAAIkRCAEAAAAkRiAEAAAAkBiBEAAAAEBiBEIAAAAAiREIAQAAACRGIAQAAACQGIEQAAAAQGIEQgAAAACJEQgBAAAAJEYgBAAAAJAYgRAAAABAYgRCAAAAAIkRCAEAAAAkRiAEAAAAkBiBEAAAAEBiBEIAAAAAiREIAQAAACRGIAQAAACQGIEQAAAAQGIEQgAAAACJEQgBAAAAJEYgBAAAAJCYkgdCc+bMiaqqqujXr1+MHj06HnnkkVbLPvroo3HUUUfFHnvsEf37948PfvCD8aMf/agbawsAAADQ8/Uu5cYXLFgQ06dPjzlz5sRRRx0VN954Y4wfPz5eeOGF2G+//VqUHzhwYFxwwQXxoQ99KAYOHBiPPvponHfeeTFw4MA499xzS7AHAAAAAD1PWSaTyZRq40cccUQcfvjhMXfu3KZ5I0eOjNNOOy2uvvrqdq3jU5/6VAwcODBuvfXWdpWvr6+PioqKWLmyLgYNGtTi8169Ivr12/b3+vWtr2uXXSL69y+s7LvvRrR25MvKIgYMKKzs//5vREND6/UYOLCwshs2RGzZUpyyAwZsrXdExMaNEe+9V5yy/ftvPc4REZs2RWzeXJyy/fptvS46Wnbz5q3lW9O3b0Tv3h0v+957W49Fa8rLI/r06XjZLVu2nrvW9OmztXxHyzY0bL3WilG2d++txyJi6z3x7rvFKduR+14bkb+sNqLjZbURW/+9s23E8kMOiYiI4S+8oI34P9qIwspqI7ba2dqIbNqIrbQRhZXVRmyljeh42e5uI+rr62PvvSuiri5/5tFMpkQ2btyY6dWrV+buu+9uNv+iiy7KHHvsse1axzPPPJMZNmxY5qc//WmrZTZs2JCpq6trml5//fVMRGQi6jJbT13zacKE5ssPGNCyTON03HHNyw4Z0nrZMWOalx0+vPWyhxzSvOwhh7Redvjw5mXHjGm97JAhzcsed1zrZQcMaF52woTWy+ZeRf/xH22XXbduW9lzzmm77OrV28p+5Sttl126dFvZSy9tu+zzz28re+WVbZf961+3lf3e99ou++CD28r++Mdtl7333m1lq6vbLvvLX24r+8tftl22unpb2Xvvbbvsj3+8reyDD7Zd9nvf21b2r39tu+yVV24r+/zzbZe99NJtZZcubbvsV76yrezq1W2XPeecbWXXrWu77H/8R6aZtspqI7ZO2ohtkzZi66SN2DppI7ZO2ohtkzZi66SN2DppI7ZO2ohtkzZi66SN2DoVp42oy0REpq6uLrM9JXuG0FtvvRVbtmyJYcOGNZs/bNiwWLVqVZvL7rvvvtG3b98YM2ZMnH/++TF16tRWy1599dVRUVHRNFVWVhal/gAAAAA9VcmGjK1cuTL22WefePzxx2Ps2LFN87/97W/HrbfeGi+99FKryy5dujTWrVsXTz75ZFx++eXx4x//OD7/+c/nLbtx48bYmNWPrb6+PiorKw0Z62BZ3Tg7XlY3zq3/rhtnYWW1EVv/XRvR8bKptRGGjLWkjSisrDZiq52tjcimjdhKG1FYWW3EVtqIjpfdkYeMlSwQ2rRpUwwYMCDuuuuuOP3005vmX3zxxbFkyZJ4+OGH27Web33rW3HrrbfGyy+/3K7yjc8Qatd4OgBgh7asqioiIvZfurTENQEAKL2OZB4lGzJWXl4eo0ePjkWLFjWbv2jRojjyyCPbvZ5MJtOsBxAAAAAAbSvpa+dnzJgRZ599dowZMybGjh0bN910U9TU1MS0adMiImLmzJmxYsWKmD9/fkRE/OQnP4n99tsvPvjBD0ZExKOPPho/+MEP4sILLyzZPgAAAAD0NCUNhCZOnBhr1qyJq666Kmpra2PUqFGxcOHCGD58eERE1NbWRk1NTVP5hoaGmDlzZixdujR69+4dBxxwQFxzzTVx3nnnlWoXAAAAAHqckj1DqFQ8QwgAdh6eIQQAsE2PeIYQAAAAAKUhEAIAAABIjEAIAAAAIDECIQAAAIDECIQAAAAAEiMQAgAAAEiMQAgAAAAgMQIhAAAAgMQIhAAAAAASIxACAAAASIxACAAAACAxAiEAAACAxAiEAAAAABIjEAIAAABIjEAIAAAAIDECIQAAAIDECIQAAAAAEiMQAgAAAEiMQAgAAAAgMQIhAAAAgMQIhAAAAAASIxACAAAASIxACAAAACAxAiEAAACAxAiEAAAAABIjEAIAAABIjEAIAAAAIDECIQAAAIDECIQAAAAAEiMQAgAAAEiMQAgAAAAgMQIhAAAAgMQIhAAAAAASIxACAAAASIxACAAAACAxBQVCS5cuLXY9AAAAAOgmBQVCBx54YBx//PFx2223xYYNG4pdJwAAAAC6UEGB0LPPPhuHHXZYfPWrX4299torzjvvvPjrX/9a7LoBAAAA0AUKCoRGjRoVs2fPjhUrVkR1dXWsWrUqjj766Dj00ENj9uzZ8eabbxa7ngAAAAAUSaceKt27d+84/fTT45e//GV897vfjddeey0uvfTS2HfffWPSpElRW1tbrHoCAAAAUCSdCoSeeuqp+MpXvhLvf//7Y/bs2XHppZfGa6+9Fn/6059ixYoVceqppxarngAAAAAUSe9CFpo9e3ZUV1fHyy+/HBMmTIj58+fHhAkTYpddtuZLVVVVceONN8YHP/jBolYWAAAAgM4rKBCaO3dufPGLX4wvfOELsddee+Uts99++8W8efM6VTkAAAAAiq+gQOjVV1/dbpny8vI455xzClk9AAAAAF2ooGcIVVdXx1133dVi/l133RW33HJLpysFAAAAQNcpKBC65pprYsiQIS3mDx06NL7zne90ulIAAAAAdJ2CAqHly5dHVVVVi/nDhw+PmpqaTlcKAAAAgK5TUCA0dOjQeO6551rMf/bZZ2OPPfbodKUAAAAA6DoFBUKf+9zn4qKLLooHH3wwtmzZElu2bIk//elPcfHFF8fnPve5YtcRAAAAgCIq6C1j3/rWt2L58uVxwgknRO/eW1fR0NAQkyZN8gwhAAAAgB1cQYFQeXl5LFiwIL75zW/Gs88+G/37949/+Zd/ieHDhxe7fgAAAAAUWUGBUKMRI0bEiBEjilUXAAAAALpBQYHQli1b4uabb44//vGPsXr16mhoaGj2+Z/+9KeiVA4AAACA4isoELr44ovj5ptvjlNOOSVGjRoVZWVlxa4XAAAAAF2koEDozjvvjF/+8pcxYcKEYtcHAAAAgC5W0Gvny8vL48ADDyx2XQAAAADoBgUFQl/96lfjuuuui0wmU+z6AAAAANDFChoy9uijj8aDDz4Yv//97+PQQw+NPn36NPv87rvvLkrlAAAAACi+ggKh3XffPU4//fRi1wUAAACAblBQIFRdXV3segAAAADQTQp6hlBExHvvvRd/+MMf4sYbb4y1a9dGRMTKlStj3bp1RascAAAAAMVXUA+h5cuXx8knnxw1NTWxcePGOPHEE2O33XaL733ve7Fhw4a44YYbil1PAAAAAIqkoB5CF198cYwZMybeeeed6N+/f9P8008/Pf74xz8WrXIAAAAAFF/Bbxl77LHHory8vNn84cOHx4oVK4pSMQAAAAC6RkE9hBoaGmLLli0t5r/xxhux2267dbpSAAAAAHSdggKhE088Ma699tqmv8vKymLdunVx5ZVXxoQJE4pVNwAAAAC6QEFDxn70ox/F8ccfH4ccckhs2LAhzjjjjHj11VdjyJAhcccddxS7jgAAAAAUUUGB0N577x1LliyJO+64I5555ploaGiIKVOmxJlnntnsIdMAAAAA7HjKMplMptSV6E719fVRUVERdXV1MWjQoFJXBwDohGVVVRERsf/SpSWuCQBA6XUk8yioh9D8+fPb/HzSpEmFrBYAAACAblBQIHTxxRc3+3vz5s3x7rvvRnl5eQwYMEAgBAAAALADK+gtY++8806zad26dfHyyy/H0Ucf7aHSAAAAADu4ggKhfA466KC45pprWvQeAgAAAGDHUrRAKCKiV69esXLlymKuEgAAAIAiK+gZQvfcc0+zvzOZTNTW1saPf/zjOOqoo4pSMQAAAAC6RkGB0Gmnndbs77Kysthzzz3j4x//ePzwhz8sRr0AAAAA6CIFBUINDQ3FrgcAAAAA3aSozxACAAAAYMdXUA+hGTNmtLvs7NmzC9kEAAAAAF2koEBo8eLF8cwzz8R7770XBx98cEREvPLKK9GrV684/PDDm8qVlZUVp5YAAAAAFE1BgdAnP/nJ2G233eKWW26JwYMHR0TEO++8E1/4whfimGOOia9+9atFrSQAAAAAxVOWyWQyHV1on332iQceeCAOPfTQZvOff/75GDduXKxcubJoFSy2+vr6qKioiLq6uhg0aFCpqwMAdMKyqqqIiNh/6dIS1wQAoPQ6knkU9FDp+vr6+Oc//9li/urVq2Pt2rWFrBIAAACAblJQIHT66afHF77whfjVr34Vb7zxRrzxxhvxq1/9KqZMmRKf+tSnil1HAAAAAIqooGcI3XDDDXHppZfGWWedFZs3b966ot69Y8qUKfH973+/qBUEAAAAoLgKeoZQo/Xr18drr70WmUwmDjzwwBg4cGAx69YlPEMIAHYeniEEALBNlz9DqFFtbW3U1tbGiBEjYuDAgdGJbAkAAACAblJQILRmzZo44YQTYsSIETFhwoSora2NiIipU6d65TwAAADADq6gQOiSSy6JPn36RE1NTQwYMKBp/sSJE+O+++4rWuUAAAAAKL6CHir9wAMPxP333x/77rtvs/kHHXRQLF++vCgVAwAAAKBrFNRDaP369c16BjV66623om/fvp2uFAAAAABdp6BA6Nhjj4358+c3/V1WVhYNDQ3x/e9/P44//viiVQ4AAACA4itoyNj3v//9+NjHPhZPPfVUbNq0Kb72ta/F3/72t3j77bfjscceK3YdAQAAACiignoIHXLIIfHcc8/FRz/60TjxxBNj/fr18alPfSoWL14cBxxwQLHrCAAAAEARdbiH0ObNm2PcuHFx4403xqxZs7qiTgAAAAB0oQ73EOrTp088//zzUVZW1hX1AQAAAKCLFTRkbNKkSTFv3rxi1wUAAACAblDQQ6U3bdoUP/vZz2LRokUxZsyYGDhwYLPPZ8+eXZTKAQAAAFB8HQqE/vGPf8T+++8fzz//fBx++OEREfHKK680K2MoGQAAAMCOrUOB0EEHHRS1tbXx4IMPRkTExIkT4/rrr49hw4Z1SeUAAAAAKL4OPUMok8k0+/v3v/99rF+/vqgVAgAAAKBrFfRQ6Ua5AREAAAAAO74OBUJlZWUtnhHkmUEAAAAAPUuHniGUyWRi8uTJ0bdv34iI2LBhQ0ybNq3FW8buvvvu4tUQAAAAgKLqUA+hc845J4YOHRoVFRVRUVERZ511Vuy9995NfzdOHTFnzpyoqqqKfv36xejRo+ORRx5ptezdd98dJ554Yuy5554xaNCgGDt2bNx///0d2h4AAABA6jrUQ6i6urqoG1+wYEFMnz495syZE0cddVTceOONMX78+HjhhRdiv/32a1H+z3/+c5x44onxne98J3bfffeorq6OT37yk/GXv/wlDjvssKLWDQAAAGBnVZYp4ZOhjzjiiDj88MNj7ty5TfNGjhwZp512Wlx99dXtWsehhx4aEydOjG984xvtKl9fXx8VFRVRV1cXgwYNKqjeAMCOYVlVVURE7L90aYlrAgBQeh3JPDr1lrHO2LRpUzz99NMxbty4ZvPHjRsXjz/+eLvW0dDQEGvXro33ve99rZbZuHFj1NfXN5sAAAAAUlayQOitt96KLVu2xLBhw5rNHzZsWKxatapd6/jhD38Y69evj89+9rOtlrn66qubPd+osrKyU/UGAAAA6OlKFgg1yn1tfSaTader7O+44474f//v/8WCBQti6NChrZabOXNm1NXVNU2vv/56p+sMAAAA0JN16KHSxTRkyJDo1atXi95Aq1evbtFrKNeCBQtiypQpcdddd8W///u/t1m2b9++0bdv307XFwAAAGBnUbIeQuXl5TF69OhYtGhRs/mLFi2KI488stXl7rjjjpg8eXLcfvvtccopp3R1NQEAAAB2OiXrIRQRMWPGjDj77LNjzJgxMXbs2LjpppuipqYmpk2bFhFbh3utWLEi5s+fHxFbw6BJkybFddddF//2b//W1Luof//+UVFRUbL9AAAAAOhJShoITZw4MdasWRNXXXVV1NbWxqhRo2LhwoUxfPjwiIiora2NmpqapvI33nhjvPfee3H++efH+eef3zT/nHPOiZtvvrm7qw8AAADQI5VlMplMqSvRnerr66OioiLq6upi0KBBpa4OANAJy6qqIiJi/6VLS1wTAIDS60jmUfK3jAEAAADQvQRCAAAAAIkRCAEAAAAkRiAEAAAAkBiBEAAAAEBiBEIAAAAAiREIAQAAACRGIAQAAACQGIEQAAAAQGIEQgAAAACJEQgBAAAAJEYgBAAAAJAYgRAAAABAYgRCAAAAAIkRCAEAAAAkRiAEAAAAkBiBEAAAAEBiBEIAAAAAiREIAQAAACRGIAQAAACQGIEQAAAAQGIEQgAAAACJEQgBAAAAJEYgBAAAAJAYgRAAAABAYgRCAAAAAIkRCAEAAAAkRiAEAAAAkBiBEAAAAEBiBEIAAAAAiREIAQAAACRGIAQAAACQGIEQAAAAQGIEQgAAAACJEQgBAAAAJEYgBAAAAJAYgRAAAABAYgRCAAAAAIkRCAEAAAAkRiAEAAAAkBiBEAAAAEBiBEIAAAAAiREIAQAAACRGIAQAAACQGIEQAAAAQGIEQgAAAACJEQgBAAAAJEYgBAAAAJAYgRAAAABAYgRCAAAAAIkRCAEAAAAkRiAEAAAAkBiBEAAAAEBiBEIAAAAAiREIAQAAACRGIAQAAACQGIEQAAAAQGIEQgAAAACJEQgBAAAAJEYgBAAAAJAYgRAAAABAYgRCAAAAAIkRCAEAAAAkRiAEAAAAkBiBEAAAAEBiBEIAAAAAiREIAQAAACRGIAQAAACQGIEQAAAAQGIEQgAAAACJEQgBAAAAJEYgBAAAAJAYgRAAAABAYgRCAAAAAIkRCAEAAAAkRiAEAAAAkBiBEAAAAEBiBEIAAAAAiREIAQAAACRGIAQAAACQGIEQAAAAQGIEQgAAAACJEQgBAAAAJEYgBAAAAJAYgRAAAABAYgRCAAAAAIkRCAEAAAAkRiAEAAAAkBiBEAAAAEBiBEIAAAAAiREIAQAAACRGIAQAAACQGIEQAAAAQGIEQgAAAACJEQgBAAAAJEYgBAAAAJAYgRAAAABAYgRCAAAAAIkRCAEAAAAkRiAEAAAAkBiBEAAAAEBiBEIAAAAAiREIAQAAACRGIAQAAACQmJIHQnPmzImqqqro169fjB49Oh555JFWy9bW1sYZZ5wRBx98cOyyyy4xffr07qsoAAAAwE6ipIHQggULYvr06XHFFVfE4sWL45hjjonx48dHTU1N3vIbN26MPffcM6644or48Ic/3M21BQAAANg5lDQQmj17dkyZMiWmTp0aI0eOjGuvvTYqKytj7ty5ecvvv//+cd1118WkSZOioqKim2sLAAAAsHMoWSC0adOmePrpp2PcuHHN5o8bNy4ef/zxom1n48aNUV9f32wCAAAASFnJAqG33nortmzZEsOGDWs2f9iwYbFq1aqibefqq6+OioqKpqmysrJo6wYAAADoiUr+UOmysrJmf2cymRbzOmPmzJlRV1fXNL3++utFWzcAAABAT9S7VBseMmRI9OrVq0VvoNWrV7foNdQZffv2jb59+xZtfQAAAAA9Xcl6CJWXl8fo0aNj0aJFzeYvWrQojjzyyBLVCgAAAGDnV7IeQhERM2bMiLPPPjvGjBkTY8eOjZtuuilqampi2rRpEbF1uNeKFSti/vz5TcssWbIkIiLWrVsXb775ZixZsiTKy8vjkEMOKcUuAAAAAPQ4JQ2EJk6cGGvWrImrrroqamtrY9SoUbFw4cIYPnx4RETU1tZGTU1Ns2UOO+ywpn9/+umn4/bbb4/hw4fHsmXLurPqAAAAAD1WWSaTyZS6Et2pvr4+Kioqoq6uLgYNGlTq6gAAnbCsqioiIvZfurTENQEAKL2OZB4lf8sYAAAAAN1LIAQAAACQGIEQAAAAQGIEQgAAAACJEQgBAAAAJEYgBAAAAJAYgRAAAABAYgRCAAAAAIkRCAEAAAAkRiAEAAAAkBiBEAAAAEBiBEIAAAAAiREIAQAAACRGIAQAAACQGIEQAAAAQGIEQgAAAACJEQgBAAAAJEYgBAAAAJAYgRAAAABAYgRCAAAAAIkRCAEAAAAkRiAEAAAAkBiBEAAAAEBiBEIAAAAAiREIAQAAACRGIAQAAACQGIEQAAAAQGIEQgAAAACJEQgBAAAAJEYgBAAAAJAYgRAAAABAYgRCAAAAAIkRCAEAALBDmzVrVsyaNavU1YCdikAIAAAAIDECIQAAAIDECIQAAAAAEiMQAgAAAEiMQAgAAAAgMQIhAAAAgMQIhAAAAAASIxACAAAASIxACAAAACAxAiEAAACAxAiEAAAAABIjEAIAAABIjEAIAAAAIDECIQAAAIDECIQAAAAAEiMQAgAAAEiMQAgAAAAgMQIhAAAAgMQIhAAAAAASIxACAAAASIxACAAAACAxAiEAAACAxAiEAAAAABIjEAIAAABIjEAIAAAAIDECIQAAAIDECIQAAAAAEiMQAgAAAEiMQAgAAAAgMQIhAAAAgMQIhAAAAAASIxACAAAASIxACAAAACAxAiEAAACAxAiEAAAAABIjEAIAgB5oWVVVLKuqKuo6Z82aFbNmzSrqOoGOcQ/uuHa2cyMQAgAAAEiMQAgAAAAgMQIhAAAAgMQIhAAAAAASIxACAAAASIxACAAAACAxAiEAAACAxAiEAAAAABIjEKJNs2bNilmzZpW6GgAAtIPfbnSG6wfSIhACAAAASIxACAAAACAxAiEAkqM7PNDTaLfo6ZZVVcWyqqpSV6PHWrasKpYt65rjN/iSwTH4ksFdsm52bAIhAAAAgMQIhMjLA+V6JueN7uI6A4B0tPd7v9Dfon7DFk7PHjpDIAQAAACQGIEQAAAAQGIEQrRLe7pwNnZXrKpa1sW1YXt0uwUA6F5+A3dMRx6Q7HdtYRw3tkcgBAAAAJAYgRAAAABAYgRCQFH01C6pHemuvDPzhgqA9unp3xuzZs2KqqplhjfRpZZVFec+6am/L6GnEAgBAAAAJEYgBAAAAJAYgdAOoCe9Eaqxrj2lvpSW66T4uvL+a23dhhVQCtoPegK/iZrb0Y/F4EsGd8kQ6baGEXb3d2ixrslC1pHa8PNiDYvr8HZ7+LDVHVmKbbpACAAAACAxAqEutmxZlRQ3h2MC6SrW/53tyDq219501f8xLqbGdjP3/1zt6PXOtbP9XzffZWnId932xP+LvKNcr509bttbfllVVcE9N2bNmtWh36k7cg/anniNpqYj92QxrrWeek3sqP/t2BOPZT4CIQAAAIDECIQAAAAAEiMQooX2DkFo7L7Xk4Ys9KS67ux21O6f2Rrr11Y9811TjUOQStWVtCNdgtt7T1RVLduhu8YXU7Guy462N11x3nYUufvW3fdGW8MC27q2i91Odfc91Nn6t2f5UrXluec03xCh1upWaH13luEBbWnPUKtUXjDSeI1lX2ddvd9dcS91dp1t7W/jZ22V2VF+O3T19Vrq4ef5rtXt2dnv4dYUep521nZPIAQAAACQGIEQAAAAQGIEQl2gra6Zjd3MqqqWtestCLndnZt1j+4BQ27a6lrXort31r6059g0rjtft7/G5fN13+zI0Jc2t1HkruhtLdeV57mtY7G9OnX39Tf4ksFtnr9Sd9dtTXu7l3ZnV9TOvIFlu294yXNtdObctHe5fN37OyO37c1eb+M1WOhx7ArdOSwr33433pvZQwg6ck135s1AXS233WnPd0ju2+A6OmStVN/vjdstZMhjW59t7zroyPa6s53s7msy39sE2yNf+cb6t7UPbX2ee212dPh0d8jdblu/y4pZx7bu6dZ+H82aNavTw6+399urI8OO89V/R/q9UvDv6Zxrekf/b6XtKdZ12xW/D4o5VLC1a7LQ/9bqyHlftqyqZI9G2BH+e14gBAAAAJAYgRAAAABAYgRCrch+S1B2d7ju6ErW2e7JHX0bQld0883t/tYdx62YXVdbO275hkA0DR3phjcTLVu2/e7frS3XqD1DrtrqvljIWwzarFsXdsfvyHXX0eGbXanpPFRt/7x1tHt4u4aqFmmYRLu69ObZTltDHzq0/azyjfdt7npyh4HlO8YdHWLa3ffE9oaKdbZL96xZs5qu/9x1FeP7o1hDDzq7nsZ7pPGYtHuoWwFtQzHbvdau2/YOMe/IMMtivTmvPce2PcOdtrftRl25j7naO+Qr+55qrWzjZ7nnd1lV68MbCv29Vehxzndsc89bR+uU3ZY2tj8RbQ/jzaet66c9b6nLV69iKWR4Slf+lm5vG9Ydv4Xac//kK5N7TAs9X437mLufHd3vtt5Cm7v+7H8v1ndZR+T+hmitDrnfk9lae1xHvuNZcDuVs572vOWuUI3rLHR4WjHvk64ezlbyQGjOnDlRVVUV/fr1i9GjR8cjjzzSZvmHH344Ro8eHf369YsPfOADccMNN3RTTQEAAAB2DiUNhBYsWBDTp0+PK664IhYvXhzHHHNMjB8/PmpqavKWX7p0aUyYMCGOOeaYWLx4cfznf/5nXHTRRfHrX/+6m2sOAAAA0HOVNBCaPXt2TJkyJaZOnRojR46Ma6+9NiorK2Pu3Ll5y99www2x3377xbXXXhsjR46MqVOnxhe/+MX4wQ9+UNR6tadLVr6nkecbntCoPV33OtIVbHtPYW+rq19u2ablu+FNGq11MWzt78buhx3pEtjetyY0nr/c7RdzCETjcc4egpDbpXJ73S63t43tDTHLXaYr5Xbzzq1H9tCTtpbPfVtRPi2627Zx/+UdOpSnDk3XRBt1bO3cNNa3tTo3DX9o7Dpbte2YFKI9w4Iar61Cu5m2517IblM6MoQhX1fs1q6b3HW1GD7RjcP58m1/e927c8977ttPWlu+s+cvXx0bt9PaOckeFtYdb5tpa7+z1537Brm2rs3G5fO9dS67LdreevJp6zrL/b5qj/YM6c6dGue3tnybdWzjbUsdeUtYW29Laupmn+c3Re61v2xZy+EFbX1ft/VZe/ajLV3xW6C134n5jk0p3m7TXp0Zhpb7/Zv979m/CTr7nVjo+erK7472Dh1sTZtvMssZUtme/cjbJpbozY6tDf3pqI625dnHrfG7uXH+9r6LG7fX1u/cxrq0NlysLZ0d6lVK7b2O2vP4g/b8900+rZ2b7DcU5x7jYj42pbUhvtly/9sv9+2l7X2rY7Zi/Abu3amlO2HTpk3x9NNPx+WXX95s/rhx4+Lxxx/Pu8wTTzwR48aNazbvpJNOinnz5sXmzZujT58+LZbZuHFjbNy4senvurq6iIior69vtW4NDWsjszETEREbNmxoVr6hYW3U19fH2rUN0dCwttlnaxsamj5rXG7t2oaIiMhszER9fX1s2LAh6uvro6FhbayNhmbbaFr3/60ne/nGvxvXk9mYafHZ2rUNTZ81rrNR7vKNZZvNa2h9+cZ1ZM/P3bfGfzY0rG36LHffsj9rPMbZx62xHrkaj1vjP7PXnXtsMxszTevIPo+NnzWet6b9iIZm10O+/c/eRu7+Z18T+c5/4343Lp9bt9zlc+ufu+7/u2xa7H/2vOw6Zi+/YcOGFtdt4/abnZO1285B7r2SWyb73OSTu/3sdW+tQLSof2Mds89347YbP2tx/Tc0tLhv891Tuecm+7priObrzq5/dn0b65J73Wefk9xj03Rdr922nrau6cZ9yj5P2WUa69bWvZF932WvM3v57P1oOjY511/2PZO7/exrNPfcNJbJ3n6j7OOee99kH/fce6LxHGVf97ltStPy+drkhubXbfYy2fXP3n52/bOPReM2Wmsvs9vW7PPe+Hnjtlrctw3N2418bWq+85a7/dbuzdx15x7H3Gsq+75tnJfv+yL3vm2rTchXJvsc536We26bnf+G/N83uddc4z7mfl801ju3Tc7e79xrI7vdbpLVNufeky3a+4bW2+Tsuubbh+z9z7d8s/smt73Nc/xy26t8y+f7vs2933P3O9+9kb3frd232ctvO7Qtr63ce7LxuzzfddtYNvcY5atbs99ZDa3sW577Nt893ervxKw2ocX3VZ51Z1+TLb6vonkds7+Lctv7fL8FGpdvKtPQsv75lm+U7/zl/o7Lvqdb+y2V/Z2YK99vmdbahOzfSdtWsK1Mq+tu57WZu9+5v+Ha+g2cfdxzt5/v79zrNnvd2fvY2rnJXWfe/W9oeWyztfVdkK9M7m+JFt8zub+F8hzb3Hq35zul2e/DnHui2W/Q3P/GWNvyus/3G6a1ezP7+LXWXjfK/jzf78zc37DZ+5Z7bWUft9bObb7v8ty65WtTc7efLd/3ZG57ne/eaJTbpmTXNXf7ze7pPLZ3bWRf97nf843L5/sN29pv0Xyf5ftOyG0T2jp+2ee0te/yfJ/l7me+z5rakUzLa6OFTImsWLEiExGZxx57rNn8b3/725kRI0bkXeaggw7KfPvb324277HHHstERGblypV5l7nyyiszEWEymUwmk8lkMplMJpPJlMT0+uuvbzeXKVkPoUZlZWXN/s5kMi3mba98vvmNZs6cGTNmzGj6u6GhId5+++3YY4892twOAAAAQE+SyWRi7dq1sffee2+3bMkCoSFDhkSvXr1i1apVzeavXr06hg0blneZvfbaK2/53r17xx577JF3mb59+0bfvn2bzdt9990LrzgAAADADqqioqJd5Ur2UOny8vIYPXp0LFq0qNn8RYsWxZFHHpl3mbFjx7Yo/8ADD8SYMWPyPj8IAAAAgJZK+paxGTNmxM9+9rP4+c9/Hi+++GJccsklUVNTE9OmTYuIrcO9Jk2a1FR+2rRpsXz58pgxY0a8+OKL8fOf/zzmzZsXl156aal2AQAAAKDHKekzhCZOnBhr1qyJq666Kmpra2PUqFGxcOHCGD58eERE1NbWRk1NTVP5qqqqWLhwYVxyySXxk5/8JPbee++4/vrr49Of/nSpdgEAAACgxynLZNrzLjIAAAAAdhYlHTIGALCzWLZsWZSVlcWSJUvaLPexj30spk+f3i11AgBojUAIAEjK5MmTo6ysLMrKyqJPnz7xgQ98IC699NJYv359p9ZbWVnZNAQ+IuKhhx6KsrKy+J//+Z9m5e6+++745je/2altAQB0VkmfIQQAUAonn3xyVFdXx+bNm+ORRx6JqVOnxvr162Pu3LkFr7NXr16x1157bbfc+973voK3AQBQLHoIAQDJ6du3b+y1115RWVkZZ5xxRpx55pnx29/+NjZu3BgXXXRRDB06NPr16xdHH310/Nd//VfTcu+8806ceeaZseeee0b//v3joIMOiurq6ohoPmRs2bJlcfzxx0dExODBg6OsrCwmT54cES2HjL3zzjsxadKkGDx4cAwYMCDGjx8fr776atPnN998c+y+++5x//33x8iRI2PXXXeNk08+OWpra5vKPPTQQ/HRj340Bg4cGLvvvnscddRRsXz58i48ggBATycQAgCS179//9i8eXN87Wtfi1//+tdxyy23xDPPPBMHHnhgnHTSSfH2229HRMTXv/71eOGFF+L3v/99vPjiizF37twYMmRIi/VVVlbGr3/964iIePnll6O2tjauu+66vNuePHlyPPXUU3HPPffEE088EZlMJiZMmBCbN29uKvPuu+/GD37wg7j11lvjz3/+c9TU1MSll14aERHvvfdenHbaaXHcccfFc889F0888USce+65UVZWVuzDBADsRAwZAwCS9te//jVuv/32OP7442Pu3Llx8803x/jx4yMi4qc//WksWrQo5s2bF5dddlnU1NTEYYcdFmPGjImIiP333z/vOnv16tU0NGzo0KGx++675y336quvxj333BOPPfZYHHnkkRER8Ytf/CIqKyvjt7/9bXzmM5+JiIjNmzfHDTfcEAcccEBERFxwwQVx1VVXRUREfX191NXVxSc+8Ymmz0eOHNn5AwMA7NT0EAIAknPvvffGrrvuGv369YuxY8fGscceGxdeeGFs3rw5jjrqqKZyffr0iY9+9KPx4osvRkTEl7/85bjzzjvjIx/5SHzta1+Lxx9/vFP1ePHFF6N3795xxBFHNM3bY4894uCDD27aZkTEgAEDmsKeiIj3v//9sXr16ojY+kyiyZMnx0knnRSf/OQn47rrrms2nAwAIB+BEACQnOOPPz6WLFkSL7/8cmzYsCHuvvvuqKioiIhoMdQqk8k0zRs/fnwsX748pk+fHitXrowTTjihaehWITKZTKvzs+vRp0+fZp+XlZU1W7a6ujqeeOKJOPLII2PBggUxYsSIePLJJwuuFwCw8xMIAQDJGThwYBx44IExfPjwprDlwAMPjPLy8nj00Uebym3evDmeeuqpZkOw9txzz5g8eXLcdtttce2118ZNN92Udxvl5eUREbFly5ZW63HIIYfEe++9F3/5y1+a5q1ZsyZeeeWVDg/7Ouyww2LmzJnx+OOPx6hRo+L222/v0PIAQFoEQgAAsTUk+vKXvxyXXXZZ3HffffHCCy/El770pXj33XdjypQpERHxjW98I373u9/F3//+9/jb3/4W9957b6vBzfDhw6OsrCzuvffeePPNN2PdunUtyhx00EFx6qmnxpe+9KV49NFH49lnn42zzjor9tlnnzj11FPbVe+lS5fGzJkz44knnojly5fHAw88UFCgBACkRSAEAPB/rrnmmvj0pz8dZ599dhx++OHx97//Pe6///4YPHhwRGzt9TNz5sz40Ic+FMcee2z06tUr7rzzzrzr2meffWLWrFlx+eWXx7Bhw+KCCy7IW666ujpGjx4dn/jEJ2Ls2LGRyWRi4cKFLYaJtWbAgAHx0ksvxac//ekYMWJEnHvuuXHBBRfEeeedV9hBAACSUJZpbfA6AAAAADslPYQAAAAAEiMQAgAAAEiMQAgAAAAgMQIhAAAAgMQIhAAAAAASIxACAAAASIxACAAAACAxAiEAAACAxAiEAAAAABIjEAIAAABIjEAIAAAAIDH/H4uaKjiJ4qIDAAAAAElFTkSuQmCC",
      "text/plain": [
       "<Figure size 1400x700 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "colors_for_bases = {\"-\" : \"grey\", \"A\" : \"red\", \"T\" : \"blue\", \"C\" : \"green\", \"G\" : \"yellow\", \"None\" : \"black\"}\n",
    "\n",
    "freq_dist = freq_dist.rename(columns={0:\"Base\", 1:\"Frequency\"})\n",
    "\n",
    "freq_dist[\"Base\"] = freq_dist[\"Base\"].fillna(\"None\")\n",
    "\n",
    "freq_dist = freq_dist.dropna()\n",
    "\n",
    "freq_dist[\"Color\"] = freq_dist[\"Base\"].map(colors_for_bases)\n",
    "\n",
    "\n",
    "\n",
    "# Plotting\n",
    "plt.figure(figsize=(14, 7))\n",
    "sns.barplot(x=freq_dist.index, y=freq_dist[\"Frequency\"], palette=freq_dist[\"Color\"])\n",
    "#plt.title(\"Highest Non-Reference Base Frequency \")\n",
    "plt.xticks([])\n",
    "plt.xlabel(\"Positions\")\n",
    "plt.ylabel(\"Frequency\")\n",
    "# TODO Add horizontal line at 2 mean std deviation\n",
    "plt.axhline(y=0.3, color='b', linestyle='--')\n",
    "# TODO Add legend for bases\n",
    "nb_positions = []\n",
    "for index, row in freq_dist.iterrows():\n",
    "    if row[\"Frequency\"] > 0.3:\n",
    "        nb_positions.append(index)\n",
    "        plt.text(index, row[\"Frequency\"], f\"{index}{row['Base']}\", color='black', ha=\"center\", va='bottom')\n",
    "\n",
    "plt.show()\n",
    "\n",
    "#plt.savefig(\"/home/emre/github_repo/MinION/examples/figures/PgA9_variant_freq_RB02_NB92.png\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_nb_positions(freq_dist, threshold):\n",
    "    nb_positions = []\n",
    "    for index, row in freq_dist.iterrows():\n",
    "        if row[\"Frequency\"] > threshold:\n",
    "            nb_positions.append(index)\n",
    "    return nb_positions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bases_df = get_bases_from_pileup(bam_file, reference, nb_positions)\n",
    "\n",
    "heatmap_data = generate_heatmap_data(bam_file, reference, nb_positions)\n",
    "pval_matrix = np.zeros((len(nb_positions), len(nb_positions)))\n",
    "odds_ratio_matrix = np.zeros((len(nb_positions), len(nb_positions)))\n",
    "jaccard_index_matrix = np.zeros((len(nb_positions), len(nb_positions)))\n",
    "overlap_matrix = np.zeros((len(nb_positions), len(nb_positions)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Generate combinations of positions and initialize storage structures\n",
    "variant = []\n",
    "if nb_positions is not None:\n",
    "    heatmap_data = generate_heatmap_data(bam_file, reference, nb_positions)\n",
    "    position_combinations = list(combinations(nb_positions, 2))\n",
    "    jaccard_indices = {\"Pair\": [],\"Jaccard Index\": [],}\n",
    "    bases_df = get_bases_from_pileup(bam_file, reference, nb_positions)\n",
    "\n",
    "    if len(nb_positions) == 1:\n",
    "\n",
    "        variant = f\"{template[nb_positions[0] - 1]}{nb_positions[0]}{get_most_common_base([0], heatmap_data, template[nb_positions[0] - 1])[0]}\"\n",
    "\n",
    "\n",
    "    elif len(nb_positions) > 1:\n",
    "        # Process each pair of positions\n",
    "        for pos1, pos2 in position_combinations:\n",
    "            common_base1, common_base2 = [get_most_common_base(pos, heatmap_data, template[pos - 1])[0] for pos in [pos1, pos2]]\n",
    "            ref_bases = [template[pos - 1] for pos in [pos1, pos2]]\n",
    "\n",
    "            counts = {}\n",
    "            for idx, (base1, base2) in enumerate([(common_base1, common_base2), (ref_bases[0], common_base2), (common_base1, ref_bases[1]), (ref_bases[0], ref_bases[1])]):\n",
    "                counts[idx] = len(bases_df[(bases_df[pos1] == base1) & (bases_df[pos2] == base2)]) + 1\n",
    "\n",
    "            # Calculate Jaccard indices\n",
    "            jaccard_dict = {f\"A_{i}_and_B_{j}\": counts[k] / (counts[k] + counts[(k + 1) % 4] + counts[(k + 2) % 4]) for k, (i, j) in enumerate([('mut', 'mut'), ('wt', 'mut'), ('mut', 'wt'), ('wt', 'wt')])}\n",
    "\n",
    "            jaccard_indices[\"Pair\"].append((pos1, pos2))\n",
    "            jaccard_indices[\"Jaccard Index\"].append(jaccard_dict)\n",
    "\n",
    "            jaccard_index_matrix[nb_positions.index(pos1), nb_positions.index(pos2)] = jaccard_dict[\"A_mut_and_B_mut\"]\n",
    "\n",
    "            contingency_table = pd.DataFrame({\n",
    "                \"mut_A\": [jaccard_dict[\"A_mut_and_B_mut\"], jaccard_dict[\"A_mut_and_B_wt\"]],\n",
    "                \"wt_A\": [jaccard_dict[\"A_wt_and_B_mut\"], jaccard_dict[\"A_wt_and_B_wt\"]],\n",
    "            }, index=['mut_B', 'wt_B'])\n",
    "\n",
    "        mutation_details = {\"Type\": [], \"Position\": [], \"Variant\": [], \"Frequency\": []}\n",
    "        for i, (pair, jaccard_dict) in enumerate(zip(jaccard_indices[\"Pair\"], jaccard_indices[\"Jaccard Index\"])):\n",
    "            max_type = max(jaccard_dict, key=jaccard_dict.get)\n",
    "            if jaccard_dict[max(jaccard_dict, key=jaccard_dict.get)] > 0.7 and max_type != \"A_wt_and_B_wt\":\n",
    "\n",
    "                ref_bases = [template[pos - 1] for pos in pair]\n",
    "                common_base1, common_base2 = [get_most_common_base(pos, heatmap_data, template[pos - 1])[0] for pos in pair]\n",
    "                if max_type == \"A_mut_and_B_mut\":\n",
    "                    mutation_details[\"Type\"].extend([\"Co-Occuring\"])\n",
    "                    mutation_details[\"Position\"].extend(pair)\n",
    "                    mutation_details[\"Variant\"].extend([f'{ref_bases[0]}{pair[0]}{common_base1}', f'{ref_bases[1]}{pair[1]}{common_base2}'])\n",
    "                    mutation_details[\"Frequency\"].extend([jaccard_dict[\"A_mut_and_B_mut\"], jaccard_dict[\"A_mut_and_B_mut\"]])\n",
    "                elif max_type == \"A_mut_and_B_wt\":\n",
    "                    mutation_details[\"Type\"].extend([\"A\"])\n",
    "                    mutation_details[\"Position\"].append(pair[0])\n",
    "                    mutation_details[\"Variant\"].append(f'{ref_bases[0]}{pair[0]}{common_base1}')\n",
    "                    mutation_details[\"Frequency\"].append(jaccard_dict[\"A_mut_and_B_wt\"])\n",
    "                elif max_type == \"A_wt_and_B_mut\":\n",
    "                    mutation_details[\"Type\"].extend([\"B\"])\n",
    "                    mutation_details[\"Position\"].append(pair[1])\n",
    "                    mutation_details[\"Variant\"].append(f'{ref_bases[1]}{pair[1]}{common_base2}')\n",
    "                    mutation_details[\"Frequency\"].append(jaccard_dict[\"A_wt_and_B_mut\"])\n",
    "            \n",
    "\n",
    "\n",
    "        # Create a DataFrame from the mutation details of Position and Variant only\n",
    "        mutation_details_df = pd.DataFrame(mutation_details, columns=[\"Position\", \"Variant\", \"Frequency\"])\n",
    "\n",
    "\n",
    "        mutation_details_df = mutation_details_df.sort_values(by=[\"Position\"])\n",
    "        mutation_details_df = mutation_details_df[\"Variant\"].drop_duplicates()\n",
    "\n",
    "        # TODO: Try to explain why we set up a threshold of 10\n",
    "        if len(mutation_details_df) > 10:\n",
    "            variant = \"NA\"\n",
    "        else:\n",
    "\n",
    "            variant = \"_\".join(list(mutation_details_df))\n",
    "\n",
    "        #Plot heatmap for Jaccard indices\n",
    "        plt.figure(figsize=(6, 4))\n",
    "        sns.heatmap(jaccard_index_matrix, cmap='coolwarm', mask=np.tri(len(nb_positions), k=-1), linewidths=0.5, linecolor='white')\n",
    "        plt.xlabel('Position')\n",
    "        plt.ylabel('Position')\n",
    "        plt.title('Jaccard Index Heatmap')\n",
    "else:\n",
    "    variant = \"#PARENT#\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate the frequency of each unique row\n",
    "frequency_df = bases_df.apply(lambda row: '-'.join(row), axis=1).value_counts().reset_index()\n",
    "frequency_df.columns = ['Population', 'Frequency']\n",
    "frequency_df[\"Frequency\"] = frequency_df[\"Frequency\"] / frequency_df[\"Frequency\"].sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import tqdm\n",
    "\n",
    "def get_variant_name(entry, reference, nb_positions):\n",
    "    variant = []\n",
    "    for i, mut in enumerate(entry):\n",
    "        \n",
    "        ref_AA = reference[nb_positions[i] - 1]\n",
    "        \n",
    "        if mut == \"-\":\n",
    "            v = f'{ref_AA}{nb_positions[i]}DEL'\n",
    "        \n",
    "        elif mut == ref_AA:\n",
    "            continue\n",
    "\n",
    "        else:\n",
    "            v = f'{ref_AA}{nb_positions[i]}{mut}'\n",
    "\n",
    "        variant.append(v)\n",
    "\n",
    "    return \"_\".join(variant)\n",
    "\n",
    "def get_nb_positions(freq_dist, threshold):\n",
    "    nb_positions = []\n",
    "    for index, row in freq_dist.iterrows():\n",
    "        if row[\"Frequency\"] > threshold:\n",
    "            nb_positions.append(index)\n",
    "    return nb_positions\n",
    "\n",
    "def get_pop_frequency(bam_file, template, reference, nb_positions, min_freq=0.1, min_depth= 15):\n",
    "\n",
    "    # Get PileUP\n",
    "    bases_df = get_bases_from_pileup(bam_file, reference, nb_positions)\n",
    "    \n",
    "    frequency_df = bases_df.apply(get_variant_name, axis=1, args=(template, nb_positions)).value_counts().reset_index()\n",
    "\n",
    "    frequency_df.columns = ['Population', 'N_reads']\n",
    "    \n",
    "    frequency_df[\"Frequency\"] = frequency_df[\"N_reads\"] / frequency_df[\"N_reads\"].sum()\n",
    "\n",
    "    # Filter for frequency > 0.1\n",
    "    frequency_df = frequency_df[(frequency_df[\"Frequency\"] > min_freq) & (frequency_df[\"N_reads\"] > min_depth)]\n",
    "\n",
    "    return frequency_df\n",
    "\n",
    "def call_variant_pop_frequency(bam_file, template, reference, min_freq=0.1, min_depth= 15):\n",
    "    \"\"\" \n",
    "    Call Variant from BAM file based on Basecall Frequency & Alignment Frequency\n",
    "\n",
    "    Args:\n",
    "        - bam_file (str): Path to the BAM file.\n",
    "        - template (str): Template name >NAME from fasta or fastq files.\n",
    "        - reference (str): Reference name >NAME from fasta or fastq files.\n",
    "        - nb_positions (list): List of positions to call variants.\n",
    "        - min_freq (float): Minimum frequency of the variant.\n",
    "        - min_depth (int): Minimum depth of the variant.\n",
    "    \n",
    "    Returns:\n",
    "        - variant (str): Variant name.\n",
    "    \"\"\"\n",
    "\n",
    "    variants = {\"Variant\" : [], \"Position\" : [], \"Alignment Count\" : [], \"Alignment Frequency\" : []}\n",
    "\n",
    "\n",
    "    try:\n",
    "        alignment_count = int(subprocess.run(f\"samtools view -c {bam_file}\", shell=True, capture_output=True).stdout.decode(\"utf-8\").strip())\n",
    "        freq_dist = pd.DataFrame(analyser.get_highest_non_ref_base_freq(bam_file, reference, range(1,len(template)), template)[0]).T.rename(columns={0:\"Base\", 1:\"Frequency\"})\n",
    "\n",
    "        nb_positions = get_nb_positions(freq_dist, 0.4)\n",
    "\n",
    "        if nb_positions == [] and alignment_count > min_depth:\n",
    "            variants[\"Variant\"].append(\"#PARENT#\")\n",
    "            variants[\"Position\"].append(\"-\")\n",
    "            variants[\"Alignment Count\"].append(alignment_count)\n",
    "            variants[\"Alignment Frequency\"].append(\"-\")\n",
    "            \n",
    "\n",
    "        elif nb_positions == [] and alignment_count < min_depth:\n",
    "            variants[\"Variant\"].append(\"NA\")\n",
    "            variants[\"Position\"].append(\"-\")\n",
    "            variants[\"Alignment Count\"].append(alignment_count)\n",
    "            variants[\"Alignment Frequency\"].append(\"-\")\n",
    "\n",
    "        elif len(nb_positions) == 1 and alignment_count > min_depth:\n",
    "\n",
    "            ref_base = template[nb_positions[0] - 1]\n",
    "            pos = nb_positions[0]\n",
    "            base = freq_dist[\"Base\"][nb_positions[0] - 1]\n",
    "\n",
    "            variants[\"Variant\"].append(f\"{ref_base}{pos}{base}\")\n",
    "            variants[\"Position\"].append(nb_positions[0])\n",
    "            variants[\"Alignment Count\"].append(alignment_count)\n",
    "            variants[\"Alignment Frequency\"].append(freq_dist[\"Frequency\"][nb_positions[0] - 1])\n",
    "\n",
    "           \n",
    "        else:\n",
    "\n",
    "            freq_df = get_pop_frequency(bam_file, template, reference, nb_positions, min_freq=0.1)\n",
    "\n",
    "            for index, row in freq_df.iterrows():\n",
    "                variant = row[\"Population\"]\n",
    "                variants[\"Variant\"].append(variant)\n",
    "                variants[\"Position\"].append(nb_positions)\n",
    "                variants[\"Alignment Count\"].append(row[\"N_reads\"])\n",
    "                variants[\"Alignment Frequency\"].append(row[\"Frequency\"])\n",
    "\n",
    "    except:\n",
    "        print(\"Error in calling variant\")\n",
    "        variants[\"Variant\"].append(\"NA\")\n",
    "        variants[\"Position\"].append(\"-\")\n",
    "\n",
    "    return variants\n",
    "\n",
    "def get_variant_df_AF(demultiplex_folder: Path, ref_seq : Path, ref_name : str, barcode_dicts : dict = None, consensus_folder_name = \"consensus\", merge = True, min_freq=0.1, min_depth= 15):\n",
    "\n",
    "    if barcode_dicts is None:\n",
    "        barcode_dicts = IO_processor.get_barcode_dict(demultiplex_folder)\n",
    "    \n",
    "    variant_template_df = analyser.template_df(barcode_dicts, rowwise=False)\n",
    "\n",
    "    variants = {\"RBC\": [], \"FBC\": [], \"Position\": [], \"Variant\": [], \"Alignment Count\": [], \"Alignment Frequency\": []}\n",
    "\n",
    "    template = analyser.get_template_sequence(ref_seq) # Reference sequence\n",
    "\n",
    "    summary = analyser.read_summary_file(demultiplex_folder)\n",
    "    n_counts = summary.groupby([\"RBC\",\"FBC\"])[\"FBC\"].value_counts().reset_index() \n",
    "\n",
    "\n",
    "\n",
    "    for barcode_id, barcode_dict in barcode_dicts.items():\n",
    "\n",
    "        rbc = os.path.basename(barcode_id)\n",
    "\n",
    "        for front_barcode in barcode_dict:\n",
    "\n",
    "            fbc = os.path.basename(front_barcode)\n",
    "\n",
    "            count = n_counts[ (n_counts[\"RBC\"] == rbc) & (n_counts[\"FBC\"] == fbc)][\"count\"].values[0]\n",
    "\n",
    "            bam_file = front_barcode / consensus_folder_name / \"alignment.bam\"\n",
    "\n",
    "            if not bam_file.exists():\n",
    "                print(f\"{bam_file} does not exist.\")\n",
    "                variants[\"RBC\"].append(rbc)\n",
    "                variants[\"FBC\"].append(fbc)\n",
    "                variants[\"Position\"].append([\"NA\"])\n",
    "                variants[\"Variant\"].append([\"NA\"])\n",
    "                #variants[\"Reads\"].append(count)\n",
    "                variants[\"Alignment Count\"].append([\"NA\"])\n",
    "                variants[\"Alignment Frequency\"].append([\"NA\"])\n",
    "                continue\n",
    "\n",
    "            try:\n",
    "                \n",
    "                nn_variants = call_variant_pop_frequency(bam_file, template, ref_name, min_freq, min_depth) #Â TODO read header from reference file for name\n",
    "\n",
    "                for i, variant in enumerate(nn_variants[\"Variant\"]):\n",
    "                    variants[\"RBC\"].append(rbc)\n",
    "                    variants[\"FBC\"].append(fbc)\n",
    "                    variants[\"Position\"].append(nn_variants[\"Position\"][i])\n",
    "                    variants[\"Variant\"].append(variant)\n",
    "                    #variants[\"Reads\"].append(count)\n",
    "                    variants[\"Alignment Count\"].append(nn_variants[\"Alignment Count\"][i])\n",
    "                    variants[\"Alignment Frequency\"].append(nn_variants[\"Alignment Frequency\"][i])\n",
    "            \n",
    "            except:\n",
    "                print(f\"Skipping {rbc}/{fbc}\")\n",
    "                continue\n",
    "\n",
    "    variant_df = analyser.rename_barcode(pd.DataFrame(variants).merge(n_counts, on=[\"RBC\",\"FBC\"] , how=\"left\"))\n",
    "\n",
    "    if merge:\n",
    "        return variant_df.merge(variant_template_df, on=[\"Plate\", \"Well\"], how=\"right\")\n",
    "    else:\n",
    "        return variants\n",
    "\n",
    "\n",
    "\n",
    "def get_mixed_population(variant_df):\n",
    "\n",
    "    # Get row that are not unique\n",
    "    mixed_pop = variant_df[variant_df.duplicated(subset=[\"Plate\", \"Well\"], keep=False)]\n",
    "\n",
    "    return mixed_pop\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "path_to_bam = \"/home/emre/minION_results/20231119_1502_MN41105/Demultiplex_cpp_70/RB01/NB94/consensus\"\n",
    "bam_file = f'{path_to_bam}/alignment.bam'\n",
    "template = analyser.get_template_sequence(Path(\"/home/emre/PgA9.fasta\"))\n",
    "reference = \"PgA9\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/emre/minION_results/20231119_1502_MN41105/Demultiplex_cpp_70/RB03/NB44/consensus/alignment.bam does not exist.\n",
      "/home/emre/minION_results/20231119_1502_MN41105/Demultiplex_cpp_70/RB03/NB43/consensus/alignment.bam does not exist.\n"
     ]
    }
   ],
   "source": [
    "variant_df = get_variant_df_AF(Path(\"/home/emre/minION_results/20231119_1502_MN41105/Demultiplex_cpp_70\"), Path(\"/home/emre/PgA9.fasta\"), ref_name =  \"PgA9\", min_freq=0.1, min_depth= 15)\n",
    "\n",
    "\n",
    "r"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "variant_df.to_csv(\"data/PgA9_variant_df.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "variant_df = pd.read_csv(\"data/PgA9_variant_df.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>Plate</th>\n",
       "      <th>Well</th>\n",
       "      <th>Position</th>\n",
       "      <th>Variant</th>\n",
       "      <th>Reads</th>\n",
       "      <th>Alignment Count</th>\n",
       "      <th>Alignment Frequency</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>93</td>\n",
       "      <td>1</td>\n",
       "      <td>H10</td>\n",
       "      <td>[14, 62, 232, 446, 593]</td>\n",
       "      <td>A14G_A62G_A232G</td>\n",
       "      <td>398.0</td>\n",
       "      <td>154</td>\n",
       "      <td>0.3869346733668342</td>\n",
       "      <td>398.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>94</td>\n",
       "      <td>1</td>\n",
       "      <td>H10</td>\n",
       "      <td>[14, 62, 232, 446, 593]</td>\n",
       "      <td>A446T_A593DEL</td>\n",
       "      <td>398.0</td>\n",
       "      <td>133</td>\n",
       "      <td>0.3341708542713568</td>\n",
       "      <td>398.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163</th>\n",
       "      <td>163</td>\n",
       "      <td>2</td>\n",
       "      <td>F7</td>\n",
       "      <td>[200, 319]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1141.0</td>\n",
       "      <td>545</td>\n",
       "      <td>0.4776511831726556</td>\n",
       "      <td>1141.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>164</th>\n",
       "      <td>164</td>\n",
       "      <td>2</td>\n",
       "      <td>F7</td>\n",
       "      <td>[200, 319]</td>\n",
       "      <td>A200G_C319T</td>\n",
       "      <td>1141.0</td>\n",
       "      <td>533</td>\n",
       "      <td>0.4671340929009641</td>\n",
       "      <td>1141.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>172</th>\n",
       "      <td>172</td>\n",
       "      <td>2</td>\n",
       "      <td>G3</td>\n",
       "      <td>[41, 393]</td>\n",
       "      <td>A393G</td>\n",
       "      <td>914.0</td>\n",
       "      <td>405</td>\n",
       "      <td>0.4431072210065646</td>\n",
       "      <td>914.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173</th>\n",
       "      <td>173</td>\n",
       "      <td>2</td>\n",
       "      <td>G3</td>\n",
       "      <td>[41, 393]</td>\n",
       "      <td>T41C</td>\n",
       "      <td>914.0</td>\n",
       "      <td>383</td>\n",
       "      <td>0.4190371991247265</td>\n",
       "      <td>914.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>225</th>\n",
       "      <td>225</td>\n",
       "      <td>3</td>\n",
       "      <td>C7</td>\n",
       "      <td>[120, 219]</td>\n",
       "      <td>C120T_T219C</td>\n",
       "      <td>414.0</td>\n",
       "      <td>264</td>\n",
       "      <td>0.6376811594202898</td>\n",
       "      <td>414.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>226</th>\n",
       "      <td>226</td>\n",
       "      <td>3</td>\n",
       "      <td>C7</td>\n",
       "      <td>[120, 219]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>414.0</td>\n",
       "      <td>127</td>\n",
       "      <td>0.30676328502415456</td>\n",
       "      <td>414.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>243</th>\n",
       "      <td>243</td>\n",
       "      <td>3</td>\n",
       "      <td>D12</td>\n",
       "      <td>[239, 458]</td>\n",
       "      <td>A239G_T458C</td>\n",
       "      <td>243.0</td>\n",
       "      <td>146</td>\n",
       "      <td>0.6008230452674898</td>\n",
       "      <td>243.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>244</th>\n",
       "      <td>244</td>\n",
       "      <td>3</td>\n",
       "      <td>D12</td>\n",
       "      <td>[239, 458]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>243.0</td>\n",
       "      <td>70</td>\n",
       "      <td>0.2880658436213992</td>\n",
       "      <td>243.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>365</th>\n",
       "      <td>365</td>\n",
       "      <td>4</td>\n",
       "      <td>G1</td>\n",
       "      <td>[185, 367, 520]</td>\n",
       "      <td>G185T_T367C_T520C</td>\n",
       "      <td>588.0</td>\n",
       "      <td>259</td>\n",
       "      <td>0.44047619047619047</td>\n",
       "      <td>588.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>366</th>\n",
       "      <td>366</td>\n",
       "      <td>4</td>\n",
       "      <td>G1</td>\n",
       "      <td>[185, 367, 520]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>588.0</td>\n",
       "      <td>251</td>\n",
       "      <td>0.4268707482993197</td>\n",
       "      <td>588.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Unnamed: 0  Plate Well                 Position            Variant  \\\n",
       "93           93      1  H10  [14, 62, 232, 446, 593]    A14G_A62G_A232G   \n",
       "94           94      1  H10  [14, 62, 232, 446, 593]      A446T_A593DEL   \n",
       "163         163      2   F7               [200, 319]                NaN   \n",
       "164         164      2   F7               [200, 319]        A200G_C319T   \n",
       "172         172      2   G3                [41, 393]              A393G   \n",
       "173         173      2   G3                [41, 393]               T41C   \n",
       "225         225      3   C7               [120, 219]        C120T_T219C   \n",
       "226         226      3   C7               [120, 219]                NaN   \n",
       "243         243      3  D12               [239, 458]        A239G_T458C   \n",
       "244         244      3  D12               [239, 458]                NaN   \n",
       "365         365      4   G1          [185, 367, 520]  G185T_T367C_T520C   \n",
       "366         366      4   G1          [185, 367, 520]                NaN   \n",
       "\n",
       "      Reads Alignment Count  Alignment Frequency   count  \n",
       "93    398.0             154   0.3869346733668342   398.0  \n",
       "94    398.0             133   0.3341708542713568   398.0  \n",
       "163  1141.0             545   0.4776511831726556  1141.0  \n",
       "164  1141.0             533   0.4671340929009641  1141.0  \n",
       "172   914.0             405   0.4431072210065646   914.0  \n",
       "173   914.0             383   0.4190371991247265   914.0  \n",
       "225   414.0             264   0.6376811594202898   414.0  \n",
       "226   414.0             127  0.30676328502415456   414.0  \n",
       "243   243.0             146   0.6008230452674898   243.0  \n",
       "244   243.0              70   0.2880658436213992   243.0  \n",
       "365   588.0             259  0.44047619047619047   588.0  \n",
       "366   588.0             251   0.4268707482993197   588.0  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_mixed_population(variant_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = variant_df.dropna(subset=[\"Variant\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count     362.000000\n",
       "mean      816.414365\n",
       "std       636.440173\n",
       "min        18.000000\n",
       "25%       364.750000\n",
       "50%       699.000000\n",
       "75%      1107.750000\n",
       "max      5602.000000\n",
       "Name: count, dtype: float64"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[\"count\"].describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(len(variants[\"RBC\"]))\n",
    "print(len(variants[\"FBC\"]))\n",
    "#print(len(variants[\"Position\"]))\n",
    "print(len(variants[\"Variant\"]))\n",
    "print(len(variants[\"Alignment Count\"]))\n",
    "print(len(variants[\"Alignment Frequency\"]))\n",
    "\n",
    "test_df = pd.DataFrame({\"RBC\": variants[\"RBC\"], \"FBC\": variants[\"FBC\"], \"Variant\": variants[\"Variant\"], \"Alignment Count\": variants[\"Alignment Count\"], \"Alignment Frequency\": variants[\"Alignment Frequency\"]})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "test_df[test_df[\"RBC\"] == \"RB01\"].sort_values(by=[\"FBC\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "barcode_dicts = IO_processor.get_barcode_dict(Path(\"/home/emre/minION_results/TamLQV96_sup/Demultiplex_cpp_70\"))\n",
    "demultiplex_folder = Path(\"/home/emre/minION_results/TamLQV96_sup/Demultiplex_cpp_70\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bases, qualities = analyser.get_highest_non_ref_base_freq(bam_file, reference, range(1,len(ref_seq)), ref_seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "variant_df = analyser.get_variant_df_custom(demultiplex_folder, template_fasta, barcode_dicts, consensus_folder_name = \"consensus\", merge = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "variant_df.dropna(subset=[\"count\"], inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def single_plate_annotation(entry):\n",
    "    row = [\"A\", \"B\", \"C\", \"D\", \"E\", \"F\", \"G\", \"H\"]\n",
    "    new_well_name = row[int(entry[\"Plate\"]) - 1] + entry[\"Well\"][1:]\n",
    "    entry[\"Well\"] = new_well_name\n",
    "    return entry"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = variant_df.apply(single_plate_annotation, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.colors as mcolors\n",
    "\n",
    "# Define rows and columns for the 96-well plate\n",
    "rows = list('ABCDEFGH')\n",
    "columns = range(1, 13)\n",
    "\n",
    "# Generate well identifiers\n",
    "wells = [f\"{row}{col:01d}\" for row in rows for col in columns]\n",
    "\n",
    "# # Generate random reads data\n",
    "# np.random.seed(0)  # For reproducible results\n",
    "# reads = np.random.randint(50, 1000, size=len(wells))  # Random reads between 50 and 1000\n",
    "\n",
    "# # Create the DataFrame\n",
    "# df = pd.DataFrame({\n",
    "#     'Well': wells,\n",
    "#     'Reads': reads\n",
    "# })\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "variant_df.to_csv(\"data/Tam_LQV_variant_df.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create the figure and axis\n",
    "fig, ax = plt.subplots(figsize=(16, 9))\n",
    "\n",
    "# Initialize the data matrix with NaN for the heatmap\n",
    "heatmap_data = np.full((8, 12), np.nan)\n",
    "\n",
    "# Fill in the data matrix with 'Reads' values\n",
    "for _, row in df.iterrows():\n",
    "    well_row = 'ABCDEFGH'.index(row['Well'][0])\n",
    "    well_col = int(row['Well'][1:]) - 1\n",
    "    heatmap_data[well_row, well_col] = row['count']\n",
    "\n",
    "# Labels\n",
    "col_labels = ['1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12']\n",
    "row_labels = ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H']\n",
    "\n",
    "# Heatmap\n",
    "im = ax.imshow(heatmap_data, cmap=\"Blues\", aspect='auto')\n",
    "\n",
    "# Colorbar\n",
    "cbar = ax.figure.colorbar(im, ax=ax)\n",
    "cbar.ax.set_ylabel(\"# of Reads\", rotation=90, va=\"top\", size=14)\n",
    "\n",
    "# White grid\n",
    "ax.set_xticks(np.arange(-.5, 12, 1), minor=True)\n",
    "ax.set_yticks(np.arange(-.5, 8, 1), minor=True)\n",
    "ax.grid(which=\"minor\", color=\"w\", linestyle='-', linewidth=2)\n",
    "ax.tick_params(which=\"minor\", size=0)\n",
    "\n",
    "# Annotations with variants\n",
    "for (i, j), val in np.ndenumerate(heatmap_data):\n",
    "    if not np.isnan(val):  # Only annotate non-NaN values\n",
    "        variant_text = '\\n'.join(df.loc[df['Well'] == f'{row_labels[i]}{j+1:01d}', 'Variant'].values[0])\n",
    "        ax.text(j, i, variant_text, ha='center', va='center', color='red', fontsize=8)\n",
    "    elif np.isnan(val):\n",
    "        ax.text(j, i, \"NA\", ha='center', va='center', color='black', fontsize=8)\n",
    "\n",
    "\n",
    "\n",
    "# Set the ticks\n",
    "ax.set_xticks(np.arange(len(col_labels)))\n",
    "ax.set_yticks(np.arange(len(row_labels)))\n",
    "ax.set_xticklabels(col_labels)\n",
    "ax.set_yticklabels(row_labels)\n",
    "\n",
    "# Turn off the tick labels if not needed\n",
    "ax.tick_params(axis=u'both', which=u'both',length=0)\n",
    "\n",
    "plt.savefig(\"figures/heatmap_tam.png\", dpi=300, bbox_inches='tight')\n",
    "\n",
    "plt.show()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "variant_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Backup\n",
    "\n",
    "# Calculate for each pair chi-squared test\n",
    "#nb_positions = extract_positions_from_vcf(f'{path_to_bam}/filtered_variants.vcf')\n",
    "comb = list(combinations(nb_positions, 2))\n",
    "\n",
    "\n",
    "# Initialize a matrix with zeros for storing the -log10(p-values)\n",
    "jaccard_index_meta = {\"Pair\" : [], \"Jaccard Index Dict\" : []}\n",
    "\n",
    "bases_df = get_bases_from_pileup(bam_file, reference, nb_positions)\n",
    "\n",
    "for pair in comb:\n",
    "    \n",
    "    max_nucleotide_pair1, _ = get_most_common_base(pair[0], heatmap_data, template[pair[0] - 1])\n",
    "    max_nucleotide_pair2, _ = get_most_common_base(pair[1], heatmap_data, template[pair[1] - 1])\n",
    "\n",
    "    ref_nuc_pair1 = template[pair[0] - 1]\n",
    "    ref_nuc_pair2 = template[pair[1] - 1]\n",
    "\n",
    "\n",
    "    # Mutations\n",
    "    A_and_B = len(bases_df[(bases_df[pair[0]] == max_nucleotide_pair1) & (bases_df[pair[1]] == max_nucleotide_pair2)])+ 1\n",
    "    A_wo_B = len(bases_df[(bases_df[pair[0]] == max_nucleotide_pair1) & (bases_df[pair[1]] != max_nucleotide_pair2)]) + 1\n",
    "    B_wo_A = len(bases_df[(bases_df[pair[0]] != max_nucleotide_pair1) & (bases_df[pair[1]] == max_nucleotide_pair2)]) + 1\n",
    "    wo_A_and_B = len(bases_df[(bases_df[pair[0]] != max_nucleotide_pair1) & (bases_df[pair[1]] != max_nucleotide_pair2)]) + 1\n",
    "\n",
    "\n",
    "    # A wt vs Mutations\n",
    "    A_wt_and_B = len(bases_df[(bases_df[pair[0]] == ref_nuc_pair1) & (bases_df[pair[1]] == max_nucleotide_pair2)]) + 1\n",
    "    A_wt_wo_B = len(bases_df[(bases_df[pair[0]] == ref_nuc_pair1) & (bases_df[pair[1]] != max_nucleotide_pair2)]) + 1\n",
    "    B_wo_A_wt = len(bases_df[(bases_df[pair[0]] != ref_nuc_pair1) & (bases_df[pair[1]] == max_nucleotide_pair2)]) + 1\n",
    "    wo_A_wt_and_B = len(bases_df[(bases_df[pair[0]] != ref_nuc_pair1) & (bases_df[pair[1]] != max_nucleotide_pair2)]) + 1\n",
    "\n",
    "    # B wt vs Mutations\n",
    "    A_and_B_wt  = len(bases_df[(bases_df[pair[0]] == max_nucleotide_pair1) & (bases_df[pair[1]] == ref_nuc_pair2)]) + 1\n",
    "    A_wo_B_wt = len(bases_df[(bases_df[pair[0]] == max_nucleotide_pair1) & (bases_df[pair[1]] != ref_nuc_pair2)]) + 1\n",
    "    B_wo_A_wt = len(bases_df[(bases_df[pair[0]] != max_nucleotide_pair1) & (bases_df[pair[1]] == ref_nuc_pair2)]) + 1\n",
    "    wo_A_and_B_wt = len(bases_df[(bases_df[pair[0]] != max_nucleotide_pair1) & (bases_df[pair[1]] != ref_nuc_pair2)]) + 1\n",
    "\n",
    "    # A wt vs B wt\n",
    "    A_wt_and_B_wt = len(bases_df[(bases_df[pair[0]] == ref_nuc_pair1) & (bases_df[pair[1]] == ref_nuc_pair2)]) + 1\n",
    "    A_wt_wo_B_wt = len(bases_df[(bases_df[pair[0]] == ref_nuc_pair1) & (bases_df[pair[1]] != ref_nuc_pair2)]) + 1\n",
    "    B_wo_A_wt_wt = len(bases_df[(bases_df[pair[0]] != ref_nuc_pair1) & (bases_df[pair[1]] == ref_nuc_pair2)]) + 1\n",
    "    wo_A_wt_and_B_wt = len(bases_df[(bases_df[pair[0]] != ref_nuc_pair1) & (bases_df[pair[1]] != ref_nuc_pair2)]) + 1\n",
    "\n",
    "\n",
    "    # Jaccardi index dictionary\n",
    "    print(\"Pair:\", pair)\n",
    "    jaccard_index_dict = {  \"A_and_B\": A_and_B / (A_and_B + A_wo_B + B_wo_A),\n",
    "                            \"A_and_B_wt\": A_and_B_wt / (A_and_B_wt + A_wo_B_wt + B_wo_A_wt),\n",
    "                            \"A_wt_and_B\": A_wt_and_B / (A_wt_and_B + A_wt_wo_B + B_wo_A_wt),\n",
    "                            \"A_wt_and_B_wt\": A_wt_and_B_wt / (A_wt_and_B_wt + A_wt_wo_B_wt + B_wo_A_wt_wt)}\n",
    "\n",
    "    jaccard_index_meta[\"Pair\"].append(pair)\n",
    "    jaccard_index_meta[\"Jaccard Index Dict\"].append(jaccard_index_dict)\n",
    "\n",
    "    print(jaccard_index_dict)\n",
    "\n",
    "\n",
    "    #print(\"Jaccard index:\", jaccard_index)\n",
    "\n",
    "    # contingency_table = pd.DataFrame({\n",
    "    #     'B': [A_and_B, A_wo_B],\n",
    "    #     'Not B': [B_wo_A, wo_A_and_B]\n",
    "    # }, index=['A', 'Not A'])\n",
    "\n",
    "    contingency_table = pd.DataFrame({\n",
    "        \"mut_A\": [jaccard_index_dict[\"A_and_B\"], jaccard_index_dict[\"A_and_B_wt\"]], \n",
    "        \"wt_A\": [jaccard_index_dict[\"A_wt_and_B\"], jaccard_index_dict[\"A_wt_and_B_wt\"]]}, index=['mut_B', 'wt_B'])\n",
    "\n",
    "    print(\"Contingency Table for pair\", pair)\n",
    "    print(contingency_table)\n",
    "\n",
    "\n",
    "# Get final combination based on threshold\n",
    "jacc_ind_the = 0.7\n",
    "overlap_the = 0.7\n",
    "\n",
    "# Where non wt are max and above threshold\n",
    "mutation = {\"Type\" : [], \"Position\" : [], \"Variant\" : [], \"Frequency\" : []}\n",
    "\n",
    "for i in range(len(jaccard_index_meta[\"Pair\"])):\n",
    "    max_type = max(jaccard_index_meta[\"Jaccard Index Dict\"][i], key=jaccard_index_meta[\"Jaccard Index Dict\"][i].get)\n",
    "    if (max(jaccard_index_meta[\"Jaccard Index Dict\"][i].values()) > jacc_ind_the) and (max_type != \"A_wt_and_B_wt\"):\n",
    "        mutation[\"Type\"].append(max_type)\n",
    "        if max_type == \"A_and_B\":\n",
    "            mutation[\"Position\"] += [jaccard_index_meta[\"Pair\"][i][0], jaccard_index_meta[\"Pair\"][i][1]]\n",
    "            mutation[\"Variant\"] +=  [f'{ref_nuc_pair1}{jaccard_index_meta[\"Pair\"][i][0]}{max_nucleotide_pair1}', f'{ref_nuc_pair2}{jaccard_index_meta[\"Pair\"][i][1]}{max_nucleotide_pair2}']\n",
    "            mutation[\"Frequency\"] += [jaccard_index_meta[\"Jaccard Index Dict\"][i][\"A_and_B\"], jaccard_index_meta[\"Jaccard Index Dict\"][i][\"A_and_B\"]]\n",
    "        elif max_type == \"A_and_B_wt\":\n",
    "            mutation[\"Position\"].append(jaccard_index_meta[\"Pair\"][i][0])\n",
    "            mutation[\"Variant\"].append(f'{ref_nuc_pair1}{jaccard_index_meta[\"Pair\"][i][0]}{max_nucleotide_pair1}')\n",
    "            mutation[\"Frequency\"].append(jaccard_index_meta[\"Jaccard Index Dict\"][i][\"A_and_B_wt\"])\n",
    "        elif max_type == \"A_wt_and_B\":\n",
    "            mutation[\"Position\"].append(jaccard_index_meta[\"Pair\"][i][1])\n",
    "            mutation[\"Variant\"].append(f'{ref_nuc_pair1}{jaccard_index_meta[\"Pair\"][i][1]}{max_nucleotide_pair1}')\n",
    "            mutation[\"Frequency\"].append(jaccard_index_meta[\"Jaccard Index Dict\"][i][\"A_wt_and_B\"])\n",
    "        \n",
    "    else:\n",
    "        continue\n",
    "\n",
    "        \n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# Masking the lower triangle of the matrix\n",
    "mask = np.tri(pval_matrix.shape[0], k=-1)\n",
    "\n",
    "# Plotting the heatmap\n",
    "plt.figure(figsize=(6, 4))\n",
    "#sns.heatmap(odds_ratio_matrix, cmap='coolwarm', mask=mask, linewidths=0.5, linecolor='white')\n",
    "sns.heatmap(jaccard_index_matrix, cmap='coolwarm', mask=mask, linewidths=0.5, linecolor='white')\n",
    "sns.xticklabels = nb_positions\n",
    "sns.yticklabels = nb_positions\n",
    "#Label the axes\n",
    "plt.xlabel('Position')\n",
    "plt.ylabel('Position')\n",
    "plt.title(\"Jacard Index\")\n",
    "plt.show()\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "minion",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
